// kds-react/build/es/components/Breadcrumbs/Breadcrumbs.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React from 'react';
import cx from 'classnames';
import Text from '../Text/Text.js';
import BreadcrumbLink from './_BreadcrumbLink.js';

/**
 * Breadcrumbs are used to show the User where they are inside the site's hierarchy and give them the ability to move to a higher level.
 */

var Breadcrumbs = function Breadcrumbs(_ref) {
  var breadcrumbs = _ref.breadcrumbs,
      inverse = _ref.inverse,
      withRouter = _ref.withRouter,
      className = _ref.className;
  return React.createElement("nav", {
    "aria-labelledby": "breadcrumb-heading",
    className: cx(className, 'kds-Breadcrumb')
  }, React.createElement("h1", {
    id: "breadcrumb-heading",
    className: "screen-reader"
  }, "Breadcrumb"), React.createElement("ol", {
    className: "list-reset"
  }, breadcrumbs.filter(function (b) {
    return b;
  }).map(function (breadcrumb, index) {
    var component = breadcrumb.component,
        title = breadcrumb.title,
        url = breadcrumb.url,
        linkProps = _objectWithoutProperties(breadcrumb, ["component", "title", "url"]);

    return index !== breadcrumbs.length - 1 ? React.createElement(BreadcrumbLink, _extends({
      inverse: inverse,
      withRouter: withRouter,
      title: title,
      url: url,
      key: index,
      component: component
    }, linkProps)) : React.createElement("li", {
      className: "inline",
      key: url
    }, React.createElement(Text, {
      className: cx({
        'text-default-50': inverse
      }),
      size: "l",
      tabIndex: "0",
      "aria-current": "page"
    }, title));
  })));
};

Breadcrumbs.defaultProps = {
  breadcrumbs: [],
  inverse: false,
  withRouter: false
};

export default Breadcrumbs;
