// kds-react/build/cjs/components/Breadcrumbs/Breadcrumbs.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var Text = require('../Text/Text.js');
var _BreadcrumbLink = require('./_BreadcrumbLink.js');

/**
 * Breadcrumbs are used to show the User where they are inside the site's hierarchy and give them the ability to move to a higher level.
 */

var Breadcrumbs = function Breadcrumbs(_ref) {
  var breadcrumbs = _ref.breadcrumbs,
      inverse = _ref.inverse,
      withRouter = _ref.withRouter,
      className = _ref.className;
  return React__default.createElement("nav", {
    "aria-labelledby": "breadcrumb-heading",
    className: cx(className, 'kds-Breadcrumb')
  }, React__default.createElement("h1", {
    id: "breadcrumb-heading",
    className: "screen-reader"
  }, "Breadcrumb"), React__default.createElement("ol", {
    className: "list-reset"
  }, breadcrumbs.filter(function (b) {
    return b;
  }).map(function (breadcrumb, index) {
    var component = breadcrumb.component,
        title = breadcrumb.title,
        url = breadcrumb.url,
        linkProps = _objectWithoutProperties(breadcrumb, ["component", "title", "url"]);

    return index !== breadcrumbs.length - 1 ? React__default.createElement(_BreadcrumbLink.default, _extends({
      inverse: inverse,
      withRouter: withRouter,
      title: title,
      url: url,
      key: index,
      component: component
    }, linkProps)) : React__default.createElement("li", {
      className: "inline",
      key: url
    }, React__default.createElement(Text.default, {
      className: cx({
        'text-default-50': inverse
      }),
      size: "l",
      tabIndex: "0",
      "aria-current": "page"
    }, title));
  })));
};

Breadcrumbs.defaultProps = {
  breadcrumbs: [],
  inverse: false,
  withRouter: false
};

exports.default = Breadcrumbs;
