// kds-react/build/cjs/components/Breadcrumbs/SchemaForBreadcrumbs.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var JsonLd = require('../JsonLd/JsonLd.js');

var generateData = function generateData() {
  var breadcrumbs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  return {
    '@context': 'https://schema.org',
    '@type': 'BreadcrumbList',
    itemListElement: breadcrumbs && breadcrumbs.map(function (breadcrumb, index) {
      return index === 0 ? {
        '@type': 'ListItem',
        position: index + 1,
        name: breadcrumb && breadcrumb.title,
        item: "https://www.kroger.com"
      } : {
        '@type': 'ListItem',
        position: index + 1,
        name: breadcrumb && breadcrumb.title,
        item: "https://www.kroger.com".concat(breadcrumb && breadcrumb.url)
      };
    })
  };
};

var SchemaForBreadcrumbs = function SchemaForBreadcrumbs(_ref) {
  var breadcrumbs = _ref.breadcrumbs;
  return breadcrumbs ? React__default.createElement(JsonLd.default, {
    data: generateData(breadcrumbs)
  }) : null;
};

exports.default = SchemaForBreadcrumbs;
exports.generateData = generateData;
