// kds-react/build/cjs/components/Stepper/Stepper.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
require('../Icon/icons/AcceptEntry.js');
require('../Icon/icons/AccountSummary.js');
require('../Icon/icons/Account.js');
require('../Icon/icons/AddCard.js');
require('../Icon/icons/AddColumn.js');
require('../Icon/icons/AddPatient.js');
require('../Icon/icons/AddPlusCard.js');
require('../Icon/icons/AddPrescription.js');
require('../Icon/icons/AdultBeverage.js');
require('../Icon/icons/Airplane.js');
require('../Icon/icons/Android.js');
require('../Icon/icons/Animals.js');
require('../Icon/icons/ApiAuthorize.js');
require('../Icon/icons/ApiProduct.js');
require('../Icon/icons/Approved.js');
require('../Icon/icons/Ar.js');
require('../Icon/icons/Associate.js');
require('../Icon/icons/Asterisk.js');
require('../Icon/icons/Audience.js');
require('../Icon/icons/AutoRefill.js');
require('../Icon/icons/Baby.js');
require('../Icon/icons/Backstock.js');
require('../Icon/icons/BagBarcode.js');
require('../Icon/icons/Bags.js');
require('../Icon/icons/BarGraph.js');
require('../Icon/icons/BarLineGraph.js');
require('../Icon/icons/Barcode.js');
require('../Icon/icons/BasketTime.js');
require('../Icon/icons/Basket.js');
require('../Icon/icons/BeautyPersonalCare.js');
require('../Icon/icons/Beta.js');
require('../Icon/icons/BlankBox.js');
require('../Icon/icons/BuyWireless.js');
require('../Icon/icons/Calendar.js');
require('../Icon/icons/Camera.js');
require('../Icon/icons/Candy.js');
require('../Icon/icons/CapsLock.js');
require('../Icon/icons/CardView.js');
require('../Icon/icons/CarouselView.js');
require('../Icon/icons/CaretLeft.js');
require('../Icon/icons/CaretRight.js');
require('../Icon/icons/Cart.js');
require('../Icon/icons/CashBack.js');
require('../Icon/icons/CashRegister.js');
require('../Icon/icons/CcAmex.js');
require('../Icon/icons/CcDinersClub.js');
require('../Icon/icons/CcDiscover.js');
require('../Icon/icons/CcMastercard.js');
require('../Icon/icons/CcVisa.js');
require('../Icon/icons/CellPhone.js');
require('../Icon/icons/CenterAlign.js');
require('../Icon/icons/CheckMark.js');
require('../Icon/icons/ChevronDown.js');
require('../Icon/icons/ChevronUp.js');
require('../Icon/icons/CivicOrganizations.js');
require('../Icon/icons/CleaningHousehold.js');
require('../Icon/icons/Clock.js');
require('../Icon/icons/Close.js');
require('../Icon/icons/CodePhone.js');
require('../Icon/icons/Coffee.js');
require('../Icon/icons/ContainsDairy.js');
require('../Icon/icons/ContainsArtificialIngredients.js');
require('../Icon/icons/Comment.js');
require('../Icon/icons/CompanyDirect.js');
require('../Icon/icons/ContractViewport.js');
require('../Icon/icons/ControlPayment.js');
require('../Icon/icons/Convenience.js');
require('../Icon/icons/CollapseAll.js');
require('../Icon/icons/Copy.js');
require('../Icon/icons/Coupons.js');
require('../Icon/icons/CreateNew.js');
require('../Icon/icons/CreditCard.js');
require('../Icon/icons/CurrentLocationAndroid.js');
require('../Icon/icons/CurrentLocationIos.js');
require('../Icon/icons/DamagedItem.js');
require('../Icon/icons/DarkLight.js');
require('../Icon/icons/Dashboard.js');
require('../Icon/icons/DebitCard.js');
require('../Icon/icons/DeliBakery.js');
require('../Icon/icons/Delivery.js');
require('../Icon/icons/DiabeticNeeds.js');
require('../Icon/icons/Directions.js');
require('../Icon/icons/Diversity.js');
require('../Icon/icons/DownArrow.js');
require('../Icon/icons/Download.js');
require('../Icon/icons/DragHandle.js');
require('../Icon/icons/Draggable.js');
require('../Icon/icons/DrugInteractions.js');
require('../Icon/icons/DrugSearch.js');
require('../Icon/icons/Ebt.js');
require('../Icon/icons/Edit.js');
require('../Icon/icons/Education.js');
require('../Icon/icons/EggFree.js');
require('../Icon/icons/Ellipsis.js');
require('../Icon/icons/Email.js');
require('../Icon/icons/EmergencyContact.js');
require('../Icon/icons/Environmental.js');
require('../Icon/icons/Error.js');
require('../Icon/icons/ExclusiveSavings.js');
require('../Icon/icons/Exit.js');
require('../Icon/icons/ExpandAll.js');
require('../Icon/icons/ExpandViewport.js');
require('../Icon/icons/ExpenseReport.js');
require('../Icon/icons/Expiration.js');
require('../Icon/icons/Export.js');
require('../Icon/icons/External.js');
require('../Icon/icons/EyeSlash.js');
require('../Icon/icons/Eye.js');
require('../Icon/icons/Failure.js');
require('../Icon/icons/FairService.js');
require('../Icon/icons/FaithBased.js');
require('../Icon/icons/Faq.js');
require('../Icon/icons/FatFree.js');
require('../Icon/icons/FavoriteSolid.js');
require('../Icon/icons/Favorite.js');
require('../Icon/icons/Fee.js');
require('../Icon/icons/Feedback.js');
require('../Icon/icons/File.js');
require('../Icon/icons/Filters.js');
require('../Icon/icons/Flag.js');
require('../Icon/icons/FlagSolid.js');
require('../Icon/icons/FlashOff.js');
require('../Icon/icons/FlashOn.js');
require('../Icon/icons/Flashlight.js');
require('../Icon/icons/Floral.js');
require('../Icon/icons/FmJewelers.js');
require('../Icon/icons/FuelPump.js');
require('../Icon/icons/Fuel.js');
require('../Icon/icons/GanttChart.js');
require('../Icon/icons/GiftCardMall.js');
require('../Icon/icons/GiftCard.js');
require('../Icon/icons/GlutenFree.js');
require('../Icon/icons/GoodService.js');
require('../Icon/icons/GoogleHome.js');
require('../Icon/icons/GridView.js');
require('../Icon/icons/GroceriesHome.js');
require('../Icon/icons/Groceries.js');
require('../Icon/icons/HandSanitizer.js');
require('../Icon/icons/HamburgerMenuNative.js');
require('../Icon/icons/HamburgerMenu.js');
require('../Icon/icons/HealthNutrition.js');
require('../Icon/icons/HealthWellness.js');
require('../Icon/icons/HeatEat.js');
require('../Icon/icons/Hierarchy.js');
require('../Icon/icons/HighProtein.js');
require('../Icon/icons/History.js');
require('../Icon/icons/HomeElectronics.js');
require('../Icon/icons/Home.js');
require('../Icon/icons/HungerRelief.js');
require('../Icon/icons/ImageMissing.js');
require('../Icon/icons/Import.js');
require('../Icon/icons/Info.js');
require('../Icon/icons/Ingredients.js');
require('../Icon/icons/InProgress.js');
require('../Icon/icons/InsuranceInfo.js');
require('../Icon/icons/IosShare.js');
require('../Icon/icons/Ios.js');
require('../Icon/icons/Kosher.js');
require('../Icon/icons/KrogerCreditCard.js');
require('../Icon/icons/KrogerRewards.js');
require('../Icon/icons/LactoseFree.js');
require('../Icon/icons/Layers.js');
require('../Icon/icons/LeftAlign.js');
require('../Icon/icons/LeftArrow.js');
require('../Icon/icons/Leftover.js');
require('../Icon/icons/LightBulb.js');
require('../Icon/icons/Link.js');
require('../Icon/icons/ListView.js');
require('../Icon/icons/LoadedSavings.js');
require('../Icon/icons/LocationA.js');
require('../Icon/icons/LocationB.js');
require('../Icon/icons/LocationRx.js');
require('../Icon/icons/Location.js');
require('../Icon/icons/Locked.js');
require('../Icon/icons/LockerPasscode.js');
require('../Icon/icons/LowFat.js');
require('../Icon/icons/Map.js');
require('../Icon/icons/MealKit.js');
require('../Icon/icons/MeatSeafood.js');
require('../Icon/icons/MedicationToolkit.js');
require('../Icon/icons/Members.js');
require('../Icon/icons/Merge.js');
require('../Icon/icons/MilitaryPoliceFire.js');
var Minus = require('../Icon/icons/Minus.js');
require('../Icon/icons/MissingItem.js');
require('../Icon/icons/MoneyServices.js');
require('../Icon/icons/Moon.js');
require('../Icon/icons/MoreOptions.js');
require('../Icon/icons/Natural.js');
require('../Icon/icons/Newest.js');
require('../Icon/icons/NoFees.js');
require('../Icon/icons/NonGmo.js');
require('../Icon/icons/NotAvailable.js');
require('../Icon/icons/Notification.js');
require('../Icon/icons/NuroDeliver.js');
require('../Icon/icons/NuroPickup.js');
require('../Icon/icons/NutFree.js');
require('../Icon/icons/Offline.js');
require('../Icon/icons/OnlinePayment.js');
require('../Icon/icons/OptUp.js');
require('../Icon/icons/Organic.js');
require('../Icon/icons/OtherMisc.js');
require('../Icon/icons/Oven.js');
require('../Icon/icons/Paleo.js');
require('../Icon/icons/Partnership.js');
require('../Icon/icons/Paste.js');
require('../Icon/icons/PatientAccess.js');
require('../Icon/icons/PatientWellness.js');
require('../Icon/icons/PaymentSuccess.js');
require('../Icon/icons/Pencil.js');
require('../Icon/icons/PersonHome.js');
require('../Icon/icons/PersonStore.js');
require('../Icon/icons/PersonalCare.js');
require('../Icon/icons/PetMedications.js');
require('../Icon/icons/PharmacyNotifications.js');
require('../Icon/icons/PharmacySearch.js');
require('../Icon/icons/Pharmacy.js');
require('../Icon/icons/Phone.js');
require('../Icon/icons/Pickup.js');
require('../Icon/icons/PieChart.js');
require('../Icon/icons/PiggyBank.js');
require('../Icon/icons/PillIdentifier.js');
require('../Icon/icons/PlusCard.js');
var Plus = require('../Icon/icons/Plus.js');
require('../Icon/icons/PoorService.js');
require('../Icon/icons/PoorSubstitution.js');
require('../Icon/icons/Popularity.js');
require('../Icon/icons/PreferredCard.js');
require('../Icon/icons/PrescriptionHistory.js');
require('../Icon/icons/PrescriptionStatus.js');
require('../Icon/icons/Print.js');
require('../Icon/icons/PrivacyPolicy.js');
require('../Icon/icons/Produce.js');
require('../Icon/icons/PromotionOnAd.js');
require('../Icon/icons/PromotionOnDisplay.js');
require('../Icon/icons/Promotions.js');
require('../Icon/icons/Publish.js');
require('../Icon/icons/PurchaseHistory.js');
require('../Icon/icons/PushNotification.js');
require('../Icon/icons/PushPin.js');
require('../Icon/icons/PushPinSolid.js');
require('../Icon/icons/PuzzlePiece.js');
require('../Icon/icons/QrCode.js');
require('../Icon/icons/RadioChecked.js');
require('../Icon/icons/RadioUnchecked.js');
require('../Icon/icons/RatingsHalf.js');
require('../Icon/icons/RatingsSolid.js');
require('../Icon/icons/Ratings.js');
require('../Icon/icons/ReWeigh.js');
require('../Icon/icons/React.js');
require('../Icon/icons/Recall.js');
require('../Icon/icons/Receipt.js');
require('../Icon/icons/RecentSearch.js');
require('../Icon/icons/Recipes.js');
require('../Icon/icons/Recycle.js');
require('../Icon/icons/Redo.js');
require('../Icon/icons/RefillFromList.js');
require('../Icon/icons/RefillPrograms.js');
require('../Icon/icons/RejectEntry.js');
require('../Icon/icons/Relevance.js');
require('../Icon/icons/RemoveColumn.js');
require('../Icon/icons/Repurchase.js');
require('../Icon/icons/Reset.js');
require('../Icon/icons/Restrooms.js');
require('../Icon/icons/Returns.js');
require('../Icon/icons/Revert.js');
require('../Icon/icons/Rewards.js');
require('../Icon/icons/RightArrow.js');
require('../Icon/icons/RxBarcode.js');
require('../Icon/icons/Route.js');
require('../Icon/icons/Sandwich.js');
require('../Icon/icons/SaveTime.js');
require('../Icon/icons/Scan.js');
require('../Icon/icons/ScanBagGo.js');
require('../Icon/icons/ScheduleAppointment.js');
require('../Icon/icons/Search.js');
require('../Icon/icons/SecurityQuestions.js');
require('../Icon/icons/SecurityTag.js');
require('../Icon/icons/Send.js');
require('../Icon/icons/Settings.js');
require('../Icon/icons/Share.js');
require('../Icon/icons/ShipToHome.js');
require('../Icon/icons/Ship.js');
require('../Icon/icons/ShopInStore.js');
require('../Icon/icons/ShopOnline.js');
require('../Icon/icons/Shop.js');
require('../Icon/icons/ShoppingList.js');
require('../Icon/icons/SkipToEnd.js');
require('../Icon/icons/SkipToStart.js');
require('../Icon/icons/SocialServices.js');
require('../Icon/icons/Sort.js');
require('../Icon/icons/SortAz.js');
require('../Icon/icons/SortZa.js');
require('../Icon/icons/SoyFree.js');
require('../Icon/icons/Status.js');
require('../Icon/icons/Stencil.js');
require('../Icon/icons/Stethoscope.js');
require('../Icon/icons/Stock.js');
require('../Icon/icons/StoreLocker.js');
require('../Icon/icons/Store.js');
require('../Icon/icons/Success.js');
require('../Icon/icons/Sun.js');
require('../Icon/icons/Supercharged.js');
require('../Icon/icons/TableView.js');
require('../Icon/icons/TermsConditions.js');
require('../Icon/icons/ThumbsDown.js');
require('../Icon/icons/ThumbsUp.js');
require('../Icon/icons/Transfer.js');
require('../Icon/icons/TransferCircle.js');
require('../Icon/icons/TransferCircleSolid.js');
var Trash = require('../Icon/icons/Trash.js');
require('../Icon/icons/Trending.js');
require('../Icon/icons/Undo.js');
require('../Icon/icons/Unlocked.js');
require('../Icon/icons/UpArrow.js');
require('../Icon/icons/Upload.js');
require('../Icon/icons/VaccinationsMadeEasy.js');
require('../Icon/icons/Vaccinations.js');
require('../Icon/icons/Value.js');
require('../Icon/icons/Vegan.js');
require('../Icon/icons/Vegetarian.js');
require('../Icon/icons/Voice.js');
require('../Icon/icons/Wallet.js');
require('../Icon/icons/Warning.js');
require('../Icon/icons/WeeklyAd.js');
require('../Icon/icons/Weigh.js');
require('../Icon/icons/WorldwideAcceptance.js');
require('../Icon/icons/WrongItem.js');
require('../Icon/icons/YouthPrograms.js');
require('../Icon/icons/YtdSavings.js');
require('../Icon/icons/ZoomIn.js');
require('../Icon/icons/ZoomOut.js');
var functionUtils = require('../../utils/functionUtils.js');
var Text = require('../Text/Text.js');
var Button = require('../Button/Button.js');
var numberUtils = require('../../utils/numberUtils.js');

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

function cleanValue(oldValue, newValue, min, max) {
  newValue = parseInt(newValue) || 0;
  var value = oldValue;

  if (min <= newValue && newValue <= max) {
    value = newValue;
  } else if (newValue > max) {
    value = max;
  } else if (newValue < min) {
    var direction = newValue - value;
    value = direction > 0 ? min // This means we're adding, snap to min
    : 0; // if we're subtracting, snap to zero
  }

  return value;
}

var kinds = ['primary', 'secondary'];
/**
 * `Stepper` lets users increment and decrement quantities. Use them for things like adding/removing items to/from a cart or list.
 *
 * It handles local state and passes the current value to the `onChange` callback for you to use as needed.
 *
 * Because it wraps the buttons and input in a `div`, additional props will be spread onto the `div`.
 *
 * Stepper comes built in with a 400ms debounce on increment/decrement and input actions.
 *
 * ### Accessibility
 *
 * This component meets the criteria for a `spinbutton` as defined in <a href="https://www.w3.org/TR/wai-aria-1.1/#spinbutton" rel="external">WAI-ARIA 1.1: 5.4 Definition of Roles: spinbutton</a>
 *
 * And it follows patterns informed by <a href="https://www.w3.org/TR/wai-aria-practices-1.1/#spinbutton" rel="external">WAI-ARIA Authoring Practices: Design Patterns and Widgets: Spinbutton</a>
 *
 * >**Note:** IE 11 doesn't natively support keyboard controls for input `type="number"` other than manual numeric entry. Because the KDS defers to native browser behavior for input controls, we do not add this functionality.
 *
 * > ```Message { "props": { "kind": "warning" } }
 * > *THIS COMPONENT IS ON A DEPRECATION PATH.*
 * >
 */

var Stepper =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Stepper, _PureComponent);

  var _super = _createSuper(Stepper);

  function Stepper() {
    var _this;

    _classCallCheck(this, Stepper);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      clampedMax: numberUtils.clamp(_this.props.max, 1, 999),
      inputIsFocusedAndZero: false,
      shouldSetFocusOnCTA: false,
      value: _this.props.value || 0,
      updatedValue: _this.props.value || 0,
      locked: false
    });

    _defineProperty(_assertThisInitialized(_this), "ctaButtonRef", React__default.createRef());

    _defineProperty(_assertThisInitialized(_this), "lastEmittedValue", _this.state.value);

    _defineProperty(_assertThisInitialized(_this), "debouncedOnChange", functionUtils.debounce(function (value) {
      _this.props.onChange(value);

      _this.lastEmittedValue = value;
      _this.state.locked = false;
    }, 800));

    _defineProperty(_assertThisInitialized(_this), "incrBtnRef", React__default.createRef());

    _defineProperty(_assertThisInitialized(_this), "getMinimumQuantity", function () {
      return Math.max(1, _this.props.minimumQuantity);
    });

    _defineProperty(_assertThisInitialized(_this), "isOutOfRange", function (value) {
      return value < 0 || value > _this.state.clampedMax;
    });

    _defineProperty(_assertThisInitialized(_this), "isLessThanMinQuantity", function (value) {
      return value > 0 && value < _this.getMinimumQuantity();
    });

    _defineProperty(_assertThisInitialized(_this), "isInvalid", function (value) {
      return _this.isOutOfRange(value) || _this.isLessThanMinQuantity(value);
    });

    _defineProperty(_assertThisInitialized(_this), "isEqualToLastEmittedValue", function (value) {
      return value === _this.lastEmittedValue;
    });

    _defineProperty(_assertThisInitialized(_this), "tryOnChangeCallback", function () {
      var debounce = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;

      /* Get the newly updated value from state. */
      var value = _this.state.value;
      /* If value is equal to last emitted value, out of range, or otherwise invalid, don't execute callback function. */

      if (!_this.isInvalid(value) && !_this.isEqualToLastEmittedValue(value)) {
        if (value === 0 || !debounce) {
          /* If the value is 0 we want to fire onChange immediately,
           * so we cancel the debounced version and call it directly. */
          _this.state.locked = false;

          _this.debouncedOnChange.cancel();

          _this.props.onChange(value);

          _this.lastEmittedValue = value;
        } else {
          _this.state.locked = true;

          _this.debouncedOnChange(value);
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "cleanValue", function (newValue) {
      return cleanValue(_this.state.value, newValue, _this.getMinimumQuantity(), _this.props.max);
    });

    _defineProperty(_assertThisInitialized(_this), "focusOnCTA", function () {
      var callToAction = _this.props.callToAction;
      var _this$state = _this.state,
          inputIsFocusedAndZero = _this$state.inputIsFocusedAndZero,
          shouldSetFocusOnCTA = _this$state.shouldSetFocusOnCTA,
          value = _this$state.value;
      /* If callToAction prop is set
       * and value is 0
       * and current state indicates that focus should be set on the CTA button
       * and current state indicates that the input is NOT focused with a zero value, ... */

      callToAction && value === 0 && shouldSetFocusOnCTA && !inputIsFocusedAndZero &&
      /* then in the next frame, ... */
      setTimeout(function () {
        /* if the CTA button exists and is focusable, then set focus on it. */
        _this.ctaButtonRef && _this.ctaButtonRef.current && typeof _this.ctaButtonRef.current.focus === 'function' && _this.ctaButtonRef.current.focus();
        _this.state.shouldSetFocusOnCTA = false;
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onCTAButtonClick", function () {
      _this.setState({
        value: _this.getMinimumQuantity()
      }, function () {
        _this.tryOnChangeCallback(_this.state.value);
        /* We want the input to take focus when it renders in place of the "call to action" button (for accessibility) */


        _this.incrBtnRef.current.focus();
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onDecrementClick", function () {
      _this.setState(function (_ref) {
        var value = _ref.value;

        var cleanedValue = _this.cleanValue(value - 1);

        return {
          /** Snap manually entered (typed) invalid value to clamped max. */
          value: cleanedValue,
          inputIsFocusedAndZero: false,
          shouldSetFocusOnCTA: cleanedValue - 1 < _this.getMinimumQuantity()
        };
      }, _this.tryOnChangeCallback);
    });

    _defineProperty(_assertThisInitialized(_this), "onIncrementClick", function () {
      _this.setState(function (_ref2) {
        var value = _ref2.value;

        var cleanedValue = _this.cleanValue(value + 1);

        return {
          /** Snap manually entered (typed) invalid value to 0. */
          value: cleanedValue
        };
      }, _this.tryOnChangeCallback);
    });

    _defineProperty(_assertThisInitialized(_this), "onInputChange", function (e) {
      /* In this event we "clean" the value first because a user could have typed anything into the input. */
      var cleanedValue = parseInt(e.target.value) || 0;

      _this.setState({
        inputIsFocusedAndZero: cleanedValue === 0,
        // indicate when input has focus and zero value
        value: cleanedValue
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onInputKeyPress", function (e) {
      var value = _this.state.value;
      /* We want users to have to "commit" a zero typed directly into the focused input. */

      if (e.key === 'Enter') {
        _this.setState({
          /* We fake blur by telling state that the input is not focused, so we don't trigger the onInputBlur handler; */
          inputIsFocusedAndZero: false,

          /* In this event, we want to transfer focus back to the CTA button, unlike in a true blur event. */
          shouldSetFocusOnCTA: value === 0
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onInputBlur", function () {
      /* On blur, set state accordingly ... */
      _this.setState({
        inputIsFocusedAndZero: false,
        shouldSetFocusOnCTA: false
      }, function () {
        return _this.tryOnChangeCallback(false);
      });
    });

    return _this;
  }

  _createClass(Stepper, [{
    key: "componentDidUpdate",

    /* We want the "call to action" button to take focus when value is decreased to zero (for accessibility) */
    value: function componentDidUpdate(prevProps) {
      if (this.props.value !== prevProps.value) {
        this.lastEmittedValue = this.props.value;
      }

      this.focusOnCTA();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
          clampedMax = _this$state2.clampedMax,
          inputIsFocusedAndZero = _this$state2.inputIsFocusedAndZero,
          value = _this$state2.value;

      var _this$props = this.props,
          className = _this$props.className,
          compact = _this$props.compact,
          callToAction = _this$props.callToAction,
          description = _this$props.description,
          destructive = _this$props.destructive,
          fullWidth = _this$props.fullWidth,
          what = _this$props.what,
          kind = _this$props.kind,
          childrenOmitted = _this$props.children,
          maxOmitted = _this$props.max,
          minimumQuantityOmitted = _this$props.minimumQuantity,
          onChangeOmitted = _this$props.onChange,
          valueOmitted = _this$props.value,
          divProps = _objectWithoutProperties(_this$props, ["className", "compact", "callToAction", "description", "destructive", "fullWidth", "what", "kind", "children", "max", "minimumQuantity", "onChange", "value"]);

      var destructiveDecrement = destructive && value <= this.getMinimumQuantity();
      var hideCTAButton = value !== 0 || inputIsFocusedAndZero;
      var hideStepper = callToAction && !hideCTAButton;
      var disableDecrementBtn = value <= 0;
      var disableIncrementBtn = value >= clampedMax;
      var isInvalid = this.isInvalid(value);
      var invalidMessage = isInvalid ? this.isLessThanMinQuantity(value) ? "".concat(value, " is less than the minimum quantity.") : "".concat(value, " is more than the maximum quantity.") : '';
      var rangeLimitMessage = '';
      /** aria-live message for out-of-range values. */

      if (value <= 0) {
        rangeLimitMessage = 'Minimum Quantity Reached,';
      } else if (value >= clampedMax) {
        rangeLimitMessage = 'Maximum Quantity Reached,';
      }

      return React__default.createElement("div", _extends({}, divProps, {
        className: cx('kds-Stepper', className)
      }), callToAction && React__default.createElement(Button.default, {
        "aria-label": "".concat(callToAction, ": ").concat(what),
        className: cx('kds-Stepper-ctaButton', {
          'kds-Stepper-ctaButton--compact': compact,
          'w-full': fullWidth,
          hidden: hideCTAButton
        }),
        compact: compact,
        "data-testid": "kds-Stepper-ctaButton",
        onClick: this.onCTAButtonClick,
        ref: this.ctaButtonRef,
        kind: kinds.includes(kind) ? kind : undefined
      }, callToAction), React__default.createElement("div", {
        className: cx('kds-Stepper-wrapper', {
          'text-center': fullWidth,
          hidden: hideStepper
        }),
        "data-testid": "kds-Stepper-wrapper"
      }, React__default.createElement(Button.default, {
        "aria-label": "Decrement ".concat(what).concat(description ? ", ".concat(description) : ''),
        tabIndex: disableDecrementBtn ? -1 : 0,
        compact: compact,
        "data-testid": "kds-Stepper-decButton",
        disabled: disableDecrementBtn,
        kind: destructiveDecrement ? 'secondary-negative' : 'cancel',
        onClick: this.onDecrementClick
      }, destructiveDecrement ? React__default.createElement(Trash.IconTrash, null) : React__default.createElement(Minus.IconMinus, null)), React__default.createElement("div", {
        className: cx('kds-Stepper-inputWrapper', {
          'mx-8': compact
        }),
        "data-testid": "kds-Stepper-inputWrapper"
      }, React__default.createElement("input", {
        "aria-invalid": isInvalid,
        "aria-live": "polite",
        "aria-label": "".concat(invalidMessage || rangeLimitMessage || 'Quantity of', " ").concat(what).concat(description ? ", ".concat(description) : ''),
        className: cx('kds-Stepper-input', {
          'kds-Stepper-input--descriptive': description
        }),
        "data-testid": "kds-Stepper-input",
        onBlur: this.onInputBlur,
        onChange: this.onInputChange,
        onKeyPress: this.onInputKeyPress,
        "aria-valuemax": clampedMax,
        max: clampedMax,
        "aria-valuemin": "0",
        min: "0",
        type: "text",
        inputMode: "numeric",
        "aria-valuenow": value,
        value:
        /* toString() is necessary to force the browser to strip leading zeros. ¯\_(ツ)_/¯ */
        value.toString(),
        style: {
          /* The input is resized to be wide enough for the value with a little wiggle room */
          width: "".concat(value.toString().length * 1.5, "ch")
        },
        tabIndex:
        /* This is required to fix a bug where this button is not perceived as focusable by some versions of VO + Safari. */
        0
      }), description && React__default.createElement(Text.default, {
        "aria-hidden": "true",
        className: "kds-Stepper-description",
        size: "s"
      }, description)), React__default.createElement(Button.default, {
        "aria-label": "Increment ".concat(what).concat(description ? ", ".concat(description) : ''),
        tabIndex: disableIncrementBtn ? -1 : 0,
        ref: this.incrBtnRef,
        compact: compact,
        "data-testid": "kds-Stepper-incButton",
        disabled: disableIncrementBtn,
        onClick: this.onIncrementClick
      }, React__default.createElement(Plus.IconPlus, null))));
    }
  }], [{
    key: "getDerivedStateFromProps",

    /* Handle updates to updatedValue prop and set value on state */
    value: function getDerivedStateFromProps(_ref3, _ref4) {
      var propValue = _ref3.value;
      var updatedValue = _ref4.updatedValue,
          locked = _ref4.locked;

      if (!locked && propValue !== updatedValue) {
        /* In this event we "clean" the value first because a user could have typed anything into the input. */
        var cleanedValue = parseInt(propValue) || 0;
        return {
          inputIsFocusedAndZero: cleanedValue !== 0,
          // indicate when input has focus and zero value
          value: cleanedValue,
          updatedValue: propValue
        };
      } else {
        return {
          updatedValue: propValue
        };
      }
    }
  }]);

  return Stepper;
}(React.PureComponent);

_defineProperty(Stepper, "defaultProps", {
  callToAction: '',
  compact: false,
  description: '',
  destructive: false,
  fullWidth: false,
  value: 0,
  onChange: function onChange() {},
  max: 999,
  minimumQuantity: 1,
  kind: 'primary'
});

exports.default = Stepper;
