// kds-react/build/cjs/components/TagGroup/TagGroup.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var _slicedToArray = _interopDefault(require('@babel/runtime/helpers/slicedToArray'));

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

/**
 * TagGroup render a list of one or more Tag components.
 */
var TagGroup =
/*#__PURE__*/
function (_Component) {
  _inherits(TagGroup, _Component);

  var _super = _createSuper(TagGroup);

  function TagGroup(props) {
    var _this;

    _classCallCheck(this, TagGroup);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "removeTag", function (idxToRemove) {
      var state = _this.state.tagState.get(idxToRemove);

      state.isRemoved = true;

      _this.state.tagState.set(idxToRemove, state);

      _this.moveFocus(idxToRemove);
    });

    _defineProperty(_assertThisInitialized(_this), "moveFocus", function (originalIdx) {
      //  looking for possible prev element
      var prevExistingTags = [];
      var nextExistingTags = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = _this.state.tagState[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var _step$value = _slicedToArray(_step.value, 2),
              key = _step$value[0],
              value = _step$value[1];

          if (key < originalIdx && !value.isRemoved) {
            prevExistingTags.push(value);
          } else if (key > originalIdx && !value.isRemoved) {
            nextExistingTags.push(value);
          } else {
            continue;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      if (prevExistingTags.length > 0) {
        prevExistingTags[prevExistingTags.length - 1].ref.current.focus();
        return;
      }

      if (nextExistingTags.length > 0) {
        nextExistingTags[0].ref.current.focus();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "isNonMixedTagsKinds", function () {
      var childrenArr = React__default.Children.toArray(_this.props.children).filter(function (child) {
        return child.type.displayName && child.type.displayName === 'Tag';
      });

      if (childrenArr.length === 0) {
        return true;
      }

      if (childrenArr[0].props.kind === 'interactive' || childrenArr[0].props.kind === 'action') {
        return childrenArr.every(function (child) {
          return child.type !== 'Tag' || child.type === 'Tag' && (child.props.kind === 'interactive' || child.props.kind === 'action');
        });
      }

      if (childrenArr[0].props.kind !== 'interactive' && childrenArr[0].props.kind !== 'action') {
        return childrenArr.every(function (child) {
          return child.props.kind !== 'interactive' && child.props.kind !== 'action';
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "renderChildren", function () {
      var childrenArr = React__default.Children.toArray(_this.props.children);
      return childrenArr.map(function (child, index) {
        if (child.type.displayName && child.type.displayName === 'Tag') {
          return React__default.cloneElement(child, {
            tag: 'li',
            role: 'listitem',
            'data-index': index,
            ref: _this.state.tagState.get(index).ref,
            onRemove: _this.removeTag
          });
        }

        return child;
      });
    });

    _this.state = {
      tagState: new Map()
    };
    _this.props.children && _this.props.children.map(function (child, index) {
      if (child.type.displayName && child.type.displayName === 'Tag') {
        _this.state.tagState.set(index, {
          isRemoved: false,
          //  existing in DOM
          ref: React__default.createRef()
        });
      }
    });
    return _this;
  }

  _createClass(TagGroup, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children,
          ariaLabel = _this$props['aria-label'],
          layout = _this$props.layout,
          props = _objectWithoutProperties(_this$props, ["className", "children", "aria-label", "layout"]); // if (!this.isNonMixedTagsKinds() && process.env.NODE_ENV === 'development') {
      //   console.warn('action kind and non-action kind Tag should not be mixed in TagGroup')
      //   return null
      // }


      return React__default.createElement("section", _extends({
        className: cx('kds-TagGroup', {
          'with-scroll': layout === 'scroll'
        }, className),
        "aria-label": "Tag Group: ".concat(ariaLabel)
      }, props), React__default.createElement("ul", {
        className: "kds-TagGroup-list"
      }, this.renderChildren()));
    }
  }]);

  return TagGroup;
}(React.Component);

_defineProperty(TagGroup, "defaultProps", {
  className: '',
  layout: 'wrap'
});

exports.default = TagGroup;
