// kds-react/build/cjs/utils/createReverseContext.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}
/**
 * Similar to React.createContext() but instead of the ancestor passing data to it's descendant,
 * the descendant passes data to a single ancestor.
 *
 * @param defaultValue The default value for the reverse context
 * @returns {{ReverseConsumer: ReverseConsumer, ReverseProvider: ReverseProvider}}
 * @example
 const ClassNameContext = createReverseContext(false)

 const Parent = props => (
   <ClassNameContext.ReverseConsumer>
     {className => <div {...props} className={`Parent ${className}`} />}
   </ClassNameContext.ReverseConsumer>
 )

 const Child = props => (
   <ClassNameContext.ReverseProvider value={props.className}>
     <div className={`Child ${props.className}`} />
   </ClassNameContext.ReverseProvider>
 )

 // This will add the child className to the parent as well
 const Example = () =>
   <Parent>
     <Child className={`Child-${this.state.num}`} />
   </Parent>
 */

function createReverseContext(defaultValue) {
  var NativeContext = React__default.createContext();

  var ReverseProvider =
  /*#__PURE__*/
  function (_React$Component) {
    _inherits(ReverseProvider, _React$Component);

    var _super = _createSuper(ReverseProvider);

    function ReverseProvider() {
      var _this;

      _classCallCheck(this, ReverseProvider);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = _super.call.apply(_super, [this].concat(args));

      _defineProperty(_assertThisInitialized(_this), "setParentState", function () {});

      return _this;
    }

    _createClass(ReverseProvider, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        this.setParentState(this.props.value);
      }
    }, {
      key: "componentDidUpdate",
      value: function componentDidUpdate() {
        this.setParentState(this.props.value);
      }
    }, {
      key: "render",
      value: function render() {
        var _this2 = this;

        return React__default.createElement(NativeContext.Consumer, null, function (setState) {
          if (setState) _this2.setParentState = setState;
          return _this2.props.children;
        });
      }
    }]);

    return ReverseProvider;
  }(React__default.Component);

  var ReverseConsumer =
  /*#__PURE__*/
  function (_React$Component2) {
    _inherits(ReverseConsumer, _React$Component2);

    var _super2 = _createSuper(ReverseConsumer);

    function ReverseConsumer() {
      var _this3;

      _classCallCheck(this, ReverseConsumer);

      for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      _this3 = _super2.call.apply(_super2, [this].concat(args));

      _defineProperty(_assertThisInitialized(_this3), "state", {
        value: defaultValue
      });

      _defineProperty(_assertThisInitialized(_this3), "setParentState", function (value) {
        return _this3.setState({
          value: value === undefined ? defaultValue : value
        });
      });

      return _this3;
    }

    _createClass(ReverseConsumer, [{
      key: "render",
      value: function render() {
        return React__default.createElement(NativeContext.Provider, {
          value: this.setParentState
        }, this.props.children(this.state.value));
      }
    }]);

    return ReverseConsumer;
  }(React__default.Component);

  return {
    ReverseConsumer: ReverseConsumer,
    ReverseProvider: ReverseProvider
  };
}

exports.createReverseContext = createReverseContext;
