// kds-react/build/cjs/utils/functionUtils.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var debounce = function debounce(func, wait) {
  var timeout;

  function debounced() {
    var context = this;
    var args = arguments;

    var later = function later() {
      timeout = null;
      func.apply(context, args);
    };

    clearTimeout(timeout);
    timeout = setTimeout(later, wait || 200);

    if (!timeout) {
      func.apply(context, args);
    }
  }

  function cancel() {
    clearTimeout(timeout);
  }

  debounced.cancel = cancel;
  return debounced;
};
var callAll = function callAll() {
  for (var _len3 = arguments.length, fns = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
    fns[_key3] = arguments[_key3];
  }

  return function () {
    for (var _len4 = arguments.length, args = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
      args[_key4] = arguments[_key4];
    }

    return fns.forEach(function (fn) {
      return callSafely(fn).apply(void 0, args);
    });
  };
};
var callSafely = function callSafely(maybeFn) {
  return function () {
    if (typeof maybeFn === 'function') {
      return maybeFn.apply(void 0, arguments);
    } else {
      return maybeFn;
    }
  };
};

exports.callAll = callAll;
exports.callSafely = callSafely;
exports.debounce = debounce;
