// kds-react/build/cjs/components/QuantityStepper/QuantityStepper.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
require('../Icon/icons/AcceptEntry.js');
require('../Icon/icons/AccountSummary.js');
require('../Icon/icons/Account.js');
require('../Icon/icons/AddCard.js');
require('../Icon/icons/AddColumn.js');
require('../Icon/icons/AddPatient.js');
require('../Icon/icons/AddPlusCard.js');
require('../Icon/icons/AddPrescription.js');
require('../Icon/icons/AdultBeverage.js');
require('../Icon/icons/Airplane.js');
require('../Icon/icons/Android.js');
require('../Icon/icons/Animals.js');
require('../Icon/icons/ApiAuthorize.js');
require('../Icon/icons/ApiProduct.js');
require('../Icon/icons/Approved.js');
require('../Icon/icons/Ar.js');
require('../Icon/icons/Associate.js');
require('../Icon/icons/Asterisk.js');
require('../Icon/icons/Audience.js');
require('../Icon/icons/AutoRefill.js');
require('../Icon/icons/Baby.js');
require('../Icon/icons/Backstock.js');
require('../Icon/icons/BagBarcode.js');
require('../Icon/icons/Bags.js');
require('../Icon/icons/BarGraph.js');
require('../Icon/icons/BarLineGraph.js');
require('../Icon/icons/Barcode.js');
require('../Icon/icons/BasketTime.js');
require('../Icon/icons/Basket.js');
require('../Icon/icons/BeautyPersonalCare.js');
require('../Icon/icons/Beta.js');
require('../Icon/icons/BlankBox.js');
require('../Icon/icons/BuyWireless.js');
require('../Icon/icons/Calendar.js');
require('../Icon/icons/Camera.js');
require('../Icon/icons/Candy.js');
require('../Icon/icons/CapsLock.js');
require('../Icon/icons/CardView.js');
require('../Icon/icons/CarouselView.js');
require('../Icon/icons/CaretLeft.js');
require('../Icon/icons/CaretRight.js');
require('../Icon/icons/Cart.js');
require('../Icon/icons/CashBack.js');
require('../Icon/icons/CashRegister.js');
require('../Icon/icons/CcAmex.js');
require('../Icon/icons/CcDinersClub.js');
require('../Icon/icons/CcDiscover.js');
require('../Icon/icons/CcMastercard.js');
require('../Icon/icons/CcVisa.js');
require('../Icon/icons/CellPhone.js');
require('../Icon/icons/CenterAlign.js');
require('../Icon/icons/CheckMark.js');
require('../Icon/icons/ChevronDown.js');
require('../Icon/icons/ChevronUp.js');
require('../Icon/icons/CivicOrganizations.js');
require('../Icon/icons/CleaningHousehold.js');
require('../Icon/icons/Clock.js');
require('../Icon/icons/Close.js');
require('../Icon/icons/CodePhone.js');
require('../Icon/icons/Coffee.js');
require('../Icon/icons/ContainsDairy.js');
require('../Icon/icons/ContainsArtificialIngredients.js');
require('../Icon/icons/Comment.js');
require('../Icon/icons/CompanyDirect.js');
require('../Icon/icons/ContractViewport.js');
require('../Icon/icons/ControlPayment.js');
require('../Icon/icons/Convenience.js');
require('../Icon/icons/CollapseAll.js');
require('../Icon/icons/Copy.js');
require('../Icon/icons/Coupons.js');
require('../Icon/icons/CreateNew.js');
require('../Icon/icons/CreditCard.js');
require('../Icon/icons/CurrentLocationAndroid.js');
require('../Icon/icons/CurrentLocationIos.js');
require('../Icon/icons/DamagedItem.js');
require('../Icon/icons/DarkLight.js');
require('../Icon/icons/Dashboard.js');
require('../Icon/icons/DebitCard.js');
require('../Icon/icons/DeliBakery.js');
require('../Icon/icons/Delivery.js');
require('../Icon/icons/DiabeticNeeds.js');
require('../Icon/icons/Directions.js');
require('../Icon/icons/Diversity.js');
require('../Icon/icons/DownArrow.js');
require('../Icon/icons/Download.js');
require('../Icon/icons/DragHandle.js');
require('../Icon/icons/Draggable.js');
require('../Icon/icons/DrugInteractions.js');
require('../Icon/icons/DrugSearch.js');
require('../Icon/icons/Ebt.js');
require('../Icon/icons/Edit.js');
require('../Icon/icons/Education.js');
require('../Icon/icons/EggFree.js');
require('../Icon/icons/Ellipsis.js');
require('../Icon/icons/Email.js');
require('../Icon/icons/EmergencyContact.js');
require('../Icon/icons/Environmental.js');
require('../Icon/icons/Error.js');
require('../Icon/icons/ExclusiveSavings.js');
require('../Icon/icons/Exit.js');
require('../Icon/icons/ExpandAll.js');
require('../Icon/icons/ExpandViewport.js');
require('../Icon/icons/ExpenseReport.js');
require('../Icon/icons/Expiration.js');
require('../Icon/icons/Export.js');
require('../Icon/icons/External.js');
require('../Icon/icons/EyeSlash.js');
require('../Icon/icons/Eye.js');
require('../Icon/icons/Failure.js');
require('../Icon/icons/FairService.js');
require('../Icon/icons/FaithBased.js');
require('../Icon/icons/Faq.js');
require('../Icon/icons/FatFree.js');
require('../Icon/icons/FavoriteSolid.js');
require('../Icon/icons/Favorite.js');
require('../Icon/icons/Fee.js');
require('../Icon/icons/Feedback.js');
require('../Icon/icons/File.js');
require('../Icon/icons/Filters.js');
require('../Icon/icons/Flag.js');
require('../Icon/icons/FlagSolid.js');
require('../Icon/icons/FlashOff.js');
require('../Icon/icons/FlashOn.js');
require('../Icon/icons/Flashlight.js');
require('../Icon/icons/Floral.js');
require('../Icon/icons/FmJewelers.js');
require('../Icon/icons/FuelPump.js');
require('../Icon/icons/Fuel.js');
require('../Icon/icons/GanttChart.js');
require('../Icon/icons/GiftCardMall.js');
require('../Icon/icons/GiftCard.js');
require('../Icon/icons/GlutenFree.js');
require('../Icon/icons/GoodService.js');
require('../Icon/icons/GoogleHome.js');
require('../Icon/icons/GridView.js');
require('../Icon/icons/GroceriesHome.js');
require('../Icon/icons/Groceries.js');
require('../Icon/icons/HandSanitizer.js');
require('../Icon/icons/HamburgerMenuNative.js');
require('../Icon/icons/HamburgerMenu.js');
require('../Icon/icons/HealthNutrition.js');
require('../Icon/icons/HealthWellness.js');
require('../Icon/icons/HeatEat.js');
require('../Icon/icons/Hierarchy.js');
require('../Icon/icons/HighProtein.js');
require('../Icon/icons/History.js');
require('../Icon/icons/HomeElectronics.js');
require('../Icon/icons/Home.js');
require('../Icon/icons/HungerRelief.js');
require('../Icon/icons/ImageMissing.js');
require('../Icon/icons/Import.js');
require('../Icon/icons/Info.js');
require('../Icon/icons/Ingredients.js');
require('../Icon/icons/InProgress.js');
require('../Icon/icons/InsuranceInfo.js');
require('../Icon/icons/IosShare.js');
require('../Icon/icons/Ios.js');
require('../Icon/icons/Kosher.js');
require('../Icon/icons/KrogerCreditCard.js');
require('../Icon/icons/KrogerRewards.js');
require('../Icon/icons/LactoseFree.js');
require('../Icon/icons/Layers.js');
require('../Icon/icons/LeftAlign.js');
require('../Icon/icons/LeftArrow.js');
require('../Icon/icons/Leftover.js');
require('../Icon/icons/LightBulb.js');
require('../Icon/icons/Link.js');
require('../Icon/icons/ListView.js');
require('../Icon/icons/LoadedSavings.js');
require('../Icon/icons/LocationA.js');
require('../Icon/icons/LocationB.js');
require('../Icon/icons/LocationRx.js');
require('../Icon/icons/Location.js');
require('../Icon/icons/Locked.js');
require('../Icon/icons/LockerPasscode.js');
require('../Icon/icons/LowFat.js');
require('../Icon/icons/Map.js');
require('../Icon/icons/MealKit.js');
require('../Icon/icons/MeatSeafood.js');
require('../Icon/icons/MedicationToolkit.js');
require('../Icon/icons/Members.js');
require('../Icon/icons/Merge.js');
require('../Icon/icons/MilitaryPoliceFire.js');
var Minus = require('../Icon/icons/Minus.js');
require('../Icon/icons/MissingItem.js');
require('../Icon/icons/MoneyServices.js');
require('../Icon/icons/Moon.js');
require('../Icon/icons/MoreOptions.js');
require('../Icon/icons/Natural.js');
require('../Icon/icons/Newest.js');
require('../Icon/icons/NoFees.js');
require('../Icon/icons/NonGmo.js');
require('../Icon/icons/NotAvailable.js');
require('../Icon/icons/Notification.js');
require('../Icon/icons/NuroDeliver.js');
require('../Icon/icons/NuroPickup.js');
require('../Icon/icons/NutFree.js');
require('../Icon/icons/Offline.js');
require('../Icon/icons/OnlinePayment.js');
require('../Icon/icons/OptUp.js');
require('../Icon/icons/Organic.js');
require('../Icon/icons/OtherMisc.js');
require('../Icon/icons/Oven.js');
require('../Icon/icons/Paleo.js');
require('../Icon/icons/Partnership.js');
require('../Icon/icons/Paste.js');
require('../Icon/icons/PatientAccess.js');
require('../Icon/icons/PatientWellness.js');
require('../Icon/icons/PaymentSuccess.js');
require('../Icon/icons/Pencil.js');
require('../Icon/icons/PersonHome.js');
require('../Icon/icons/PersonStore.js');
require('../Icon/icons/PersonalCare.js');
require('../Icon/icons/PetMedications.js');
require('../Icon/icons/PharmacyNotifications.js');
require('../Icon/icons/PharmacySearch.js');
require('../Icon/icons/Pharmacy.js');
require('../Icon/icons/Phone.js');
require('../Icon/icons/Pickup.js');
require('../Icon/icons/PieChart.js');
require('../Icon/icons/PiggyBank.js');
require('../Icon/icons/PillIdentifier.js');
require('../Icon/icons/PlusCard.js');
var Plus = require('../Icon/icons/Plus.js');
require('../Icon/icons/PoorService.js');
require('../Icon/icons/PoorSubstitution.js');
require('../Icon/icons/Popularity.js');
require('../Icon/icons/PreferredCard.js');
require('../Icon/icons/PrescriptionHistory.js');
require('../Icon/icons/PrescriptionStatus.js');
require('../Icon/icons/Print.js');
require('../Icon/icons/PrivacyPolicy.js');
require('../Icon/icons/Produce.js');
require('../Icon/icons/PromotionOnAd.js');
require('../Icon/icons/PromotionOnDisplay.js');
require('../Icon/icons/Promotions.js');
require('../Icon/icons/Publish.js');
require('../Icon/icons/PurchaseHistory.js');
require('../Icon/icons/PushNotification.js');
require('../Icon/icons/PushPin.js');
require('../Icon/icons/PushPinSolid.js');
require('../Icon/icons/PuzzlePiece.js');
require('../Icon/icons/QrCode.js');
require('../Icon/icons/RadioChecked.js');
require('../Icon/icons/RadioUnchecked.js');
require('../Icon/icons/RatingsHalf.js');
require('../Icon/icons/RatingsSolid.js');
require('../Icon/icons/Ratings.js');
require('../Icon/icons/ReWeigh.js');
require('../Icon/icons/React.js');
require('../Icon/icons/Recall.js');
require('../Icon/icons/Receipt.js');
require('../Icon/icons/RecentSearch.js');
require('../Icon/icons/Recipes.js');
require('../Icon/icons/Recycle.js');
require('../Icon/icons/Redo.js');
require('../Icon/icons/RefillFromList.js');
require('../Icon/icons/RefillPrograms.js');
require('../Icon/icons/RejectEntry.js');
require('../Icon/icons/Relevance.js');
require('../Icon/icons/RemoveColumn.js');
require('../Icon/icons/Repurchase.js');
require('../Icon/icons/Reset.js');
require('../Icon/icons/Restrooms.js');
require('../Icon/icons/Returns.js');
require('../Icon/icons/Revert.js');
require('../Icon/icons/Rewards.js');
require('../Icon/icons/RightArrow.js');
require('../Icon/icons/RxBarcode.js');
require('../Icon/icons/Route.js');
require('../Icon/icons/Sandwich.js');
require('../Icon/icons/SaveTime.js');
require('../Icon/icons/Scan.js');
require('../Icon/icons/ScanBagGo.js');
require('../Icon/icons/ScheduleAppointment.js');
require('../Icon/icons/Search.js');
require('../Icon/icons/SecurityQuestions.js');
require('../Icon/icons/SecurityTag.js');
require('../Icon/icons/Send.js');
require('../Icon/icons/Settings.js');
require('../Icon/icons/Share.js');
require('../Icon/icons/ShipToHome.js');
require('../Icon/icons/Ship.js');
require('../Icon/icons/ShopInStore.js');
require('../Icon/icons/ShopOnline.js');
require('../Icon/icons/Shop.js');
require('../Icon/icons/ShoppingList.js');
require('../Icon/icons/SkipToEnd.js');
require('../Icon/icons/SkipToStart.js');
require('../Icon/icons/SocialServices.js');
require('../Icon/icons/Sort.js');
require('../Icon/icons/SortAz.js');
require('../Icon/icons/SortZa.js');
require('../Icon/icons/SoyFree.js');
require('../Icon/icons/Status.js');
require('../Icon/icons/Stencil.js');
require('../Icon/icons/Stethoscope.js');
require('../Icon/icons/Stock.js');
require('../Icon/icons/StoreLocker.js');
require('../Icon/icons/Store.js');
require('../Icon/icons/Success.js');
require('../Icon/icons/Sun.js');
require('../Icon/icons/Supercharged.js');
require('../Icon/icons/TableView.js');
require('../Icon/icons/TermsConditions.js');
require('../Icon/icons/ThumbsDown.js');
require('../Icon/icons/ThumbsUp.js');
require('../Icon/icons/Transfer.js');
require('../Icon/icons/TransferCircle.js');
require('../Icon/icons/TransferCircleSolid.js');
var Trash = require('../Icon/icons/Trash.js');
require('../Icon/icons/Trending.js');
require('../Icon/icons/Undo.js');
require('../Icon/icons/Unlocked.js');
require('../Icon/icons/UpArrow.js');
require('../Icon/icons/Upload.js');
require('../Icon/icons/VaccinationsMadeEasy.js');
require('../Icon/icons/Vaccinations.js');
require('../Icon/icons/Value.js');
require('../Icon/icons/Vegan.js');
require('../Icon/icons/Vegetarian.js');
require('../Icon/icons/Voice.js');
require('../Icon/icons/Wallet.js');
require('../Icon/icons/Warning.js');
require('../Icon/icons/WeeklyAd.js');
require('../Icon/icons/Weigh.js');
require('../Icon/icons/WorldwideAcceptance.js');
require('../Icon/icons/WrongItem.js');
require('../Icon/icons/YouthPrograms.js');
require('../Icon/icons/YtdSavings.js');
require('../Icon/icons/ZoomIn.js');
require('../Icon/icons/ZoomOut.js');
var generationUtils = require('../../utils/generationUtils.js');
var Text = require('../Text/Text.js');
var Button = require('../Button/Button.js');

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

function parseInteger(value) {
  return Math.max(parseInt(value) || 0, 0);
}

function snapValue(oldValue, newValue, minQty, max) {
  newValue = parseInteger(newValue);
  var value = oldValue;

  if (minQty <= newValue && newValue <= max) {
    value = newValue;
  } else if (newValue > max) {
    value = max;
  } else if (newValue < minQty) {
    var direction = newValue - value;
    value = direction > 0 ? minQty // This means we're adding, snap to minimum quantity
    : 0; // if we're subtracting, snap to zero
  }

  return value;
}

function focusRef(ref) {
  ref && ref.current && typeof ref.current.focus === 'function' && setTimeout(function () {
    return ref.current && ref.current.focus();
  }); // made asynchronous to prevent triggering CTA click event when focused on from manualy entering value of 0
}

var kinds = ['primary', 'secondary'];
/**
 * `QuantityStepper` lets users increment and decrement quantities. Use them for things like adding/removing items to/from a cart or list.
 *
 * QuantityStepper updates optimistically by updating local state immediately and triggering `onChange` callback passing current value from local state.
 *
 * **It is the responsibility of the consumer to debounce/handle any necessary resolutions for any functions or async requests triggered by `onChange`**
 *
 * Because it wraps the buttons and input in a `div`, additional props will be spread onto the `div`.
 *
 * > ```Message { "props": { "kind": "warning" } }
 * > *THIS COMPONENT IS EXPERIMENTAL.*
 * >
 */

var QuantityStepper =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(QuantityStepper, _PureComponent);

  var _super = _createSuper(QuantityStepper);

  function QuantityStepper() {
    var _this;

    _classCallCheck(this, QuantityStepper);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      lastPropValue: parseInteger(_this.props.value),
      // last value passed in from props
      lastSetValue: parseInteger(_this.props.value),
      // last value set by onChange
      value: parseInteger(_this.props.value),
      // value displayed in input
      showCTAButton: parseInteger(_this.props.value) === 0
    });

    _defineProperty(_assertThisInitialized(_this), "inputId", _this.props.id ? "".concat(_this.props.id, "-input") : "kds-QuantityStepper-input-".concat(generationUtils.generateUniqueString()));

    _defineProperty(_assertThisInitialized(_this), "ctaButtonRef", React__default.createRef());

    _defineProperty(_assertThisInitialized(_this), "incrBtnRef", React__default.createRef());

    _defineProperty(_assertThisInitialized(_this), "getMinimumQuantity", function () {
      return Math.max(1, _this.props.minimumQuantity);
    });

    _defineProperty(_assertThisInitialized(_this), "isLessThanMinQuantity", function (value) {
      return value > 0 && value < _this.getMinimumQuantity();
    });

    _defineProperty(_assertThisInitialized(_this), "isInvalid", function (value) {
      var isOutOfRange = value < 0 || value > _this.props.max;
      return isOutOfRange || _this.isLessThanMinQuantity(value);
    });

    _defineProperty(_assertThisInitialized(_this), "setValue", function (value) {
      var snap = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var callback = arguments.length > 2 ? arguments[2] : undefined;
      var event = arguments.length > 3 ? arguments[3] : undefined;
      var lastSetValue = _this.state.lastSetValue;
      var _this$props = _this.props,
          max = _this$props.max,
          onChange = _this$props.onChange;
      var newValue = snap ? snapValue(lastSetValue, value, _this.getMinimumQuantity(), max) : value;
      var showCTAButton = newValue === 0;
      event.persist(); // required so the 'event' object can be accessed in async callbacks (https://reactjs.org/docs/events.html#event-pooling)

      if (newValue !== lastSetValue) {
        _this.setState({
          lastSetValue: newValue,
          value: newValue,
          showCTAButton: showCTAButton
        }, function () {
          onChange(newValue, lastSetValue, event);
          callback && callback();
        });
      } else {
        /*
          snapping value back to an already set value (newValue === lastSetValue)
          we must update 'value' to snapped 'newValue' for the UI
        */
        _this.setState({
          value: newValue
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onCTAButtonClick", function (e) {
      _this.setValue(_this.getMinimumQuantity(), true, function () {
        return focusRef(_this.incrBtnRef);
      }, e);
    });

    _defineProperty(_assertThisInitialized(_this), "onDecrementClick", function (e) {
      _this.setValue(_this.state.value - 1, true, function () {
        return _this.state.showCTAButton && focusRef(_this.ctaButtonRef);
      }, e);
    });

    _defineProperty(_assertThisInitialized(_this), "onIncrementClick", function (e) {
      _this.setValue(_this.state.value + 1, true, null, e);
    });

    _defineProperty(_assertThisInitialized(_this), "onInputKeyDown", function (e) {
      if (e.key === 'Enter') {
        _this.setValue(_this.state.value, null, function () {
          return _this.state.showCTAButton && focusRef(_this.ctaButtonRef);
        }, e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onInputChange", function (e) {
      var value = parseInteger(e.target.value);

      _this.setState({
        value: value
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onInputBlur", function (e) {
      var value = _this.state.value;

      _this.setValue(value, false, null, e);
    });

    return _this;
  }

  _createClass(QuantityStepper, [{
    key: "render",
    value: function render() {
      var _this$state = this.state,
          value = _this$state.value,
          showCTAButton = _this$state.showCTAButton;

      var _this$props2 = this.props,
          className = _this$props2.className,
          compact = _this$props2.compact,
          callToAction = _this$props2.callToAction,
          callToActionProps = _this$props2.callToActionProps,
          decButtonProps = _this$props2.decButtonProps,
          description = _this$props2.description,
          destructive = _this$props2.destructive,
          disabled = _this$props2.disabled,
          fullWidth = _this$props2.fullWidth,
          incButtonProps = _this$props2.incButtonProps,
          what = _this$props2.what,
          kind = _this$props2.kind,
          max = _this$props2.max,
          childrenOmitted = _this$props2.children,
          minimumQuantityOmitted = _this$props2.minimumQuantity,
          onChangeOmitted = _this$props2.onChange,
          valueOmitted = _this$props2.value,
          divProps = _objectWithoutProperties(_this$props2, ["className", "compact", "callToAction", "callToActionProps", "decButtonProps", "description", "destructive", "disabled", "fullWidth", "incButtonProps", "what", "kind", "max", "children", "minimumQuantity", "onChange", "value"]);

      var callToActionClassName = callToActionProps.className,
          ctaProps = _objectWithoutProperties(callToActionProps, ["className"]);

      var decButtonClassName = decButtonProps.className,
          decProps = _objectWithoutProperties(decButtonProps, ["className"]);

      var incButtonClassName = incButtonProps.className,
          incProps = _objectWithoutProperties(incButtonProps, ["className"]);

      var destructiveDecrement = destructive && value <= this.getMinimumQuantity();
      var hideQuantityStepper = callToAction && showCTAButton;
      var isInvalid = this.isInvalid(value);
      var disableDecrementBtn = value <= 0 || disabled;
      var disableIncrementBtn = value >= max || disabled;
      var invalidMessage = isInvalid ? this.isLessThanMinQuantity(value) ? "".concat(value, " is less than the minimum quantity.") : "".concat(value, " is more than the maximum quantity.") : '';
      var rangeLimitMessage = '';
      /** aria-live message for out-of-range values. */

      if (value <= 0) {
        rangeLimitMessage = 'Minimum Quantity Reached,';
      } else if (value >= max) {
        rangeLimitMessage = 'Maximum Quantity Reached,';
      }

      return React__default.createElement("div", _extends({}, divProps, {
        className: cx('kds-QuantityStepper', className, {
          'kds-QuantityStepper--fullWidth': fullWidth,
          'kds-QuantityStepper--compact': compact,
          'is-disabled': disabled
        }),
        "data-testid": "kds-QuantityStepper"
      }), callToAction && React__default.createElement(Button.default, _extends({
        "aria-label": "".concat(callToAction, ": ").concat(what),
        className: cx('kds-QuantityStepper-ctaButton', {
          'kds-QuantityStepper-ctaButton--hidden': !showCTAButton
        }, callToActionClassName),
        compact: compact,
        "data-testid": "kds-QuantityStepper-ctaButton",
        disabled: disabled,
        onClick: this.onCTAButtonClick,
        ref: this.ctaButtonRef,
        kind: kinds.includes(kind) ? kind : undefined
      }, ctaProps), callToAction), React__default.createElement("div", {
        className: cx('kds-QuantityStepper-wrapper', {
          'kds-QuantityStepper-wrapper--hidden': hideQuantityStepper
        }),
        "data-testid": "kds-QuantityStepper-wrapper",
        role: "group"
      }, React__default.createElement(Button.default, _extends({
        "aria-label": "".concat(destructiveDecrement ? 'Remove' : 'Decrement', " ").concat(what).concat(description ? ", ".concat(description) : ''),
        "aria-controls": this.inputId,
        className: cx('kds-QuantityStepper-decButton', decButtonClassName),
        tabIndex: disableDecrementBtn ? -1 : 0,
        compact: compact,
        "data-testid": "kds-QuantityStepper-decButton",
        disabled: disableDecrementBtn,
        kind: destructiveDecrement ? 'secondary-negative' : 'cancel',
        onClick: this.onDecrementClick
      }, decProps), destructiveDecrement ? React__default.createElement(Trash.IconTrash, null) : React__default.createElement(Minus.IconMinus, null)), React__default.createElement("div", {
        className: "kds-QuantityStepper-inputWrapper",
        "data-testid": "kds-QuantityStepper-inputWrapper"
      }, React__default.createElement("input", {
        id: this.inputId,
        "aria-invalid": isInvalid,
        "aria-live": "polite",
        "aria-label": "".concat(invalidMessage || rangeLimitMessage || 'Quantity of', " ").concat(what).concat(description ? ", ".concat(description) : ''),
        className: cx('kds-QuantityStepper-input', {
          'kds-QuantityStepper-input--descriptive': description
        }),
        "data-testid": "kds-QuantityStepper-input",
        disabled: disabled,
        onBlur: this.onInputBlur,
        onChange: this.onInputChange,
        onKeyDown: this.onInputKeyDown,
        "aria-valuemax": max,
        max: max,
        "aria-valuemin": "0",
        min: "0",
        type: "text",
        inputMode: "numeric",
        "aria-valuenow": value,
        value:
        /* toString() is necessary to force the browser to strip leading zeros. ¯\_(ツ)_/¯ */
        value.toString(),
        style: {
          /* The input is resized to be wide enough for the max value with a little wiggle room */
          width: "".concat(String(max).length, ".5ch")
        },
        tabIndex:
        /* This is required to fix a bug where this button is not perceived as focusable by some versions of VO + Safari. */
        0
      }), description && React__default.createElement(Text.default, {
        "aria-hidden": "true",
        className: "kds-QuantityStepper-description",
        size: compact ? 'xs' : 's'
      }, description)), React__default.createElement(Button.default, _extends({
        "aria-label": "Increment ".concat(what).concat(description ? ", ".concat(description) : ''),
        "aria-controls": this.inputId,
        className: cx('kds-QuantityStepper-incButton', incButtonClassName),
        tabIndex: disableIncrementBtn ? -1 : 0,
        ref: this.incrBtnRef,
        compact: compact,
        "data-testid": "kds-QuantityStepper-incButton",
        disabled: disableIncrementBtn,
        onClick: this.onIncrementClick
      }, incProps), React__default.createElement(Plus.IconPlus, null))));
    }
  }], [{
    key: "getDerivedStateFromProps",

    /* Handle updates to value prop and set value on state */
    value: function getDerivedStateFromProps(props, state) {
      var newPropValue = parseInteger(props.value);

      if (newPropValue !== state.lastPropValue) {
        var showCTAButton = newPropValue === 0;
        return {
          value: newPropValue,
          lastSetValue: newPropValue,
          lastPropValue: newPropValue,
          // Set specifically for getDerivedStateFromProps
          showCTAButton: showCTAButton
        };
      }

      return null;
    }
  }]);

  return QuantityStepper;
}(React.PureComponent);

_defineProperty(QuantityStepper, "defaultProps", {
  callToAction: '',
  callToActionProps: {},
  compact: false,
  decButtonProps: {},
  description: '',
  destructive: false,
  disabled: false,
  fullWidth: false,
  incButtonProps: {},
  value: 0,
  onChange: function onChange() {},
  max: 999,
  minimumQuantity: 1,
  kind: 'primary'
});

exports.default = QuantityStepper;
