// kds-react/build/cjs/components/Form/FormCheckbox/FormCheckbox.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var Label = require('../../Label/Label.js');
var Checkbox = require('../../Checkbox/Checkbox.js');
var objectUtils = require('../../../utils/objectUtils.js');
var formUtils = require('../../../utils/formUtils.js');
var _prebuiltValidations = require('../validations/_prebuiltValidations.js');
var FormField = require('../FormField/FormField.js');

var CHECKBOX_REQUIRED_DEFAULT_TEXT = 'Check mark is required.';
var CheckboxWithLabel = React__default.forwardRef(function (_ref, ref) {
  var checkboxLabel = _ref.checkboxLabel,
      name = _ref.name,
      value = _ref.value,
      props = _objectWithoutProperties(_ref, ["checkboxLabel", "name", "value"]);

  return React__default.createElement(Label.default, null, React__default.createElement(Checkbox.default, _extends({
    ref: ref
  }, props, {
    name: formUtils.valOrCamelCasedFallback(name, checkboxLabel),
    value: objectUtils.isTruthyOrZero(value) ? value : checkboxLabel
  })), checkboxLabel);
});
/**
 * `FormCheckbox` will add a checkbox to a [`Form`](./form).
 *
 * > **Note:** `<FormCheckbox />` renders a `<FormField />` with most props spread on.
 * For more details, see the [FormField](./formfield) and [`Checkbox`](./checkbox) components.
 *
 * It comes with accessible form features:
 *
 * - label
 * - validation messaging
 * - aria attributes
 *
 * **Requires *custom-event* polyfill for custom validations to work in IE11** [See Polyfills](./polyfills)
 */

var FormCheckbox = function FormCheckbox(_ref2) {
  var label = _ref2.label,
      required = _ref2.required,
      validations = _ref2.validations,
      props = _objectWithoutProperties(_ref2, ["label", "required", "validations"]);

  return React__default.createElement(FormField.default, _extends({}, props, {
    as: CheckboxWithLabel,
    checkboxLabel: label,
    validations: required ? _prebuiltValidations.addRequiredToValidations(validations, CHECKBOX_REQUIRED_DEFAULT_TEXT) : validations
  }));
};

exports.CHECKBOX_REQUIRED_DEFAULT_TEXT = CHECKBOX_REQUIRED_DEFAULT_TEXT;
exports.default = FormCheckbox;
