// kds-react/build/es/components/Toast/ToastProvider.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import React, { Component } from 'react';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { canUseDOM } from '../../utils/windowUtils.js';
import { generateUniqueString } from '../../utils/generationUtils.js';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import Toast from './Toast.js';
import { Provider } from './_ToastContext.js';
import Portal from '../Portal/_Portal.js';

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}
/** See [`Toast`](./toast) documentation for further details. */

var ToastProvider =
/*#__PURE__*/
function (_Component) {
  _inherits(ToastProvider, _Component);

  var _super = _createSuper(ToastProvider);

  function ToastProvider(props) {
    var _this;

    _classCallCheck(this, ToastProvider);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "timeoutId", null);

    _defineProperty(_assertThisInitialized(_this), "add", function (content) {
      var id = generateUniqueString();

      if (_this.state.toasts.length > 0) {
        _this.remove(_this.state.toasts[0] && _this.state.toasts[0].id);
      }

      _this.setState(function (state) {
        var toasts = state.toasts.slice();
        var toast = {
          content: content,
          id: id
        };
        toasts.push(toast);
        return {
          toasts: toasts
        };
      });

      _this.timeoutId = setTimeout(_this.onDismiss(id), _this.readTime(content['text']));
    });

    _defineProperty(_assertThisInitialized(_this), "readTime", function (str) {
      var words = str.split(' ').length;
      return 5000 + words * 300;
    });

    _defineProperty(_assertThisInitialized(_this), "remove", function (id) {
      clearTimeout(_this.timeoutId);

      _this.setState(function (state) {
        var toasts = state.toasts.filter(function (t) {
          return t.id !== id;
        });
        return {
          toasts: toasts
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onDismiss", function (id) {
      return function () {
        _this.remove(id);
      };
    });

    _this.state = {
      toasts: []
    };
    return _this;
  }
  /** The `add` function can be wrapped around an element that will trigger a `Toast` message. */


  _createClass(ToastProvider, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var add = this.add;
      var toasts = this.state.toasts;
      return React.createElement(Provider, {
        value: add
      }, this.props.children, canUseDOM() ? React.createElement(Portal, {
        id: "toast"
      }, React.createElement(TransitionGroup, {
        className: "kds-ToastGroup"
      }, toasts.map(function (_ref) {
        var content = _ref.content,
            id = _ref.id;
        return React.createElement(CSSTransition, {
          key: id,
          timeout: 400,
          classNames: {
            enter: 'isVisible',
            exit: 'isRemoved'
          }
        }, React.createElement(Toast, {
          kind: content.kind,
          heading: content.heading,
          onClose: _this2.onDismiss(id)
        }, content.text));
      }))) : null);
    }
  }]);

  return ToastProvider;
}(Component);

export default ToastProvider;
