// kds-react/src/components/Toast/Toast.js
import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import DismissalButton from '../DismissalButton/DismissalButton'
import Heading from '../Heading/Heading'
import Paragraph from '../Paragraph/Paragraph'
import { IconInfo, IconError, IconWarning, IconSuccess } from '../Icon/index'

const toasts = {
  success: IconSuccess,
  error: IconError,
  warning: IconWarning,
  info: IconInfo
}

const getKind = kind => {
  const legacyKindMap = {
    success: 'positive',
    error: 'negative',
    warning: 'callout',
    info: 'action'
  }

  return legacyKindMap[kind] || kind
}

const ToastIcon = ({ kind }) => {
  const Glyph = toasts[kind]

  return <Glyph size="utility-l" className="kds-Toast-icon" />
}

/**
 * Toasts require a combination of the following components to render correctly: `ToastProvider` and `ToastConsumer`. The `Toast` component is a child of the `ToastProvider`.
 */

const Toast = ({ children, className, heading, kind, onClose }) => (
  <section
    role="alert"
    aria-live="assertive"
    className={cx('kds-Toast', `kds-Toast--${kind}`, className)}
  >
    <header className="kds-Toast-header">
      <ToastIcon kind={kind} />
      <Heading tag="h1" className="kds-Toast-heading" size="s">
        {heading}
      </Heading>
    </header>
    <Paragraph className="kds-Toast-body" size="m">
      {children}
    </Paragraph>
    <DismissalButton
      kind={getKind(kind)}
      className="kds-Toast-closeButton"
      data-testid="kds-Toast-closeButton"
      onClick={onClose}
    />
  </section>
)

Toast.propTypes = {
  /** Kind of Toast to display, determines the type and color of icon to display. */
  kind: PropTypes.oneOf(['success', 'warning', 'error', 'info']).isRequired,
  /** The Toast heading text. */
  heading: PropTypes.string.isRequired,
  /** The message following the Toast heading. Optional. */
  children: PropTypes.string,
  /** A function that fires when the "Close" button is clicked. */
  onClose: PropTypes.func,
  /** Pass optional utility classes to the Toast container. */
  className: PropTypes.string
}

Toast.defaultProps = {
  kind: 'success',
  heading: 'Success',
  onClose: () => {}
}

export default React.memo(Toast)
