// kds-stencil/dist/collection/components/kds-checkbox/kds-checkbox.js
import { Component, Element, State, Event, Method, Prop, h } from "@stencil/core";
import { getHostAttributes, getSpreadableAttributes, setDefaultHostAttributes, setStyleAttribute } from "../../utils/attributes/attributeUtils";
import { registerAttributeObserver } from "../../utils/registerObserver";
/**
 * Checkboxes should be used when a user needs to make a selection or select multiple options within a list. They can be used to opt in or out of something.
 *
 * > **Note:** `<kds-checkbox />` is a [primitive web component](primitives/) that accepts all props listed below and any attributes prefixed with `inner-`.
 * `inner-*` attributes are stripped of their `inner-` prefix and placed onto a rendered native `<input type="checkbox">` element.
 *
 * ```Message { "props": { "kind": "warning", "className": "mb-16" } }
 * **Angular Consumers:** This component requires at least version 1.0.0 of kds-stencil-accessors for use with Angular Forms.
 * Refer to our [getting started documentation](/stencil/getting-started#Angular%20Accessors) for installation and usage instructions.
 * ```
 */
export class KdsCheckBox {
  constructor() {
    /** Indicates currently-selected item. */
    this.checked = false;
    /** Disables if present. */
    this.disabled = false;
    /** Sets indeterminate state. */
    this.indeterminate = false;
    /** Applies the compact size. */
    this.compact = false;
    this.onChange = (ev) => {
      const input = ev.target;
      if (input) {
        this.checked = input.checked;
        this.indeterminate = input.indeterminate;
      }
    };
  }
  componentWillLoad() {
    this.innerAttrs = getHostAttributes(this.hostEl);
    setDefaultHostAttributes(this.hostEl);
    this.observer = registerAttributeObserver(this);
  }
  componentDidLoad() {
    this.observer.observe(this.hostEl, { attributes: true });
    this.ready.emit(this.checkboxElement);
    this.hostEl.nativeElement = this.checkboxElement;
  }
  disconnectedCallback() {
    this.observer && this.observer.disconnect();
  }
  /**
   * Returns the native `<input>` element used under the hood.
   */
  getNativeElement() {
    return Promise.resolve(this.checkboxElement);
  }
  render() {
    return (h("input", Object.assign({}, getSpreadableAttributes(this.innerAttrs), { autofocus: this.autofocus, checked: this.checked, disabled: this.disabled, indeterminate: this.indeterminate, name: this.name, required: this.required, onChange: this.onChange, class: {
        "kds-Checkbox": true,
        [this.innerAttrs.class]: this.innerAttrs.class,
        "kds-Checkbox--compact": this.compact
      }, ref: (el) => {
        this.checkboxElement = el;
        setStyleAttribute(el, this.innerAttrs);
      }, type: "checkbox", value: this.value })));
  }
  static get is() { return "kds-checkbox"; }
  static get originalStyleUrls() { return {
    "$": ["kds-checkbox.css"]
  }; }
  static get styleUrls() { return {
    "$": ["kds-checkbox.css"]
  }; }
  static get properties() { return {
    "autofocus": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Takes focus when the form is presented."
      },
      "attribute": "autofocus",
      "reflect": false
    },
    "checked": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Indicates currently-selected item."
      },
      "attribute": "checked",
      "reflect": false,
      "defaultValue": "false"
    },
    "disabled": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Disables if present."
      },
      "attribute": "disabled",
      "reflect": false,
      "defaultValue": "false"
    },
    "indeterminate": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Sets indeterminate state."
      },
      "attribute": "indeterminate",
      "reflect": false,
      "defaultValue": "false"
    },
    "name": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Sets the name."
      },
      "attribute": "name",
      "reflect": false
    },
    "required": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Requires value before submission."
      },
      "attribute": "required",
      "reflect": false
    },
    "value": {
      "type": "any",
      "mutable": true,
      "complexType": {
        "original": "string | string[] | number",
        "resolved": "number | string | string[]",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Contains the current value."
      },
      "attribute": "value",
      "reflect": false
    },
    "compact": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Applies the compact size."
      },
      "attribute": "compact",
      "reflect": false,
      "defaultValue": "false"
    }
  }; }
  static get states() { return {
    "innerAttrs": {}
  }; }
  static get events() { return [{
      "method": "ready",
      "name": "ready",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Emits when the component has been loaded.\nEvent's `detail` contains a reference to the inner input element."
      },
      "complexType": {
        "original": "HTMLInputElement",
        "resolved": "HTMLInputElement",
        "references": {
          "HTMLInputElement": {
            "location": "global"
          }
        }
      }
    }]; }
  static get methods() { return {
    "getNativeElement": {
      "complexType": {
        "signature": "() => Promise<HTMLInputElement>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          },
          "HTMLInputElement": {
            "location": "global"
          }
        },
        "return": "Promise<HTMLInputElement>"
      },
      "docs": {
        "text": "Returns the native `<input>` element used under the hood.",
        "tags": []
      }
    }
  }; }
  static get elementRef() { return "hostEl"; }
}
