// kds-react/build/es/components/SolitarySearch/SolitarySearch.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useRef } from 'react';
import cx from 'classnames';
import { IconSearch } from '../Icon/icons/Search.js';
import { callSafely } from '../../utils/functionUtils.js';
import PropTypes from 'prop-types';
import Button from '../Button/Button.js';
import Input from '../Input/Input.js';
import { ofEnterKey } from '../../utils/keyboardUtils.js';

/**
 * `<SolitarySearch />` allows the User to search by term. Pressing enter or interacting with the search button will submit the search term.
 * `<SolitarySearch />` should only be in a form by itself, and should not have sibling inputs.
 *
 * > **Note:** `<SolitarySearch />` renders an `<Input>` with all props spread on, like so: `<Input {...props} />`.
 * So you can pass any native props such as `id`, `className`, `aria-*` attributes, etc.
 *
 */

var SolitarySearch = function SolitarySearch(_ref, externalRef) {
  var onSearch = _ref.onSearch,
      compact = _ref.compact,
      className = _ref.className,
      inverse = _ref.inverse,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      inputProps = _objectWithoutProperties(_ref, ["onSearch", "compact", "className", "inverse", "disabled"]);

  var internalRef = useRef();
  var inputRef = externalRef || internalRef;

  var callback = function callback(e) {
    callSafely(onSearch)(inputRef.current.value, e.nativeEvent);
  };

  var ariaLabel = inputProps['aria-label'] || inputProps.placeholder || 'Search';
  /**
   * Enter Key on input is advised for accessibility
   * https://www.w3.org/TR/WCAG20-TECHS/SCR29.html
   */

  return React.createElement("div", {
    className: cx('kds-SolitarySearch-wrapper flex w-full', {
      'kds-SolitarySearch-wrapper--inverse': inverse
    }, className)
  }, React.createElement(Input, _extends({}, inputProps, {
    "aria-label": ariaLabel,
    disabled: disabled,
    className: "kds-SolitarySearch-input kds-Input--search min-w-0",
    compact: compact,
    onKeyPress: ofEnterKey(callback),
    ref: inputRef,
    type: "search"
  })), React.createElement(Button, {
    "aria-label": ariaLabel,
    disabled: disabled,
    className: "kds-SolitarySearch-button",
    compact: compact,
    kind: "primary-inverse",
    onClick: callback
  }, React.createElement(IconSearch, {
    className: "-ml-1"
  })));
}; // we do this because forwardRef render functions do not support propTypes or defaultProps


var SolitarySearchWithForwardRef = React.forwardRef(SolitarySearch);
SolitarySearchWithForwardRef.displayName = 'SolitarySearch';
SolitarySearchWithForwardRef.defaultProps = {
  compact: false,
  maxLength: 125,
  inverse: false
};
SolitarySearchWithForwardRef.propTypes = {
  /** Decreases the height. */
  compact: PropTypes.bool,
  maxLength: PropTypes.number,

  /** Pass utility classes to the container element */
  className: PropTypes.string,

  /** To indicate if component applies inverse style */
  inverse: PropTypes.bool,

  /**
   * A callback for when a search is performed; i.e. when the `Enter` key is pressed or the button is clicked
   *
   * @param { string } term The search term entered into the search input
   * @param { Event } event The native event (e.g. `KeyboardEvent`, `MouseEvent`) that triggered the callback
   */
  onSearch: PropTypes.func
};

export default SolitarySearchWithForwardRef;
