// kds-react/build/cjs/components/Form/FormField/FormField.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var _objectSpread = _interopDefault(require('@babel/runtime/helpers/objectSpread2'));
require('../../Icon/icons/AcceptEntry.js');
require('../../Icon/icons/AccountSummary.js');
require('../../Icon/icons/Account.js');
require('../../Icon/icons/AddCard.js');
require('../../Icon/icons/AddColumn.js');
require('../../Icon/icons/AddPatient.js');
require('../../Icon/icons/AddPlusCard.js');
require('../../Icon/icons/AddPrescription.js');
require('../../Icon/icons/AdultBeverage.js');
require('../../Icon/icons/Airplane.js');
require('../../Icon/icons/Android.js');
require('../../Icon/icons/Animals.js');
require('../../Icon/icons/ApiAuthorize.js');
require('../../Icon/icons/ApiProduct.js');
require('../../Icon/icons/Approved.js');
require('../../Icon/icons/Ar.js');
require('../../Icon/icons/Associate.js');
require('../../Icon/icons/Asterisk.js');
require('../../Icon/icons/Audience.js');
require('../../Icon/icons/AutoRefill.js');
require('../../Icon/icons/Baby.js');
require('../../Icon/icons/Backstock.js');
require('../../Icon/icons/BagBarcode.js');
require('../../Icon/icons/Bags.js');
require('../../Icon/icons/BarGraph.js');
require('../../Icon/icons/BarLineGraph.js');
require('../../Icon/icons/Barcode.js');
require('../../Icon/icons/BasketTime.js');
require('../../Icon/icons/Basket.js');
require('../../Icon/icons/BeautyPersonalCare.js');
require('../../Icon/icons/Beta.js');
require('../../Icon/icons/BlankBox.js');
require('../../Icon/icons/BuyWireless.js');
require('../../Icon/icons/Calendar.js');
require('../../Icon/icons/Camera.js');
require('../../Icon/icons/Candy.js');
require('../../Icon/icons/CapsLock.js');
require('../../Icon/icons/CardView.js');
require('../../Icon/icons/CarouselView.js');
require('../../Icon/icons/CaretLeft.js');
require('../../Icon/icons/CaretRight.js');
require('../../Icon/icons/Cart.js');
require('../../Icon/icons/CashBack.js');
require('../../Icon/icons/CashRegister.js');
require('../../Icon/icons/CcAmex.js');
require('../../Icon/icons/CcDinersClub.js');
require('../../Icon/icons/CcDiscover.js');
require('../../Icon/icons/CcMastercard.js');
require('../../Icon/icons/CcVisa.js');
require('../../Icon/icons/CellPhone.js');
require('../../Icon/icons/CenterAlign.js');
require('../../Icon/icons/CheckMark.js');
require('../../Icon/icons/ChevronDown.js');
require('../../Icon/icons/ChevronUp.js');
require('../../Icon/icons/CivicOrganizations.js');
require('../../Icon/icons/CleaningHousehold.js');
require('../../Icon/icons/Clock.js');
require('../../Icon/icons/Close.js');
require('../../Icon/icons/CodePhone.js');
require('../../Icon/icons/Coffee.js');
require('../../Icon/icons/ContainsDairy.js');
require('../../Icon/icons/ContainsArtificialIngredients.js');
require('../../Icon/icons/Comment.js');
require('../../Icon/icons/CompanyDirect.js');
require('../../Icon/icons/ContractViewport.js');
require('../../Icon/icons/ControlPayment.js');
require('../../Icon/icons/Convenience.js');
require('../../Icon/icons/CollapseAll.js');
require('../../Icon/icons/Copy.js');
require('../../Icon/icons/Coupons.js');
require('../../Icon/icons/CreateNew.js');
require('../../Icon/icons/CreditCard.js');
require('../../Icon/icons/CurrentLocationAndroid.js');
require('../../Icon/icons/CurrentLocationIos.js');
require('../../Icon/icons/DamagedItem.js');
require('../../Icon/icons/DarkLight.js');
require('../../Icon/icons/Dashboard.js');
require('../../Icon/icons/DebitCard.js');
require('../../Icon/icons/DeliBakery.js');
require('../../Icon/icons/Delivery.js');
require('../../Icon/icons/DiabeticNeeds.js');
require('../../Icon/icons/Directions.js');
require('../../Icon/icons/Diversity.js');
require('../../Icon/icons/DownArrow.js');
require('../../Icon/icons/Download.js');
require('../../Icon/icons/DragHandle.js');
require('../../Icon/icons/Draggable.js');
require('../../Icon/icons/DrugInteractions.js');
require('../../Icon/icons/DrugSearch.js');
require('../../Icon/icons/Ebt.js');
require('../../Icon/icons/Edit.js');
require('../../Icon/icons/Education.js');
require('../../Icon/icons/EggFree.js');
require('../../Icon/icons/Ellipsis.js');
require('../../Icon/icons/Email.js');
require('../../Icon/icons/EmergencyContact.js');
require('../../Icon/icons/Environmental.js');
require('../../Icon/icons/Error.js');
require('../../Icon/icons/ExclusiveSavings.js');
require('../../Icon/icons/Exit.js');
require('../../Icon/icons/ExpandAll.js');
require('../../Icon/icons/ExpandViewport.js');
require('../../Icon/icons/ExpenseReport.js');
require('../../Icon/icons/Expiration.js');
require('../../Icon/icons/Export.js');
require('../../Icon/icons/External.js');
require('../../Icon/icons/EyeSlash.js');
require('../../Icon/icons/Eye.js');
require('../../Icon/icons/Failure.js');
require('../../Icon/icons/FairService.js');
require('../../Icon/icons/FaithBased.js');
require('../../Icon/icons/Faq.js');
require('../../Icon/icons/FatFree.js');
require('../../Icon/icons/FavoriteSolid.js');
require('../../Icon/icons/Favorite.js');
require('../../Icon/icons/Fee.js');
require('../../Icon/icons/Feedback.js');
require('../../Icon/icons/File.js');
require('../../Icon/icons/Filters.js');
require('../../Icon/icons/Flag.js');
require('../../Icon/icons/FlagSolid.js');
require('../../Icon/icons/FlashOff.js');
require('../../Icon/icons/FlashOn.js');
require('../../Icon/icons/Flashlight.js');
require('../../Icon/icons/Floral.js');
require('../../Icon/icons/FmJewelers.js');
require('../../Icon/icons/FuelPump.js');
require('../../Icon/icons/Fuel.js');
require('../../Icon/icons/GanttChart.js');
require('../../Icon/icons/GiftCardMall.js');
require('../../Icon/icons/GiftCard.js');
require('../../Icon/icons/GlutenFree.js');
require('../../Icon/icons/GoodService.js');
require('../../Icon/icons/GoogleHome.js');
require('../../Icon/icons/GridView.js');
require('../../Icon/icons/GroceriesHome.js');
require('../../Icon/icons/Groceries.js');
require('../../Icon/icons/HandSanitizer.js');
require('../../Icon/icons/HamburgerMenuNative.js');
require('../../Icon/icons/HamburgerMenu.js');
require('../../Icon/icons/HealthNutrition.js');
require('../../Icon/icons/HealthWellness.js');
require('../../Icon/icons/HeatEat.js');
require('../../Icon/icons/Hierarchy.js');
require('../../Icon/icons/HighProtein.js');
require('../../Icon/icons/History.js');
require('../../Icon/icons/HomeElectronics.js');
require('../../Icon/icons/Home.js');
require('../../Icon/icons/HungerRelief.js');
require('../../Icon/icons/ImageMissing.js');
require('../../Icon/icons/Import.js');
var Info = require('../../Icon/icons/Info.js');
require('../../Icon/icons/Ingredients.js');
require('../../Icon/icons/InProgress.js');
require('../../Icon/icons/InsuranceInfo.js');
require('../../Icon/icons/IosShare.js');
require('../../Icon/icons/Ios.js');
require('../../Icon/icons/Kosher.js');
require('../../Icon/icons/KrogerCreditCard.js');
require('../../Icon/icons/KrogerRewards.js');
require('../../Icon/icons/LactoseFree.js');
require('../../Icon/icons/Layers.js');
require('../../Icon/icons/LeftAlign.js');
require('../../Icon/icons/LeftArrow.js');
require('../../Icon/icons/Leftover.js');
require('../../Icon/icons/LightBulb.js');
require('../../Icon/icons/Link.js');
require('../../Icon/icons/ListView.js');
require('../../Icon/icons/LoadedSavings.js');
require('../../Icon/icons/LocationA.js');
require('../../Icon/icons/LocationB.js');
require('../../Icon/icons/LocationRx.js');
require('../../Icon/icons/Location.js');
require('../../Icon/icons/Locked.js');
require('../../Icon/icons/LockerPasscode.js');
require('../../Icon/icons/LowFat.js');
require('../../Icon/icons/Map.js');
require('../../Icon/icons/MealKit.js');
require('../../Icon/icons/MeatSeafood.js');
require('../../Icon/icons/MedicationToolkit.js');
require('../../Icon/icons/Members.js');
require('../../Icon/icons/Merge.js');
require('../../Icon/icons/MilitaryPoliceFire.js');
require('../../Icon/icons/Minus.js');
require('../../Icon/icons/MissingItem.js');
require('../../Icon/icons/MoneyServices.js');
require('../../Icon/icons/Moon.js');
require('../../Icon/icons/MoreOptions.js');
require('../../Icon/icons/Natural.js');
require('../../Icon/icons/Newest.js');
require('../../Icon/icons/NoFees.js');
require('../../Icon/icons/NonGmo.js');
require('../../Icon/icons/NotAvailable.js');
require('../../Icon/icons/Notification.js');
require('../../Icon/icons/NuroDeliver.js');
require('../../Icon/icons/NuroPickup.js');
require('../../Icon/icons/NutFree.js');
require('../../Icon/icons/Offline.js');
require('../../Icon/icons/OnlinePayment.js');
require('../../Icon/icons/OptUp.js');
require('../../Icon/icons/Organic.js');
require('../../Icon/icons/OtherMisc.js');
require('../../Icon/icons/Oven.js');
require('../../Icon/icons/Paleo.js');
require('../../Icon/icons/Partnership.js');
require('../../Icon/icons/Paste.js');
require('../../Icon/icons/PatientAccess.js');
require('../../Icon/icons/PatientWellness.js');
require('../../Icon/icons/PaymentSuccess.js');
require('../../Icon/icons/Pencil.js');
require('../../Icon/icons/PersonHome.js');
require('../../Icon/icons/PersonStore.js');
require('../../Icon/icons/PersonalCare.js');
require('../../Icon/icons/PetMedications.js');
require('../../Icon/icons/PharmacyNotifications.js');
require('../../Icon/icons/PharmacySearch.js');
require('../../Icon/icons/Pharmacy.js');
require('../../Icon/icons/Phone.js');
require('../../Icon/icons/Pickup.js');
require('../../Icon/icons/PieChart.js');
require('../../Icon/icons/PiggyBank.js');
require('../../Icon/icons/PillIdentifier.js');
require('../../Icon/icons/PlusCard.js');
require('../../Icon/icons/Plus.js');
require('../../Icon/icons/PoorService.js');
require('../../Icon/icons/PoorSubstitution.js');
require('../../Icon/icons/Popularity.js');
require('../../Icon/icons/PreferredCard.js');
require('../../Icon/icons/PrescriptionHistory.js');
require('../../Icon/icons/PrescriptionStatus.js');
require('../../Icon/icons/Print.js');
require('../../Icon/icons/PrivacyPolicy.js');
require('../../Icon/icons/Produce.js');
require('../../Icon/icons/PromotionOnAd.js');
require('../../Icon/icons/PromotionOnDisplay.js');
require('../../Icon/icons/Promotions.js');
require('../../Icon/icons/Publish.js');
require('../../Icon/icons/PurchaseHistory.js');
require('../../Icon/icons/PushNotification.js');
require('../../Icon/icons/PushPin.js');
require('../../Icon/icons/PushPinSolid.js');
require('../../Icon/icons/PuzzlePiece.js');
require('../../Icon/icons/QrCode.js');
require('../../Icon/icons/RadioChecked.js');
require('../../Icon/icons/RadioUnchecked.js');
require('../../Icon/icons/RatingsHalf.js');
require('../../Icon/icons/RatingsSolid.js');
require('../../Icon/icons/Ratings.js');
require('../../Icon/icons/ReWeigh.js');
require('../../Icon/icons/React.js');
require('../../Icon/icons/Recall.js');
require('../../Icon/icons/Receipt.js');
require('../../Icon/icons/RecentSearch.js');
require('../../Icon/icons/Recipes.js');
require('../../Icon/icons/Recycle.js');
require('../../Icon/icons/Redo.js');
require('../../Icon/icons/RefillFromList.js');
require('../../Icon/icons/RefillPrograms.js');
require('../../Icon/icons/RejectEntry.js');
require('../../Icon/icons/Relevance.js');
require('../../Icon/icons/RemoveColumn.js');
require('../../Icon/icons/Repurchase.js');
require('../../Icon/icons/Reset.js');
require('../../Icon/icons/Restrooms.js');
require('../../Icon/icons/Returns.js');
require('../../Icon/icons/Revert.js');
require('../../Icon/icons/Rewards.js');
require('../../Icon/icons/RightArrow.js');
require('../../Icon/icons/RxBarcode.js');
require('../../Icon/icons/Route.js');
require('../../Icon/icons/Sandwich.js');
require('../../Icon/icons/SaveTime.js');
require('../../Icon/icons/Scan.js');
require('../../Icon/icons/ScanBagGo.js');
require('../../Icon/icons/ScheduleAppointment.js');
require('../../Icon/icons/Search.js');
require('../../Icon/icons/SecurityQuestions.js');
require('../../Icon/icons/SecurityTag.js');
require('../../Icon/icons/Send.js');
require('../../Icon/icons/Settings.js');
require('../../Icon/icons/Share.js');
require('../../Icon/icons/ShipToHome.js');
require('../../Icon/icons/Ship.js');
require('../../Icon/icons/ShopInStore.js');
require('../../Icon/icons/ShopOnline.js');
require('../../Icon/icons/Shop.js');
require('../../Icon/icons/ShoppingList.js');
require('../../Icon/icons/SkipToEnd.js');
require('../../Icon/icons/SkipToStart.js');
require('../../Icon/icons/SocialServices.js');
require('../../Icon/icons/Sort.js');
require('../../Icon/icons/SortAz.js');
require('../../Icon/icons/SortZa.js');
require('../../Icon/icons/SoyFree.js');
require('../../Icon/icons/Status.js');
require('../../Icon/icons/Stencil.js');
require('../../Icon/icons/Stethoscope.js');
require('../../Icon/icons/Stock.js');
require('../../Icon/icons/StoreLocker.js');
require('../../Icon/icons/Store.js');
require('../../Icon/icons/Success.js');
require('../../Icon/icons/Sun.js');
require('../../Icon/icons/Supercharged.js');
require('../../Icon/icons/TableView.js');
require('../../Icon/icons/TermsConditions.js');
require('../../Icon/icons/ThumbsDown.js');
require('../../Icon/icons/ThumbsUp.js');
require('../../Icon/icons/Transfer.js');
require('../../Icon/icons/TransferCircle.js');
require('../../Icon/icons/TransferCircleSolid.js');
require('../../Icon/icons/Trash.js');
require('../../Icon/icons/Trending.js');
require('../../Icon/icons/Undo.js');
require('../../Icon/icons/Unlocked.js');
require('../../Icon/icons/UpArrow.js');
require('../../Icon/icons/Upload.js');
require('../../Icon/icons/VaccinationsMadeEasy.js');
require('../../Icon/icons/Vaccinations.js');
require('../../Icon/icons/Value.js');
require('../../Icon/icons/Vegan.js');
require('../../Icon/icons/Vegetarian.js');
require('../../Icon/icons/Voice.js');
require('../../Icon/icons/Wallet.js');
require('../../Icon/icons/Warning.js');
require('../../Icon/icons/WeeklyAd.js');
require('../../Icon/icons/Weigh.js');
require('../../Icon/icons/WorldwideAcceptance.js');
require('../../Icon/icons/WrongItem.js');
require('../../Icon/icons/YouthPrograms.js');
require('../../Icon/icons/YtdSavings.js');
require('../../Icon/icons/ZoomIn.js');
require('../../Icon/icons/ZoomOut.js');
var functionUtils = require('../../../utils/functionUtils.js');
var generationUtils = require('../../../utils/generationUtils.js');
var Link = require('../../Link/Link.js');
var Label = require('../../Label/Label.js');
var formUtils = require('../../../utils/formUtils.js');
var _prebuiltValidations = require('../validations/_prebuiltValidations.js');
var Input = require('../../Input/Input.js');
var Message = require('../../Message/Message.js');
var MessageGroup = require('../../MessageGroup/MessageGroup.js');
var Paragraph = require('../../Paragraph/Paragraph.js');

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

var MoreInfoLink = function MoreInfoLink(_ref) {
  var _ref$linkText = _ref.linkText,
      linkText = _ref$linkText === void 0 ? 'More Info' : _ref$linkText,
      content = _ref.content,
      id = _ref.id;
  return content ? React__default.createElement(Link.default, {
    noRouter: true,
    onClick: function onClick() {
      return alert(content);
    },
    size: "m",
    className: "ml-8",
    "aria-labelledby": id
  }, linkText, React__default.createElement(Info.IconInfo, {
    className: "ml-4"
  }), React__default.createElement("span", {
    id: id,
    className: "screen-reader"
  }, linkText, ", ", content)) : null;
};

var getKinds = function getKinds(target, validations, invalidKind) {
  // Evaluate the value once outside of the map for better perf
  var value = formUtils.getValueFromTarget(target);
  return validations.map(function (_ref2) {
    var validate = _ref2.validate;
    var valid = functionUtils.callSafely(validate)(value);
    return valid ? 'success' : invalidKind;
  });
};

var _shouldShowMessages = function _shouldShowMessages(showMessages, currentState, nextState) {
  var _currentState$nextSta = _objectSpread({}, currentState, {}, nextState),
      focused = _currentState$nextSta.focused,
      validationError = _currentState$nextSta.validationError;

  if (validationError) {
    return true;
  }

  if (showMessages === 'onFocus') {
    return focused;
  }
};

var ValidationMessages = function ValidationMessages(_ref3) {
  var id = _ref3.id,
      shouldShowMessages = _ref3.shouldShowMessages,
      validations = _ref3.validations,
      messageKinds = _ref3.messageKinds,
      focused = _ref3.focused,
      invalid = _ref3.invalid,
      shouldShowAdaAlert = _ref3.shouldShowAdaAlert,
      screenReaderMessages = _ref3.screenReaderMessages,
      props = _objectWithoutProperties(_ref3, ["id", "shouldShowMessages", "validations", "messageKinds", "focused", "invalid", "shouldShowAdaAlert", "screenReaderMessages"]);

  return shouldShowMessages && validations.length > 0 ? React__default.createElement(React__default.Fragment, null, React__default.createElement(MessageGroup.default, _extends({}, props, {
    "aria-hidden": "true"
  }), validations.map(function (_ref4, index) {
    var text = _ref4.text;
    var kind = messageKinds[index];
    return React__default.createElement(Message.default, {
      key: index,
      kind: kind
    }, text);
  })),
  /*
  * If we are focused on the input we want to have the messages screenread
  * as an aria-describedby description. So we are rendering a div with an
  * id that will be used within underlying Component's aria-describedby attribute.
  */
  focused && shouldShowMessages && React__default.createElement("div", {
    id: id,
    "data-testid": id,
    className: "screen-reader"
  }, screenReaderMessages),
  /*
  * If we are NOT focused on the input and it is invalid, we want to have
  * the messages be "alerted" via aria-live="assertive".
  */
  invalid && !focused && shouldShowAdaAlert && React__default.createElement("div", {
    id: "".concat(id, "-alert"),
    "aria-live": "assertive",
    "aria-relevant": "all",
    "aria-atomic": "true",
    className: "screen-reader"
  }, screenReaderMessages)) : null;
};
/**
 * By default, `FormField` will add an [`Input`](./input) to a [`Form`](./form).
 * You can pass a component through the `as` prop, which will be rendered instead of `Input`, like so:
 * > `<FormField as={MyInputComponent}  {...props} />`.

 *
 * It comes with accessible form features:
 *
 * - label and description
 * - validation messaging
 * - aria attributes
 *
 * All additional props beyond what is specified below will be spread onto the underlying Input (or your custom component), like so:
 * > `<Input {...props} />`.
 *
 * **Requires *custom-event* polyfill for custom validations to work in IE11** [See Polyfills](./polyfills)
 */


var FormField =
/*#__PURE__*/
function (_React$Component) {
  _inherits(FormField, _React$Component);

  var _super = _createSuper(FormField);

  function FormField(props) {
    var _this;

    _classCallCheck(this, FormField);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "id", "".concat(generationUtils.generateUniqueString(), "-input"));

    _defineProperty(_assertThisInitialized(_this), "handleFocus", function () {
      var shouldShowMessages = _this.state.shouldShowMessages;
      var nextMeta = {
        focused: true
      };

      _this.setState(_objectSpread({}, nextMeta, {
        shouldShowMessages: shouldShowMessages || _shouldShowMessages(_this.props.showMessages, _this.state, nextMeta),
        shouldShowAdaAlert: false
      }));
    });

    _defineProperty(_assertThisInitialized(_this), "handleChange", function (event) {
      var validationError = _this.state.validationError;
      var messageKinds = getKinds(event.target, _this._validations, validationError ? 'error' : 'info');
      var messageKindsChanged = _this.state.messageKinds.map(function (currentKind, index) {
        return currentKind === messageKinds[index];
      }).filter(function (same) {
        return !same;
      }).length > 0;

      if (messageKindsChanged) {
        _this.setState({
          messageKinds: messageKinds
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleBlur", function (event) {
      _this.setState({
        focused: false
      });

      setTimeout(function () {
        return _this.setState(function (prevState) {
          return _objectSpread({}, prevState, {
            shouldShowAdaAlert: true
          });
        });
      }, 250);
    });

    var required = props.required,
        validations = props.validations;
    var addRequiredToValidations = _prebuiltValidations.addRequiredToValidations;
    _this._validations = required ? addRequiredToValidations(validations) : validations;
    _this.state = {
      messageKinds: _this._validations.map(function () {
        return 'info';
      }),
      focused: false,
      validationError: false,
      shouldShowMessages: false,
      shouldShowAdaAlert: false
    };
    _this.inputRef = props.inputRef || React__default.createRef();
    return _this;
  }

  _createClass(FormField, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      if (this.inputRef && this.inputRef.current) {
        var el = this.inputRef.current;
        var oldFn = el.checkValidity.bind(el);

        el.checkValidity = function () {
          var messageKinds = getKinds(el, _this2._validations, 'error');
          var validationError = messageKinds.some(function (kind) {
            return kind === 'error';
          });

          _this2.setState({
            messageKinds: messageKinds,
            validationError: validationError,
            shouldShowMessages: _shouldShowMessages(_this2.props.showMessages, _this2.state, {
              validationError: validationError
            })
          });

          return !validationError && oldFn();
        };

        el.restoreDefaultState = function () {
          _this2.setState({
            messageKinds: _this2._validations.map(function () {
              return 'info';
            }),
            focused: false,
            shouldShowMessages: false,
            shouldShowAdaAlert: false,
            submitError: false
          });
        };
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      var _this3 = this;

      var shouldSetFocus = typeof this.props.focusOn !== 'undefined' ? !prevProps.focusOn && this.props.focusOn : (!prevProps.serverMessages || prevProps.serverMessages.length === 0) && this.props.serverMessages && this.props.serverMessages.length > 0;
      var previousKinds = prevState.messageKinds;
      var messageKindsChanged = this.state.messageKinds.map(function (currentKind, index) {
        return currentKind === previousKinds[index];
      }).filter(function (same) {
        return !same;
      }).length > 0;

      if (messageKindsChanged) {
        var payload = {
          messagesShown: this.state.shouldShowMessages && this.state.messageKinds
        };
        var validationEvent;

        try {
          validationEvent = new CustomEvent('validation', {
            bubbles: true,
            detail: payload
          });
        } catch (e) {
          throw new Error('KDS FormField validations error. You may need the custom-event polyfill http://design.kroger.com/#/react/components/polyfills');
        }

        this.inputRef.current.setAttribute('data-messagesShown', JSON.stringify(this.state.shouldShowMessages ? this.state.messageKinds.map(function (kind, index) {
          return {
            messageText: _this3._validations[index].text,
            messageKind: kind
          };
        }) : []));
        this.inputRef.current.dispatchEvent(validationEvent);
      }

      if (shouldSetFocus) {
        setTimeout(function () {
          _this3.inputRef.current.focus();
        }, 0);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          Component = _this$props.as,
          name = _this$props.name,
          label = _this$props.label,
          requiredProp = _this$props.required,
          validations = _this$props.validations,
          showMessages = _this$props.showMessages,
          className = _this$props.className,
          moreInfo = _this$props.moreInfo,
          helpText = _this$props.helpText,
          button = _this$props.button,
          inputRef = _this$props.inputRef,
          _this$props$serverMes = _this$props.serverMessages,
          serverMessages = _this$props$serverMes === void 0 ? [] : _this$props$serverMes,
          props = _objectWithoutProperties(_this$props, ["as", "name", "label", "required", "validations", "showMessages", "className", "moreInfo", "helpText", "button", "inputRef", "serverMessages"]);

      var _this$state = this.state,
          messageKinds = _this$state.messageKinds,
          focused = _this$state.focused,
          validationError = _this$state.validationError,
          shouldShowMessages = _this$state.shouldShowMessages,
          shouldShowAdaAlert = _this$state.shouldShowAdaAlert;
      var required = _prebuiltValidations.contains(this._validations, _prebuiltValidations.required);
      var serverError = serverMessages.length > 0;
      var invalid = serverError || validationError && messageKinds.some(function (kind) {
        return kind === 'error';
      });
      var metaClasses = cx({
        'is-invalid': invalid,
        'is-focused': focused,
        'is-required': required
      });
      var screenReaderMessages = (!validationError && serverError ? serverMessages : this._validations.filter(function (_, i) {
        return messageKinds[i] !== 'success';
      }).map(function (_ref5) {
        var text = _ref5.text;
        return text;
      })).join('. ');
      var id = props.id || this.id;
      var helpTextId = helpText ? "".concat(id, "-helpText") : '';
      var moreInfoId = moreInfo ? "".concat(id, "-moreInfo") : '';
      var messagesId = shouldShowMessages || serverError ? "".concat(id, "-messages") : '';
      /* Create a space-separated list of ids or an empty string.
      // Order matters for AT users. Please preserve: help text, more info, validation messages */

      var ariaDescribedbyIds = [helpTextId, moreInfoId, messagesId].join(' ').trim();
      var IconButton = button && standardizeIconButton(button, id);
      return React__default.createElement("div", {
        className: cx('kds-Form--field kds-FormField', metaClasses, className)
      }, label && React__default.createElement("span", {
        className: "flex"
      }, React__default.createElement(Label.default, {
        htmlFor: id,
        className: "mb-8"
      }, label), React__default.createElement(MoreInfoLink, _extends({
        id: moreInfoId
      }, moreInfo))), helpText && React__default.createElement(Paragraph.default, {
        id: helpTextId,
        size: "s",
        className: "-mt-4 mb-8"
      }, helpText), React__default.createElement("div", {
        className: "kds-FormField-inputWrapper flex items-center"
      }, React__default.createElement(Component, _extends({}, props, {
        ref: this.inputRef,
        id: id,
        name: formUtils.valOrCamelCasedFallback(name, label),
        "aria-describedby": ariaDescribedbyIds || null,
        "aria-invalid": invalid,
        "aria-required": required,
        onFocus: functionUtils.callAll(this.handleFocus, this.props.onFocus),
        onChange: functionUtils.callAll(this.handleChange, this.props.onChange),
        onBlur: functionUtils.callAll(this.handleBlur, this.props.onBlur),
        className: cx(button && 'kds-FormField-input--withIconButton')
      })), IconButton), React__default.createElement(ValidationMessages, {
        id: messagesId,
        shouldShowMessages: shouldShowMessages || serverError,
        validations: !validationError && serverError ? serverMessages.map(function (text) {
          return {
            text: text
          };
        }) : this._validations,
        messageKinds: !validationError && serverError ? serverMessages.map(function () {
          return 'error';
        }) : messageKinds,
        focused: focused,
        invalid: invalid,
        shouldShowAdaAlert: shouldShowAdaAlert,
        screenReaderMessages: screenReaderMessages
      }));
    }
  }]);

  return FormField;
}(React__default.Component);

FormField.defaultProps = {
  as: Input.default,
  validations: []
};

function standardizeIconButton(button, inputId) {
  var onlyButton = React__default.Children.only(button);
  var onlyIcon = React__default.Children.only(button.props.children);
  return React__default.cloneElement(onlyButton, _objectSpread({}, onlyButton.props, {
    kind: 'tertiary',
    type: 'button',
    'aria-controls': inputId,
    className: cx('kds-FormField-iconButton', onlyButton.props.className),
    children: React__default.cloneElement(onlyIcon, _objectSpread({}, onlyIcon.props, {
      size: 'm',
      className: cx('kds-FormField-icon', onlyIcon.props.className),
      color: 'interactive'
    }))
  }));
}

exports.default = FormField;
