// kds-react/build/es/components/Form/FormCheckbox/FormCheckbox.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React from 'react';
import Label from '../../Label/Label.js';
import Checkbox from '../../Checkbox/Checkbox.js';
import { isTruthyOrZero } from '../../../utils/objectUtils.js';
import { valOrCamelCasedFallback } from '../../../utils/formUtils.js';
import { addRequiredToValidations } from '../validations/_prebuiltValidations.js';
import FormField from '../FormField/FormField.js';

var CHECKBOX_REQUIRED_DEFAULT_TEXT = 'Check mark is required.';
var CheckboxWithLabel = React.forwardRef(function (_ref, ref) {
  var checkboxLabel = _ref.checkboxLabel,
      name = _ref.name,
      value = _ref.value,
      props = _objectWithoutProperties(_ref, ["checkboxLabel", "name", "value"]);

  return React.createElement(Label, null, React.createElement(Checkbox, _extends({
    ref: ref
  }, props, {
    name: valOrCamelCasedFallback(name, checkboxLabel),
    value: isTruthyOrZero(value) ? value : checkboxLabel
  })), checkboxLabel);
});
/**
 * `FormCheckbox` will add a checkbox to a [`Form`](./form).
 *
 * > **Note:** `<FormCheckbox />` renders a `<FormField />` with most props spread on.
 * For more details, see the [FormField](./formfield) and [`Checkbox`](./checkbox) components.
 *
 * It comes with accessible form features:
 *
 * - label
 * - validation messaging
 * - aria attributes
 *
 * **Requires *custom-event* polyfill for custom validations to work in IE11** [See Polyfills](./polyfills)
 */

var FormCheckbox = function FormCheckbox(_ref2) {
  var label = _ref2.label,
      required = _ref2.required,
      validations = _ref2.validations,
      props = _objectWithoutProperties(_ref2, ["label", "required", "validations"]);

  return React.createElement(FormField, _extends({}, props, {
    as: CheckboxWithLabel,
    checkboxLabel: label,
    validations: required ? addRequiredToValidations(validations, CHECKBOX_REQUIRED_DEFAULT_TEXT) : validations
  }));
};

export default FormCheckbox;
export { CHECKBOX_REQUIRED_DEFAULT_TEXT };
