// kds-stencil-react-lite/dist/components.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KdsIconWarningFilled = exports.KdsIconSuccessFilled = exports.KdsIconRecallFilled = exports.KdsIconInfoFilled = exports.KdsIconFailureFilled = exports.KdsIconErrorFilled = exports.KdsIcon = exports.KdsTooltippable = void 0;
/* eslint-disable */
/* tslint:disable */
/* auto-generated react proxies */
var react_component_lib_1 = require("./react-component-lib");
require("kds-stencil-lite/dist/esm/polyfills/promise.js");
var hasWindow = typeof window !== "undefined";
// Apply forEach polyfill to DOMTokenList. Used by React bindings.
if (hasWindow && window.DOMTokenList && !DOMTokenList.prototype.forEach) {
    DOMTokenList.prototype.forEach = function (callback, thisArg) {
        thisArg = thisArg || window;
        for (var i = 0; i < this.length; i++) {
            callback.call(thisArg, this[i], i, this);
        }
    };
}
var loader = hasWindow && require('kds-stencil-lite/dist/loader');
loader && loader.applyPolyfills().then(function () { return window && window.customElements && loader.defineCustomElements(window); });
exports.KdsTooltippable = react_component_lib_1.createReactComponent('kds-tooltippable');
exports.KdsIcon = react_component_lib_1.createReactComponent('kds-icon');
exports.KdsIconErrorFilled = react_component_lib_1.createReactComponent('kds-icon-error-filled');
exports.KdsIconFailureFilled = react_component_lib_1.createReactComponent('kds-icon-failure-filled');
exports.KdsIconInfoFilled = react_component_lib_1.createReactComponent('kds-icon-info-filled');
exports.KdsIconRecallFilled = react_component_lib_1.createReactComponent('kds-icon-recall-filled');
exports.KdsIconSuccessFilled = react_component_lib_1.createReactComponent('kds-icon-success-filled');
exports.KdsIconWarningFilled = react_component_lib_1.createReactComponent('kds-icon-warning-filled');
