// kds-react/build/es/components/Stepper/Stepper.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { PureComponent } from 'react';
import cx from 'classnames';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import '../Icon/icons/AcceptEntry.js';
import '../Icon/icons/AccountSummary.js';
import '../Icon/icons/Account.js';
import '../Icon/icons/AddCard.js';
import '../Icon/icons/AddColumn.js';
import '../Icon/icons/AddPatient.js';
import '../Icon/icons/AddPlusCard.js';
import '../Icon/icons/AddPrescription.js';
import '../Icon/icons/AdultBeverage.js';
import '../Icon/icons/Airplane.js';
import '../Icon/icons/Android.js';
import '../Icon/icons/Animals.js';
import '../Icon/icons/ApiAuthorize.js';
import '../Icon/icons/ApiProduct.js';
import '../Icon/icons/Approved.js';
import '../Icon/icons/Ar.js';
import '../Icon/icons/Associate.js';
import '../Icon/icons/Asterisk.js';
import '../Icon/icons/Audience.js';
import '../Icon/icons/AutoRefill.js';
import '../Icon/icons/Baby.js';
import '../Icon/icons/Backstock.js';
import '../Icon/icons/BagBarcode.js';
import '../Icon/icons/Bags.js';
import '../Icon/icons/BarGraph.js';
import '../Icon/icons/BarLineGraph.js';
import '../Icon/icons/Barcode.js';
import '../Icon/icons/BasketTime.js';
import '../Icon/icons/Basket.js';
import '../Icon/icons/BeautyPersonalCare.js';
import '../Icon/icons/Beta.js';
import '../Icon/icons/BlankBox.js';
import '../Icon/icons/BuyWireless.js';
import '../Icon/icons/Calendar.js';
import '../Icon/icons/Camera.js';
import '../Icon/icons/Candy.js';
import '../Icon/icons/CapsLock.js';
import '../Icon/icons/CardView.js';
import '../Icon/icons/CarouselView.js';
import '../Icon/icons/CaretLeft.js';
import '../Icon/icons/CaretRight.js';
import '../Icon/icons/Cart.js';
import '../Icon/icons/CashBack.js';
import '../Icon/icons/CashRegister.js';
import '../Icon/icons/CcAmex.js';
import '../Icon/icons/CcDinersClub.js';
import '../Icon/icons/CcDiscover.js';
import '../Icon/icons/CcMastercard.js';
import '../Icon/icons/CcVisa.js';
import '../Icon/icons/CellPhone.js';
import '../Icon/icons/CenterAlign.js';
import '../Icon/icons/CheckMark.js';
import '../Icon/icons/ChevronDown.js';
import '../Icon/icons/ChevronUp.js';
import '../Icon/icons/CivicOrganizations.js';
import '../Icon/icons/CleaningHousehold.js';
import '../Icon/icons/Clock.js';
import '../Icon/icons/Close.js';
import '../Icon/icons/CodePhone.js';
import '../Icon/icons/Coffee.js';
import '../Icon/icons/ContainsDairy.js';
import '../Icon/icons/ContainsArtificialIngredients.js';
import '../Icon/icons/Comment.js';
import '../Icon/icons/CompanyDirect.js';
import '../Icon/icons/ContractViewport.js';
import '../Icon/icons/ControlPayment.js';
import '../Icon/icons/Convenience.js';
import '../Icon/icons/CollapseAll.js';
import '../Icon/icons/Copy.js';
import '../Icon/icons/Coupons.js';
import '../Icon/icons/CreateNew.js';
import '../Icon/icons/CreditCard.js';
import '../Icon/icons/CurrentLocationAndroid.js';
import '../Icon/icons/CurrentLocationIos.js';
import '../Icon/icons/DamagedItem.js';
import '../Icon/icons/DarkLight.js';
import '../Icon/icons/Dashboard.js';
import '../Icon/icons/DebitCard.js';
import '../Icon/icons/DeliBakery.js';
import '../Icon/icons/Delivery.js';
import '../Icon/icons/DiabeticNeeds.js';
import '../Icon/icons/Directions.js';
import '../Icon/icons/Diversity.js';
import '../Icon/icons/DownArrow.js';
import '../Icon/icons/Download.js';
import '../Icon/icons/DragHandle.js';
import '../Icon/icons/Draggable.js';
import '../Icon/icons/DrugInteractions.js';
import '../Icon/icons/DrugSearch.js';
import '../Icon/icons/Ebt.js';
import '../Icon/icons/Edit.js';
import '../Icon/icons/Education.js';
import '../Icon/icons/EggFree.js';
import '../Icon/icons/Ellipsis.js';
import '../Icon/icons/Email.js';
import '../Icon/icons/EmergencyContact.js';
import '../Icon/icons/Environmental.js';
import '../Icon/icons/Error.js';
import '../Icon/icons/ExclusiveSavings.js';
import '../Icon/icons/Exit.js';
import '../Icon/icons/ExpandAll.js';
import '../Icon/icons/ExpandViewport.js';
import '../Icon/icons/ExpenseReport.js';
import '../Icon/icons/Expiration.js';
import '../Icon/icons/Export.js';
import '../Icon/icons/External.js';
import '../Icon/icons/EyeSlash.js';
import '../Icon/icons/Eye.js';
import '../Icon/icons/Failure.js';
import '../Icon/icons/FairService.js';
import '../Icon/icons/FaithBased.js';
import '../Icon/icons/Faq.js';
import '../Icon/icons/FatFree.js';
import '../Icon/icons/FavoriteSolid.js';
import '../Icon/icons/Favorite.js';
import '../Icon/icons/Fee.js';
import '../Icon/icons/Feedback.js';
import '../Icon/icons/File.js';
import '../Icon/icons/Filters.js';
import '../Icon/icons/Flag.js';
import '../Icon/icons/FlagSolid.js';
import '../Icon/icons/FlashOff.js';
import '../Icon/icons/FlashOn.js';
import '../Icon/icons/Flashlight.js';
import '../Icon/icons/Floral.js';
import '../Icon/icons/FmJewelers.js';
import '../Icon/icons/FuelPump.js';
import '../Icon/icons/Fuel.js';
import '../Icon/icons/GanttChart.js';
import '../Icon/icons/GiftCardMall.js';
import '../Icon/icons/GiftCard.js';
import '../Icon/icons/GlutenFree.js';
import '../Icon/icons/GoodService.js';
import '../Icon/icons/GoogleHome.js';
import '../Icon/icons/GridView.js';
import '../Icon/icons/GroceriesHome.js';
import '../Icon/icons/Groceries.js';
import '../Icon/icons/HandSanitizer.js';
import '../Icon/icons/HamburgerMenuNative.js';
import '../Icon/icons/HamburgerMenu.js';
import '../Icon/icons/HealthNutrition.js';
import '../Icon/icons/HealthWellness.js';
import '../Icon/icons/HeatEat.js';
import '../Icon/icons/Hierarchy.js';
import '../Icon/icons/HighProtein.js';
import '../Icon/icons/History.js';
import '../Icon/icons/HomeElectronics.js';
import '../Icon/icons/Home.js';
import '../Icon/icons/HungerRelief.js';
import '../Icon/icons/ImageMissing.js';
import '../Icon/icons/Import.js';
import '../Icon/icons/Info.js';
import '../Icon/icons/Ingredients.js';
import '../Icon/icons/InProgress.js';
import '../Icon/icons/InsuranceInfo.js';
import '../Icon/icons/IosShare.js';
import '../Icon/icons/Ios.js';
import '../Icon/icons/Kosher.js';
import '../Icon/icons/KrogerCreditCard.js';
import '../Icon/icons/KrogerRewards.js';
import '../Icon/icons/LactoseFree.js';
import '../Icon/icons/Layers.js';
import '../Icon/icons/LeftAlign.js';
import '../Icon/icons/LeftArrow.js';
import '../Icon/icons/Leftover.js';
import '../Icon/icons/LightBulb.js';
import '../Icon/icons/Link.js';
import '../Icon/icons/ListView.js';
import '../Icon/icons/LoadedSavings.js';
import '../Icon/icons/LocationA.js';
import '../Icon/icons/LocationB.js';
import '../Icon/icons/LocationRx.js';
import '../Icon/icons/Location.js';
import '../Icon/icons/Locked.js';
import '../Icon/icons/LockerPasscode.js';
import '../Icon/icons/LowFat.js';
import '../Icon/icons/Map.js';
import '../Icon/icons/MealKit.js';
import '../Icon/icons/MeatSeafood.js';
import '../Icon/icons/MedicationToolkit.js';
import '../Icon/icons/Members.js';
import '../Icon/icons/Merge.js';
import '../Icon/icons/MilitaryPoliceFire.js';
import { IconMinus } from '../Icon/icons/Minus.js';
import '../Icon/icons/MissingItem.js';
import '../Icon/icons/MoneyServices.js';
import '../Icon/icons/Moon.js';
import '../Icon/icons/MoreOptions.js';
import '../Icon/icons/Natural.js';
import '../Icon/icons/Newest.js';
import '../Icon/icons/NoFees.js';
import '../Icon/icons/NonGmo.js';
import '../Icon/icons/NotAvailable.js';
import '../Icon/icons/Notification.js';
import '../Icon/icons/NuroDeliver.js';
import '../Icon/icons/NuroPickup.js';
import '../Icon/icons/NutFree.js';
import '../Icon/icons/Offline.js';
import '../Icon/icons/OnlinePayment.js';
import '../Icon/icons/OptUp.js';
import '../Icon/icons/Organic.js';
import '../Icon/icons/OtherMisc.js';
import '../Icon/icons/Oven.js';
import '../Icon/icons/Paleo.js';
import '../Icon/icons/Partnership.js';
import '../Icon/icons/Paste.js';
import '../Icon/icons/PatientAccess.js';
import '../Icon/icons/PatientWellness.js';
import '../Icon/icons/PaymentSuccess.js';
import '../Icon/icons/Pencil.js';
import '../Icon/icons/PersonHome.js';
import '../Icon/icons/PersonStore.js';
import '../Icon/icons/PersonalCare.js';
import '../Icon/icons/PetMedications.js';
import '../Icon/icons/PharmacyNotifications.js';
import '../Icon/icons/PharmacySearch.js';
import '../Icon/icons/Pharmacy.js';
import '../Icon/icons/Phone.js';
import '../Icon/icons/Pickup.js';
import '../Icon/icons/PieChart.js';
import '../Icon/icons/PiggyBank.js';
import '../Icon/icons/PillIdentifier.js';
import '../Icon/icons/PlusCard.js';
import { IconPlus } from '../Icon/icons/Plus.js';
import '../Icon/icons/PoorService.js';
import '../Icon/icons/PoorSubstitution.js';
import '../Icon/icons/Popularity.js';
import '../Icon/icons/PreferredCard.js';
import '../Icon/icons/PrescriptionHistory.js';
import '../Icon/icons/PrescriptionStatus.js';
import '../Icon/icons/Print.js';
import '../Icon/icons/PrivacyPolicy.js';
import '../Icon/icons/Produce.js';
import '../Icon/icons/PromotionOnAd.js';
import '../Icon/icons/PromotionOnDisplay.js';
import '../Icon/icons/Promotions.js';
import '../Icon/icons/Publish.js';
import '../Icon/icons/PurchaseHistory.js';
import '../Icon/icons/PushNotification.js';
import '../Icon/icons/PushPin.js';
import '../Icon/icons/PushPinSolid.js';
import '../Icon/icons/PuzzlePiece.js';
import '../Icon/icons/QrCode.js';
import '../Icon/icons/RadioChecked.js';
import '../Icon/icons/RadioUnchecked.js';
import '../Icon/icons/RatingsHalf.js';
import '../Icon/icons/RatingsSolid.js';
import '../Icon/icons/Ratings.js';
import '../Icon/icons/ReWeigh.js';
import '../Icon/icons/React.js';
import '../Icon/icons/Recall.js';
import '../Icon/icons/Receipt.js';
import '../Icon/icons/RecentSearch.js';
import '../Icon/icons/Recipes.js';
import '../Icon/icons/Recycle.js';
import '../Icon/icons/Redo.js';
import '../Icon/icons/RefillFromList.js';
import '../Icon/icons/RefillPrograms.js';
import '../Icon/icons/RejectEntry.js';
import '../Icon/icons/Relevance.js';
import '../Icon/icons/RemoveColumn.js';
import '../Icon/icons/Repurchase.js';
import '../Icon/icons/Reset.js';
import '../Icon/icons/Restrooms.js';
import '../Icon/icons/Returns.js';
import '../Icon/icons/Revert.js';
import '../Icon/icons/Rewards.js';
import '../Icon/icons/RightArrow.js';
import '../Icon/icons/RxBarcode.js';
import '../Icon/icons/Route.js';
import '../Icon/icons/Sandwich.js';
import '../Icon/icons/SaveTime.js';
import '../Icon/icons/Scan.js';
import '../Icon/icons/ScanBagGo.js';
import '../Icon/icons/ScheduleAppointment.js';
import '../Icon/icons/Search.js';
import '../Icon/icons/SecurityQuestions.js';
import '../Icon/icons/SecurityTag.js';
import '../Icon/icons/Send.js';
import '../Icon/icons/Settings.js';
import '../Icon/icons/Share.js';
import '../Icon/icons/ShipToHome.js';
import '../Icon/icons/Ship.js';
import '../Icon/icons/ShopInStore.js';
import '../Icon/icons/ShopOnline.js';
import '../Icon/icons/Shop.js';
import '../Icon/icons/ShoppingList.js';
import '../Icon/icons/SkipToEnd.js';
import '../Icon/icons/SkipToStart.js';
import '../Icon/icons/SocialServices.js';
import '../Icon/icons/Sort.js';
import '../Icon/icons/SortAz.js';
import '../Icon/icons/SortZa.js';
import '../Icon/icons/SoyFree.js';
import '../Icon/icons/Status.js';
import '../Icon/icons/Stencil.js';
import '../Icon/icons/Stethoscope.js';
import '../Icon/icons/Stock.js';
import '../Icon/icons/StoreLocker.js';
import '../Icon/icons/Store.js';
import '../Icon/icons/Success.js';
import '../Icon/icons/Sun.js';
import '../Icon/icons/Supercharged.js';
import '../Icon/icons/TableView.js';
import '../Icon/icons/TermsConditions.js';
import '../Icon/icons/ThumbsDown.js';
import '../Icon/icons/ThumbsUp.js';
import '../Icon/icons/Transfer.js';
import '../Icon/icons/TransferCircle.js';
import '../Icon/icons/TransferCircleSolid.js';
import { IconTrash } from '../Icon/icons/Trash.js';
import '../Icon/icons/Trending.js';
import '../Icon/icons/Undo.js';
import '../Icon/icons/Unlocked.js';
import '../Icon/icons/UpArrow.js';
import '../Icon/icons/Upload.js';
import '../Icon/icons/VaccinationsMadeEasy.js';
import '../Icon/icons/Vaccinations.js';
import '../Icon/icons/Value.js';
import '../Icon/icons/Vegan.js';
import '../Icon/icons/Vegetarian.js';
import '../Icon/icons/Voice.js';
import '../Icon/icons/Wallet.js';
import '../Icon/icons/Warning.js';
import '../Icon/icons/WeeklyAd.js';
import '../Icon/icons/Weigh.js';
import '../Icon/icons/WorldwideAcceptance.js';
import '../Icon/icons/WrongItem.js';
import '../Icon/icons/YouthPrograms.js';
import '../Icon/icons/YtdSavings.js';
import '../Icon/icons/ZoomIn.js';
import '../Icon/icons/ZoomOut.js';
import { debounce } from '../../utils/functionUtils.js';
import Text from '../Text/Text.js';
import Button from '../Button/Button.js';
import { clamp } from '../../utils/numberUtils.js';

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

function cleanValue(oldValue, newValue, min, max) {
  newValue = parseInt(newValue) || 0;
  var value = oldValue;

  if (min <= newValue && newValue <= max) {
    value = newValue;
  } else if (newValue > max) {
    value = max;
  } else if (newValue < min) {
    var direction = newValue - value;
    value = direction > 0 ? min // This means we're adding, snap to min
    : 0; // if we're subtracting, snap to zero
  }

  return value;
}

var kinds = ['primary', 'secondary'];
/**
 * `Stepper` lets users increment and decrement quantities. Use them for things like adding/removing items to/from a cart or list.
 *
 * It handles local state and passes the current value to the `onChange` callback for you to use as needed.
 *
 * Because it wraps the buttons and input in a `div`, additional props will be spread onto the `div`.
 *
 * Stepper comes built in with a 400ms debounce on increment/decrement and input actions.
 *
 * ### Accessibility
 *
 * This component meets the criteria for a `spinbutton` as defined in <a href="https://www.w3.org/TR/wai-aria-1.1/#spinbutton" rel="external">WAI-ARIA 1.1: 5.4 Definition of Roles: spinbutton</a>
 *
 * And it follows patterns informed by <a href="https://www.w3.org/TR/wai-aria-practices-1.1/#spinbutton" rel="external">WAI-ARIA Authoring Practices: Design Patterns and Widgets: Spinbutton</a>
 *
 * >**Note:** IE 11 doesn't natively support keyboard controls for input `type="number"` other than manual numeric entry. Because the KDS defers to native browser behavior for input controls, we do not add this functionality.
 *
 * > ```Message { "props": { "kind": "warning" } }
 * > *THIS COMPONENT IS ON A DEPRECATION PATH.*
 * >
 */

var Stepper =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Stepper, _PureComponent);

  var _super = _createSuper(Stepper);

  function Stepper() {
    var _this;

    _classCallCheck(this, Stepper);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      clampedMax: clamp(_this.props.max, 1, 999),
      inputIsFocusedAndZero: false,
      shouldSetFocusOnCTA: false,
      value: _this.props.value || 0,
      updatedValue: _this.props.value || 0,
      locked: false
    });

    _defineProperty(_assertThisInitialized(_this), "ctaButtonRef", React.createRef());

    _defineProperty(_assertThisInitialized(_this), "lastEmittedValue", _this.state.value);

    _defineProperty(_assertThisInitialized(_this), "debouncedOnChange", debounce(function (value) {
      _this.props.onChange(value);

      _this.lastEmittedValue = value;
      _this.state.locked = false;
    }, 800));

    _defineProperty(_assertThisInitialized(_this), "incrBtnRef", React.createRef());

    _defineProperty(_assertThisInitialized(_this), "getMinimumQuantity", function () {
      return Math.max(1, _this.props.minimumQuantity);
    });

    _defineProperty(_assertThisInitialized(_this), "isOutOfRange", function (value) {
      return value < 0 || value > _this.state.clampedMax;
    });

    _defineProperty(_assertThisInitialized(_this), "isLessThanMinQuantity", function (value) {
      return value > 0 && value < _this.getMinimumQuantity();
    });

    _defineProperty(_assertThisInitialized(_this), "isInvalid", function (value) {
      return _this.isOutOfRange(value) || _this.isLessThanMinQuantity(value);
    });

    _defineProperty(_assertThisInitialized(_this), "isEqualToLastEmittedValue", function (value) {
      return value === _this.lastEmittedValue;
    });

    _defineProperty(_assertThisInitialized(_this), "tryOnChangeCallback", function () {
      var debounce = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;

      /* Get the newly updated value from state. */
      var value = _this.state.value;
      /* If value is equal to last emitted value, out of range, or otherwise invalid, don't execute callback function. */

      if (!_this.isInvalid(value) && !_this.isEqualToLastEmittedValue(value)) {
        if (value === 0 || !debounce) {
          /* If the value is 0 we want to fire onChange immediately,
           * so we cancel the debounced version and call it directly. */
          _this.state.locked = false;

          _this.debouncedOnChange.cancel();

          _this.props.onChange(value);

          _this.lastEmittedValue = value;
        } else {
          _this.state.locked = true;

          _this.debouncedOnChange(value);
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "cleanValue", function (newValue) {
      return cleanValue(_this.state.value, newValue, _this.getMinimumQuantity(), _this.props.max);
    });

    _defineProperty(_assertThisInitialized(_this), "focusOnCTA", function () {
      var callToAction = _this.props.callToAction;
      var _this$state = _this.state,
          inputIsFocusedAndZero = _this$state.inputIsFocusedAndZero,
          shouldSetFocusOnCTA = _this$state.shouldSetFocusOnCTA,
          value = _this$state.value;
      /* If callToAction prop is set
       * and value is 0
       * and current state indicates that focus should be set on the CTA button
       * and current state indicates that the input is NOT focused with a zero value, ... */

      callToAction && value === 0 && shouldSetFocusOnCTA && !inputIsFocusedAndZero &&
      /* then in the next frame, ... */
      setTimeout(function () {
        /* if the CTA button exists and is focusable, then set focus on it. */
        _this.ctaButtonRef && _this.ctaButtonRef.current && typeof _this.ctaButtonRef.current.focus === 'function' && _this.ctaButtonRef.current.focus();
        _this.state.shouldSetFocusOnCTA = false;
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onCTAButtonClick", function () {
      _this.setState({
        value: _this.getMinimumQuantity()
      }, function () {
        _this.tryOnChangeCallback(_this.state.value);
        /* We want the input to take focus when it renders in place of the "call to action" button (for accessibility) */


        _this.incrBtnRef.current.focus();
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onDecrementClick", function () {
      _this.setState(function (_ref) {
        var value = _ref.value;

        var cleanedValue = _this.cleanValue(value - 1);

        return {
          /** Snap manually entered (typed) invalid value to clamped max. */
          value: cleanedValue,
          inputIsFocusedAndZero: false,
          shouldSetFocusOnCTA: cleanedValue - 1 < _this.getMinimumQuantity()
        };
      }, _this.tryOnChangeCallback);
    });

    _defineProperty(_assertThisInitialized(_this), "onIncrementClick", function () {
      _this.setState(function (_ref2) {
        var value = _ref2.value;

        var cleanedValue = _this.cleanValue(value + 1);

        return {
          /** Snap manually entered (typed) invalid value to 0. */
          value: cleanedValue
        };
      }, _this.tryOnChangeCallback);
    });

    _defineProperty(_assertThisInitialized(_this), "onInputChange", function (e) {
      /* In this event we "clean" the value first because a user could have typed anything into the input. */
      var cleanedValue = parseInt(e.target.value) || 0;

      _this.setState({
        inputIsFocusedAndZero: cleanedValue === 0,
        // indicate when input has focus and zero value
        value: cleanedValue
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onInputKeyPress", function (e) {
      var value = _this.state.value;
      /* We want users to have to "commit" a zero typed directly into the focused input. */

      if (e.key === 'Enter') {
        _this.setState({
          /* We fake blur by telling state that the input is not focused, so we don't trigger the onInputBlur handler; */
          inputIsFocusedAndZero: false,

          /* In this event, we want to transfer focus back to the CTA button, unlike in a true blur event. */
          shouldSetFocusOnCTA: value === 0
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onInputBlur", function () {
      /* On blur, set state accordingly ... */
      _this.setState({
        inputIsFocusedAndZero: false,
        shouldSetFocusOnCTA: false
      }, function () {
        return _this.tryOnChangeCallback(false);
      });
    });

    return _this;
  }

  _createClass(Stepper, [{
    key: "componentDidUpdate",

    /* We want the "call to action" button to take focus when value is decreased to zero (for accessibility) */
    value: function componentDidUpdate(prevProps) {
      if (this.props.value !== prevProps.value) {
        this.lastEmittedValue = this.props.value;
      }

      this.focusOnCTA();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
          clampedMax = _this$state2.clampedMax,
          inputIsFocusedAndZero = _this$state2.inputIsFocusedAndZero,
          value = _this$state2.value;

      var _this$props = this.props,
          className = _this$props.className,
          compact = _this$props.compact,
          callToAction = _this$props.callToAction,
          description = _this$props.description,
          destructive = _this$props.destructive,
          fullWidth = _this$props.fullWidth,
          what = _this$props.what,
          kind = _this$props.kind,
          childrenOmitted = _this$props.children,
          maxOmitted = _this$props.max,
          minimumQuantityOmitted = _this$props.minimumQuantity,
          onChangeOmitted = _this$props.onChange,
          valueOmitted = _this$props.value,
          divProps = _objectWithoutProperties(_this$props, ["className", "compact", "callToAction", "description", "destructive", "fullWidth", "what", "kind", "children", "max", "minimumQuantity", "onChange", "value"]);

      var destructiveDecrement = destructive && value <= this.getMinimumQuantity();
      var hideCTAButton = value !== 0 || inputIsFocusedAndZero;
      var hideStepper = callToAction && !hideCTAButton;
      var disableDecrementBtn = value <= 0;
      var disableIncrementBtn = value >= clampedMax;
      var isInvalid = this.isInvalid(value);
      var invalidMessage = isInvalid ? this.isLessThanMinQuantity(value) ? "".concat(value, " is less than the minimum quantity.") : "".concat(value, " is more than the maximum quantity.") : '';
      var rangeLimitMessage = '';
      /** aria-live message for out-of-range values. */

      if (value <= 0) {
        rangeLimitMessage = 'Minimum Quantity Reached,';
      } else if (value >= clampedMax) {
        rangeLimitMessage = 'Maximum Quantity Reached,';
      }

      return React.createElement("div", _extends({}, divProps, {
        className: cx('kds-Stepper', className)
      }), callToAction && React.createElement(Button, {
        "aria-label": "".concat(callToAction, ": ").concat(what),
        className: cx('kds-Stepper-ctaButton', {
          'kds-Stepper-ctaButton--compact': compact,
          'w-full': fullWidth,
          hidden: hideCTAButton
        }),
        compact: compact,
        "data-testid": "kds-Stepper-ctaButton",
        onClick: this.onCTAButtonClick,
        ref: this.ctaButtonRef,
        kind: kinds.includes(kind) ? kind : undefined
      }, callToAction), React.createElement("div", {
        className: cx('kds-Stepper-wrapper', {
          'text-center': fullWidth,
          hidden: hideStepper
        }),
        "data-testid": "kds-Stepper-wrapper"
      }, React.createElement(Button, {
        "aria-label": "Decrement ".concat(what).concat(description ? ", ".concat(description) : ''),
        tabIndex: disableDecrementBtn ? -1 : 0,
        compact: compact,
        "data-testid": "kds-Stepper-decButton",
        disabled: disableDecrementBtn,
        kind: destructiveDecrement ? 'secondary-negative' : 'cancel',
        onClick: this.onDecrementClick
      }, destructiveDecrement ? React.createElement(IconTrash, null) : React.createElement(IconMinus, null)), React.createElement("div", {
        className: cx('kds-Stepper-inputWrapper', {
          'mx-8': compact
        }),
        "data-testid": "kds-Stepper-inputWrapper"
      }, React.createElement("input", {
        "aria-invalid": isInvalid,
        "aria-live": "polite",
        "aria-label": "".concat(invalidMessage || rangeLimitMessage || 'Quantity of', " ").concat(what).concat(description ? ", ".concat(description) : ''),
        className: cx('kds-Stepper-input', {
          'kds-Stepper-input--descriptive': description
        }),
        "data-testid": "kds-Stepper-input",
        onBlur: this.onInputBlur,
        onChange: this.onInputChange,
        onKeyPress: this.onInputKeyPress,
        "aria-valuemax": clampedMax,
        max: clampedMax,
        "aria-valuemin": "0",
        min: "0",
        type: "text",
        inputMode: "numeric",
        "aria-valuenow": value,
        value:
        /* toString() is necessary to force the browser to strip leading zeros. ¯\_(ツ)_/¯ */
        value.toString(),
        style: {
          /* The input is resized to be wide enough for the value with a little wiggle room */
          width: "".concat(value.toString().length * 1.5, "ch")
        },
        tabIndex:
        /* This is required to fix a bug where this button is not perceived as focusable by some versions of VO + Safari. */
        0
      }), description && React.createElement(Text, {
        "aria-hidden": "true",
        className: "kds-Stepper-description",
        size: "s"
      }, description)), React.createElement(Button, {
        "aria-label": "Increment ".concat(what).concat(description ? ", ".concat(description) : ''),
        tabIndex: disableIncrementBtn ? -1 : 0,
        ref: this.incrBtnRef,
        compact: compact,
        "data-testid": "kds-Stepper-incButton",
        disabled: disableIncrementBtn,
        onClick: this.onIncrementClick
      }, React.createElement(IconPlus, null))));
    }
  }], [{
    key: "getDerivedStateFromProps",

    /* Handle updates to updatedValue prop and set value on state */
    value: function getDerivedStateFromProps(_ref3, _ref4) {
      var propValue = _ref3.value;
      var updatedValue = _ref4.updatedValue,
          locked = _ref4.locked;

      if (!locked && propValue !== updatedValue) {
        /* In this event we "clean" the value first because a user could have typed anything into the input. */
        var cleanedValue = parseInt(propValue) || 0;
        return {
          inputIsFocusedAndZero: cleanedValue !== 0,
          // indicate when input has focus and zero value
          value: cleanedValue,
          updatedValue: propValue
        };
      } else {
        return {
          updatedValue: propValue
        };
      }
    }
  }]);

  return Stepper;
}(PureComponent);

_defineProperty(Stepper, "defaultProps", {
  callToAction: '',
  compact: false,
  description: '',
  destructive: false,
  fullWidth: false,
  value: 0,
  onChange: function onChange() {},
  max: 999,
  minimumQuantity: 1,
  kind: 'primary'
});

export default Stepper;
