// kds-react/build/es/components/Portal/_Portal.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import { Component } from 'react';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { canUseDOM, getWindow } from '../../utils/windowUtils.js';
import ReactDOM from 'react-dom';

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

var Portal =
/*#__PURE__*/
function (_Component) {
  _inherits(Portal, _Component);

  var _super = _createSuper(Portal);

  function Portal() {
    _classCallCheck(this, Portal);

    return _super.apply(this, arguments);
  }

  _createClass(Portal, [{
    key: "UNSAFE_componentWillMount",
    value: function UNSAFE_componentWillMount() {
      var id = "kds-Portal-".concat(this.props.id);
      var className = this.props.className;
      var window = getWindow();

      if (!canUseDOM()) {
        return;
      }

      var portalEl = window.document.getElementById(id);

      if (!portalEl) {
        portalEl = window.document.createElement('div');
        portalEl.setAttribute('id', id);
        portalEl.setAttribute('class', "kds-Portal pointer-events-none ".concat(className));
        window.document.body.appendChild(portalEl);
      }

      this.portalEl = portalEl;
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.portalEl) {
        return null;
      }

      return ReactDOM.createPortal(this.props.children, this.portalEl);
    }
  }]);

  return Portal;
}(Component);

_defineProperty(Portal, "defaultProps", {
  id: 'default'
});

export default Portal;
