// kds-react/build/es/components/Accordion/ParagraphAccordion/ParagraphAccordion.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { PureComponent } from 'react';
import cx from 'classnames';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { callSafely } from '../../../utils/functionUtils.js';
import { generateUniqueString } from '../../../utils/generationUtils.js';
import Button from '../../Button/Button.js';
import Paragraph from '../../Paragraph/Paragraph.js';

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}
/**
 * Accordions allow horizontal portions of the page to show or be hidden in response to user actions.
 *
 * `<ParagraphAccordion>` is used for lengthy text that is being truncated. The `leadingText` provided is displayed with a button to show more content. Once the user interacts with the accordion, all of the remaining text is displayed.
 *
 * For creating groups of accordions, the [AccordionGroup](./accordiongroup) component can be used.
 */

var ParagraphAccordion =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(ParagraphAccordion, _PureComponent);

  var _super = _createSuper(ParagraphAccordion);

  function ParagraphAccordion() {
    var _this;

    _classCallCheck(this, ParagraphAccordion);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      open: _this.props.defaultOpen
    });

    _defineProperty(_assertThisInitialized(_this), "uniqueId", _this.props.id || "kds-ParagraphAccordion-".concat(generateUniqueString()));

    _defineProperty(_assertThisInitialized(_this), "calculateExpandedHeight", function (accordionContentEl) {
      var el = accordionContentEl || document.getElementById(_this.uniqueId).querySelector('.kds-ParagraphAccordion-content');
      return el.scrollHeight;
    });

    _defineProperty(_assertThisInitialized(_this), "handleClick", function (event) {
      var open = !_this.state.open;
      var accordion = document.getElementById(_this.uniqueId);
      var accContent = accordion.querySelector('.kds-ParagraphAccordion-content');

      var expandedHeight = _this.calculateExpandedHeight(accContent); // If collapsing, we need to set the height to px value for CSS transition to work BEFORE setting state.


      if (!open) {
        accContent.setAttribute('style', "height: ".concat(expandedHeight, "px"));
      }

      _this.setState({
        open: open,
        expandedHeight: expandedHeight
      });

      callSafely(_this.props.onClick)({
        event: event,
        isOpen: open
      });
    });

    return _this;
  }

  _createClass(ParagraphAccordion, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          className = _this$props.className,
          defaultOpen = _this$props.defaultOpen,
          leadingText = _this$props.leadingText,
          children = _this$props.children,
          onClick = _this$props.onClick,
          Tag = _this$props.tag,
          expandButtonText = _this$props.expandButtonText,
          collapseButtonText = _this$props.collapseButtonText,
          showTrailingCollapseButton = _this$props.showTrailingCollapseButton,
          restProps = _objectWithoutProperties(_this$props, ["className", "defaultOpen", "leadingText", "children", "onClick", "tag", "expandButtonText", "collapseButtonText", "showTrailingCollapseButton"]);

      var open = this.state.open;
      var ParagraphToButtonTextSizeMap = {
        l: 'm',
        m: 's',
        s: 's',
        xs: 'xs'
      };
      var ctaButton = React.createElement(Button, {
        compact: true,
        kind: "secondary",
        className: cx('kds-ParagraphAccordion-ctaButton', "body-".concat(leadingText.props && leadingText.props.size ? ParagraphToButtonTextSizeMap[leadingText.props.size] : 'm')),
        "aria-controls": "".concat(this.uniqueId, "-content"),
        "aria-expanded": open ? 'true' : 'false',
        onClick: this.handleClick
      }, open ? collapseButtonText : expandButtonText);
      return React.createElement(Tag, _extends({
        id: this.uniqueId,
        className: cx('kds-ParagraphAccordion line-length', {
          'kds-ParagraphAccordion--open': open
        }, className)
      }, restProps), leadingText && React.createElement("div", {
        className: "kds-ParagraphAccordion-leadingContent"
      }, React.isValidElement(leadingText) && leadingText.type === Paragraph && React.cloneElement(leadingText, {
        className: cx('kds-ParagraphAccordion-leadingText', leadingText.props.className, 'inline')
      }), ctaButton), React.createElement("div", {
        id: "".concat(this.uniqueId, "-content"),
        className: cx('kds-ParagraphAccordion-content', {
          hasTrailingButton: showTrailingCollapseButton
        }),
        tabIndex: "-1",
        ref: function ref(el) {
          if (el) {
            var expandedHeight = _this2.calculateExpandedHeight(el);

            el.setAttribute('style', "height: ".concat(open ? expandedHeight : 0, "px"));
            setTimeout(function () {
              open && el.setAttribute('style', "height: auto");
            }, 300);
          }
        }
      }, children, showTrailingCollapseButton && React.cloneElement(ctaButton, {
        tabIndex: open ? 0 : -1
      })));
    }
  }]);

  return ParagraphAccordion;
}(PureComponent);

_defineProperty(ParagraphAccordion, "defaultProps", {
  defaultOpen: false,
  tag: 'section',
  expandButtonText: 'Show More',
  collapseButtonText: 'Show Less',
  showTrailingCollapseButton: false
});

export default ParagraphAccordion;
