// kds-react/build/cjs/components/Portal/_Portal.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var windowUtils = require('../../utils/windowUtils.js');
var ReactDOM = _interopDefault(require('react-dom'));

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

var Portal =
/*#__PURE__*/
function (_Component) {
  _inherits(Portal, _Component);

  var _super = _createSuper(Portal);

  function Portal() {
    _classCallCheck(this, Portal);

    return _super.apply(this, arguments);
  }

  _createClass(Portal, [{
    key: "UNSAFE_componentWillMount",
    value: function UNSAFE_componentWillMount() {
      var id = "kds-Portal-".concat(this.props.id);
      var className = this.props.className;
      var window = windowUtils.getWindow();

      if (!windowUtils.canUseDOM()) {
        return;
      }

      var portalEl = window.document.getElementById(id);

      if (!portalEl) {
        portalEl = window.document.createElement('div');
        portalEl.setAttribute('id', id);
        portalEl.setAttribute('class', "kds-Portal pointer-events-none ".concat(className));
        window.document.body.appendChild(portalEl);
      }

      this.portalEl = portalEl;
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.portalEl) {
        return null;
      }

      return ReactDOM.createPortal(this.props.children, this.portalEl);
    }
  }]);

  return Portal;
}(React.Component);

_defineProperty(Portal, "defaultProps", {
  id: 'default'
});

exports.default = Portal;
