// kds-react/build/cjs/components/Tag/Tag.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var _slicedToArray = _interopDefault(require('@babel/runtime/helpers/slicedToArray'));
var DismissalButton = require('../DismissalButton/DismissalButton.js');
var reactTransitionGroup = require('react-transition-group');

/**
 * Tags are compact elements that help organize, categorize, and control content.
 */

var Tag = React__default.forwardRef(function (_ref, ref) {
  var ElementTag = _ref.tag,
      kind = _ref.kind,
      className = _ref.className,
      onRemove = _ref.onRemove,
      Icon = _ref.icon,
      role = _ref.role,
      text = _ref.text,
      compact = _ref.compact,
      ariaLabel = _ref['aria-label'],
      focusIndex = _ref['data-index'],
      props = _objectWithoutProperties(_ref, ["tag", "kind", "className", "onRemove", "icon", "role", "text", "compact", "aria-label", "data-index"]);

  var onRemoveHandler = function onRemoveHandler() {
    removable && setFadeInOutState('FadeOut'); //  fade out and remove from DOM

    if (onRemove) {
      onRemove(focusIndex); //  remove from TagGroup tag state and move focus to sibling
    }
  };

  var _useState = React.useState('FadeIn'),
      _useState2 = _slicedToArray(_useState, 2),
      fadeInOutState = _useState2[0],
      setFadeInOutState = _useState2[1];

  var removable = typeof onRemove === 'function';
  var tagClassName = cx('kds-Tag', "kds-Tag--".concat(kind), {
    'kds-Tag--compact': !removable && compact
  }, className);
  var ariaLabelTxt = ariaLabel === undefined ? text : ariaLabel;
  return React__default.createElement(reactTransitionGroup.CSSTransition, _extends({
    in: removable ? fadeInOutState == 'FadeIn' : false,
    appear: true,
    timeout: 200,
    classNames: {
      appearActive: cx(tagClassName, 'is-added'),
      enterDone: tagClassName,
      exitActive: cx(tagClassName, 'is-removed'),
      exitDone: cx(tagClassName, 'is-removed')
    },
    unmountOnExit: !!removable,
    "data-index": focusIndex
  }, props), React__default.createElement(ElementTag, {
    role: role,
    "aria-label": "Tag: ".concat(ariaLabelTxt),
    className: tagClassName
  }, Icon && React__default.createElement(Icon, {
    className: "kds-Tag-Icon mr-4",
    size: "s"
  }), React__default.createElement("span", {
    className: "kds-Tag-text"
  }, text), removable && React__default.createElement(DismissalButton.default, {
    className: "kds-Tag-dismissalButton",
    "aria-label": "remove Tag: ".concat(ariaLabelTxt),
    kind: kind,
    onClick: onRemoveHandler,
    ref: ref
  })));
});
Tag.displayName = 'Tag';
Tag.defaultProps = {
  kind: 'default',
  role: 'note',
  tag: 'div',
  compact: false
};

exports.default = Tag;
