// kds-react/build/cjs/components/StatusIndicator/StatusIndicator.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));

var kindOpts = ['positive', 'callout', 'negative'];

var getValidKind = function getValidKind(kind) {
  return kindOpts.includes(kind) ? kind : 'positive';
};
/**
 * `<StatusIndicator>` provides context to the user on the status of an order, system process, or a datapoint within the UI.
 */


var StatusIndicator = function StatusIndicator(_ref) {
  var children = _ref.children,
      Tag = _ref.tag,
      className = _ref.className,
      kind = _ref.kind,
      props = _objectWithoutProperties(_ref, ["children", "tag", "className", "kind"]);

  return children ? React__default.createElement(Tag, _extends({}, props, {
    className: cx('kds-StatusIndicator', "kds-StatusIndicator--".concat(getValidKind(kind)), className)
  }), children) : '';
};

StatusIndicator.defaultProps = {
  kind: 'positive',
  tag: 'span'
};

exports.default = StatusIndicator;
exports.kindOpts = kindOpts;
