// kds-react/src/components/Form/Form.js
import cx from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import { grabFormData } from '../../utils/formUtils'
import { callAll, callSafely } from '../../utils/functionUtils'
import * as prebuiltValidations from './validations/_prebuiltValidations'

/**
 * A `Form` is a collection of inputs that allow the user to enter data. Form inputs are automatically laid out
 * in a design-system-compliant manner.
 *
 * `Form` has a set of pre-built validations available on [`Form.validations`](./form#Validations).
 *
 * > We defer to native browser behavior of form controls in regard to functionality and style as much as possible. We minimally augment the native functionality, but also provide means to work around these features in most cases. See [ADR 7: Minimally augment browser native behavior for form controls](adr-007-minimally-augment-browser-native-behavior-for-form-controls.md)
 *
 * **Requires *custom-event* polyfill for custom validations to work in IE11** [See Polyfills](./polyfills)
 */
class Form extends React.Component {
  static propTypes = {
    /** A function to be called on form submission.
     *
     > If provided, we will call `event.preventDefault()` on your behalf because this is a majority use case scenario.
     > If you don't want this behavior, pass `false` to the `preventDefaultInOnSubmitProp` prop.
     *
     * @param { object } formData An object of input names matched to input values
     * @param { object } event The submit event object
     */
    onSubmit: PropTypes.func,
    /** A function to be called when a form submission attempt failed because of validation errors.
     * @param {Object[]} submittedFormDetails - An array of the details of all the form fields in the form, with the folowing shape:
     * @param {string} submittedFormDetails[].name - 'name' attribute of the form field.
     * @param {string} submittedFormDetails[].value - Value of the form field.
     * @param {boolean} submittedFormDetails[].isValid - Whether the form field is valid or not.
     * @param {Object[]} submittedFormDetails[].messagesShown - An array of messages shown for the form field, with the following shape:
     * @param {string} messagesShown[].messageText - Message text.
     * @param {string} messagesShown[].messageKind - One of 'error' or 'success'.
     */
    onSubmitValidationError: PropTypes.func,
    /** A function to be called when validation state changes.
     *
     * @param { object } event A validation event object
     * @param { object } event.target The element on which the validation messages changed
     * @param { array } event.detail.messagesShown A list of the message types shown to the user
     */
    onValidationChange: PropTypes.func,
    /** A function to be called when the form is reset. */
    onReset: PropTypes.func,
    /** Pass utility classes to the container element */
    className: PropTypes.string,
    /** If `false`, then `event.preventDefault()` will not be called before the `onSubmit` prop is called. */
    preventDefaultInOnSubmitProp: PropTypes.bool,
    /** The HTTP method that the form uses to create a request. */
    method: PropTypes.string,
    /** Allow users to handle the ref on the form. Can be used to reset the form if needed. */
    formRef: PropTypes.ref
  }

  static defaultProps = {
    method: 'POST',
    preventDefaultInOnSubmitProp: true
  }

  static validations = prebuiltValidations

  constructor (props) {
    super(props)

    this.formRef = props.formRef || React.createRef()
  }

  componentDidMount () {
    if (this.props.onValidationChange instanceof Function) {
      this.formRef.current.addEventListener(
        'validation',
        this.props.onValidationChange
      )
    }
  }

  componentDidUpdate (prevProps) {
    if (this.props.onValidationChange !== prevProps.onValidationChange) {
      this.formRef.current.removeEventListener(
        'validation',
        prevProps.onValidationChange
      )
      this.formRef.current.addEventListener(
        'validation',
        this.props.onValidationChange
      )
    }
  }

  componentWillUnmount () {
    if (this.props.onValidationChange instanceof Function) {
      this.formRef.current.removeEventListener(
        'validation',
        this.props.onValidationChange
      )
    }
  }

  handleSubmit = event => {
    const {
      onSubmit,
      preventDefaultInOnSubmitProp,
      onSubmitValidationError
    } = this.props
    const allFormElements = Array.from(event.target.elements).map(el => ({
      el,
      valid: el.checkValidity()
    }))
    const invalidElements = allFormElements.filter(({ valid }) => !valid)
    const isValid = invalidElements.length === 0
    const formData = grabFormData(event.target)

    if (isValid) {
      preventDefaultInOnSubmitProp && onSubmit && event.preventDefault()

      return callSafely(onSubmit)(formData, event)
    } else {
      event.preventDefault()
      setTimeout(() => {
        invalidElements[0].el.focus()

        // Invoke onSubmitValidationError prop if the submission failed because of invalid fields:
        const submittedFormDetails = Object.keys(formData).map(key => {
          // Below is for field that would have multiple elements with the same 'name' attribute (like radio buttons), and the last element with that 'name' would have the properties we need
          const elementsWithCurrentName = allFormElements.filter(
            ({ el }) => el.name === key
          )
          const current =
            elementsWithCurrentName[elementsWithCurrentName.length - 1]

          return {
            name: key,
            value: formData[key],
            isValid: current.valid,
            messagesShown: current.valid
              ? []
              : JSON.parse(current.el.getAttribute('data-messagesShown'))
          }
        })

        callSafely(onSubmitValidationError)(submittedFormDetails)
      }, 0)
    }
  }

  handleReset = event => {
    const elements = event.target.elements
    Array.from(elements).forEach(el => {
      callSafely(el.restoreDefaultState)()
    })
  }

  render () {
    const {
      preventDefaultInOnSubmitProp,
      onSubmit,
      formRef,
      onValidationChange,
      onReset,
      onSubmitValidationError,
      ...props
    } = this.props
    return (
      <form
        {...props}
        ref={this.formRef}
        className={cx('kds-Form', this.props.className)}
        onSubmit={this.handleSubmit}
        onReset={callAll(this.handleReset, onReset)}
      />
    )
  }
}

export default Form
