// kds-react/src/components/ProminentLink/ProminentLink.js
import cx from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import Link from '../Link/Link'
import { IconRightArrow } from '../Icon/index'

export const sizeOpts = ['l', 'xl']

/**
 * ProminentLink is used to add additional visual weight to a link. They have a built-in appropriately sized right-arrow icon after the link text. See: [`ProminentLink` Design Guidelines](http://design.kroger.com/#/design/components/prominentlink)
 *
 * > **Note:** `<ProminentLink />` renders a `<Link />` with most props spread on.
 * For more details, see the [Link](./link) component.
 */

const ProminentLink = ({
  className,
  /* `implied` and `explicit` are destructured to prevent from passing into Link as ProminentLinks can not be implied or explicit */
  implied,
  explicit,
  size,
  children,
  ...props
}) => {
  return (
    <Link
      {...props}
      className={cx(
        'kds-ProminentLink',
        `kds-ProminentLink--${sizeOpts.includes(size) ? size : 'l'}`,
        className
      )}
      size="manual"
    >
      {children}
      <IconRightArrow className="kds-ProminentLink-icon" />
    </Link>
  )
}

ProminentLink.propTypes = {
  /** Optional classname */
  className: PropTypes.string,
  /** Set to `true` if the link is placed on a dark background. */
  inverse: PropTypes.string.bool,
  /** Set to `true` if you want to render a React Router `NavLink` */
  withRouter: PropTypes.bool,
  size: PropTypes.oneOf(sizeOpts)
}

ProminentLink.defaultProps = {
  href: '#',
  inverse: false,
  withRouter: false,
  size: 'l'
}

export default ProminentLink
