// kds-react/build/es/components/Modal/Modal.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { PureComponent } from 'react';
import cx from 'classnames';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import { getWindow } from '../../utils/windowUtils.js';
import '../Icon/icons/AcceptEntry.js';
import '../Icon/icons/AccountSummary.js';
import '../Icon/icons/Account.js';
import '../Icon/icons/AddCard.js';
import '../Icon/icons/AddColumn.js';
import '../Icon/icons/AddPatient.js';
import '../Icon/icons/AddPlusCard.js';
import '../Icon/icons/AddPrescription.js';
import '../Icon/icons/AdultBeverage.js';
import '../Icon/icons/Airplane.js';
import '../Icon/icons/Android.js';
import '../Icon/icons/Animals.js';
import '../Icon/icons/ApiAuthorize.js';
import '../Icon/icons/ApiProduct.js';
import '../Icon/icons/Approved.js';
import '../Icon/icons/Ar.js';
import '../Icon/icons/Associate.js';
import '../Icon/icons/Asterisk.js';
import '../Icon/icons/Audience.js';
import '../Icon/icons/AutoRefill.js';
import '../Icon/icons/Baby.js';
import '../Icon/icons/Backstock.js';
import '../Icon/icons/BagBarcode.js';
import '../Icon/icons/Bags.js';
import '../Icon/icons/BarGraph.js';
import '../Icon/icons/BarLineGraph.js';
import '../Icon/icons/Barcode.js';
import '../Icon/icons/BasketTime.js';
import '../Icon/icons/Basket.js';
import '../Icon/icons/BeautyPersonalCare.js';
import '../Icon/icons/Beta.js';
import '../Icon/icons/BlankBox.js';
import '../Icon/icons/BuyWireless.js';
import '../Icon/icons/Calendar.js';
import '../Icon/icons/Camera.js';
import '../Icon/icons/Candy.js';
import '../Icon/icons/CapsLock.js';
import '../Icon/icons/CardView.js';
import '../Icon/icons/CarouselView.js';
import '../Icon/icons/CaretLeft.js';
import '../Icon/icons/CaretRight.js';
import '../Icon/icons/Cart.js';
import '../Icon/icons/CashBack.js';
import '../Icon/icons/CashRegister.js';
import '../Icon/icons/CcAmex.js';
import '../Icon/icons/CcDinersClub.js';
import '../Icon/icons/CcDiscover.js';
import '../Icon/icons/CcMastercard.js';
import '../Icon/icons/CcVisa.js';
import '../Icon/icons/CellPhone.js';
import '../Icon/icons/CenterAlign.js';
import '../Icon/icons/CheckMark.js';
import '../Icon/icons/ChevronDown.js';
import '../Icon/icons/ChevronUp.js';
import '../Icon/icons/CivicOrganizations.js';
import '../Icon/icons/CleaningHousehold.js';
import '../Icon/icons/Clock.js';
import '../Icon/icons/Close.js';
import '../Icon/icons/CodePhone.js';
import '../Icon/icons/Coffee.js';
import '../Icon/icons/ContainsDairy.js';
import '../Icon/icons/ContainsArtificialIngredients.js';
import '../Icon/icons/Comment.js';
import '../Icon/icons/CompanyDirect.js';
import '../Icon/icons/ContractViewport.js';
import '../Icon/icons/ControlPayment.js';
import '../Icon/icons/Convenience.js';
import '../Icon/icons/CollapseAll.js';
import '../Icon/icons/Copy.js';
import '../Icon/icons/Coupons.js';
import '../Icon/icons/CreateNew.js';
import '../Icon/icons/CreditCard.js';
import '../Icon/icons/CurrentLocationAndroid.js';
import '../Icon/icons/CurrentLocationIos.js';
import '../Icon/icons/DamagedItem.js';
import '../Icon/icons/DarkLight.js';
import '../Icon/icons/Dashboard.js';
import '../Icon/icons/DebitCard.js';
import '../Icon/icons/DeliBakery.js';
import '../Icon/icons/Delivery.js';
import '../Icon/icons/DiabeticNeeds.js';
import '../Icon/icons/Directions.js';
import '../Icon/icons/Diversity.js';
import '../Icon/icons/DownArrow.js';
import '../Icon/icons/Download.js';
import '../Icon/icons/DragHandle.js';
import '../Icon/icons/Draggable.js';
import '../Icon/icons/DrugInteractions.js';
import '../Icon/icons/DrugSearch.js';
import '../Icon/icons/Ebt.js';
import '../Icon/icons/Edit.js';
import '../Icon/icons/Education.js';
import '../Icon/icons/EggFree.js';
import '../Icon/icons/Ellipsis.js';
import '../Icon/icons/Email.js';
import '../Icon/icons/EmergencyContact.js';
import '../Icon/icons/Environmental.js';
import '../Icon/icons/Error.js';
import '../Icon/icons/ExclusiveSavings.js';
import '../Icon/icons/Exit.js';
import '../Icon/icons/ExpandAll.js';
import '../Icon/icons/ExpandViewport.js';
import '../Icon/icons/ExpenseReport.js';
import '../Icon/icons/Expiration.js';
import '../Icon/icons/Export.js';
import '../Icon/icons/External.js';
import '../Icon/icons/EyeSlash.js';
import '../Icon/icons/Eye.js';
import '../Icon/icons/Failure.js';
import '../Icon/icons/FairService.js';
import '../Icon/icons/FaithBased.js';
import '../Icon/icons/Faq.js';
import '../Icon/icons/FatFree.js';
import '../Icon/icons/FavoriteSolid.js';
import '../Icon/icons/Favorite.js';
import '../Icon/icons/Fee.js';
import '../Icon/icons/Feedback.js';
import '../Icon/icons/File.js';
import '../Icon/icons/Filters.js';
import '../Icon/icons/Flag.js';
import '../Icon/icons/FlagSolid.js';
import '../Icon/icons/FlashOff.js';
import '../Icon/icons/FlashOn.js';
import '../Icon/icons/Flashlight.js';
import '../Icon/icons/Floral.js';
import '../Icon/icons/FmJewelers.js';
import '../Icon/icons/FuelPump.js';
import '../Icon/icons/Fuel.js';
import '../Icon/icons/GanttChart.js';
import '../Icon/icons/GiftCardMall.js';
import '../Icon/icons/GiftCard.js';
import '../Icon/icons/GlutenFree.js';
import '../Icon/icons/GoodService.js';
import '../Icon/icons/GoogleHome.js';
import '../Icon/icons/GridView.js';
import '../Icon/icons/GroceriesHome.js';
import '../Icon/icons/Groceries.js';
import '../Icon/icons/HandSanitizer.js';
import '../Icon/icons/HamburgerMenuNative.js';
import '../Icon/icons/HamburgerMenu.js';
import '../Icon/icons/HealthNutrition.js';
import '../Icon/icons/HealthWellness.js';
import '../Icon/icons/HeatEat.js';
import '../Icon/icons/Hierarchy.js';
import '../Icon/icons/HighProtein.js';
import '../Icon/icons/History.js';
import '../Icon/icons/HomeElectronics.js';
import '../Icon/icons/Home.js';
import '../Icon/icons/HungerRelief.js';
import '../Icon/icons/ImageMissing.js';
import '../Icon/icons/Import.js';
import '../Icon/icons/Info.js';
import '../Icon/icons/Ingredients.js';
import '../Icon/icons/InProgress.js';
import '../Icon/icons/InsuranceInfo.js';
import '../Icon/icons/IosShare.js';
import '../Icon/icons/Ios.js';
import '../Icon/icons/Kosher.js';
import '../Icon/icons/KrogerCreditCard.js';
import '../Icon/icons/KrogerRewards.js';
import '../Icon/icons/LactoseFree.js';
import '../Icon/icons/Layers.js';
import '../Icon/icons/LeftAlign.js';
import '../Icon/icons/LeftArrow.js';
import '../Icon/icons/Leftover.js';
import '../Icon/icons/LightBulb.js';
import '../Icon/icons/Link.js';
import '../Icon/icons/ListView.js';
import '../Icon/icons/LoadedSavings.js';
import '../Icon/icons/LocationA.js';
import '../Icon/icons/LocationB.js';
import '../Icon/icons/LocationRx.js';
import '../Icon/icons/Location.js';
import '../Icon/icons/Locked.js';
import '../Icon/icons/LockerPasscode.js';
import '../Icon/icons/LowFat.js';
import '../Icon/icons/Map.js';
import '../Icon/icons/MealKit.js';
import '../Icon/icons/MeatSeafood.js';
import '../Icon/icons/MedicationToolkit.js';
import '../Icon/icons/Members.js';
import '../Icon/icons/Merge.js';
import '../Icon/icons/MilitaryPoliceFire.js';
import '../Icon/icons/Minus.js';
import '../Icon/icons/MissingItem.js';
import '../Icon/icons/MoneyServices.js';
import '../Icon/icons/Moon.js';
import '../Icon/icons/MoreOptions.js';
import '../Icon/icons/Natural.js';
import '../Icon/icons/Newest.js';
import '../Icon/icons/NoFees.js';
import '../Icon/icons/NonGmo.js';
import '../Icon/icons/NotAvailable.js';
import '../Icon/icons/Notification.js';
import '../Icon/icons/NuroDeliver.js';
import '../Icon/icons/NuroPickup.js';
import '../Icon/icons/NutFree.js';
import '../Icon/icons/Offline.js';
import '../Icon/icons/OnlinePayment.js';
import '../Icon/icons/OptUp.js';
import '../Icon/icons/Organic.js';
import '../Icon/icons/OtherMisc.js';
import '../Icon/icons/Oven.js';
import '../Icon/icons/Paleo.js';
import '../Icon/icons/Partnership.js';
import '../Icon/icons/Paste.js';
import '../Icon/icons/PatientAccess.js';
import '../Icon/icons/PatientWellness.js';
import '../Icon/icons/PaymentSuccess.js';
import '../Icon/icons/Pencil.js';
import '../Icon/icons/PersonHome.js';
import '../Icon/icons/PersonStore.js';
import '../Icon/icons/PersonalCare.js';
import '../Icon/icons/PetMedications.js';
import '../Icon/icons/PharmacyNotifications.js';
import '../Icon/icons/PharmacySearch.js';
import '../Icon/icons/Pharmacy.js';
import '../Icon/icons/Phone.js';
import '../Icon/icons/Pickup.js';
import '../Icon/icons/PieChart.js';
import '../Icon/icons/PiggyBank.js';
import '../Icon/icons/PillIdentifier.js';
import '../Icon/icons/PlusCard.js';
import '../Icon/icons/Plus.js';
import '../Icon/icons/PoorService.js';
import '../Icon/icons/PoorSubstitution.js';
import '../Icon/icons/Popularity.js';
import '../Icon/icons/PreferredCard.js';
import '../Icon/icons/PrescriptionHistory.js';
import '../Icon/icons/PrescriptionStatus.js';
import '../Icon/icons/Print.js';
import '../Icon/icons/PrivacyPolicy.js';
import '../Icon/icons/Produce.js';
import '../Icon/icons/PromotionOnAd.js';
import '../Icon/icons/PromotionOnDisplay.js';
import '../Icon/icons/Promotions.js';
import '../Icon/icons/Publish.js';
import '../Icon/icons/PurchaseHistory.js';
import '../Icon/icons/PushNotification.js';
import '../Icon/icons/PushPin.js';
import '../Icon/icons/PushPinSolid.js';
import '../Icon/icons/PuzzlePiece.js';
import '../Icon/icons/QrCode.js';
import '../Icon/icons/RadioChecked.js';
import '../Icon/icons/RadioUnchecked.js';
import '../Icon/icons/RatingsHalf.js';
import '../Icon/icons/RatingsSolid.js';
import '../Icon/icons/Ratings.js';
import '../Icon/icons/ReWeigh.js';
import '../Icon/icons/React.js';
import '../Icon/icons/Recall.js';
import '../Icon/icons/Receipt.js';
import '../Icon/icons/RecentSearch.js';
import '../Icon/icons/Recipes.js';
import '../Icon/icons/Recycle.js';
import '../Icon/icons/Redo.js';
import '../Icon/icons/RefillFromList.js';
import '../Icon/icons/RefillPrograms.js';
import '../Icon/icons/RejectEntry.js';
import '../Icon/icons/Relevance.js';
import '../Icon/icons/RemoveColumn.js';
import '../Icon/icons/Repurchase.js';
import '../Icon/icons/Reset.js';
import '../Icon/icons/Restrooms.js';
import '../Icon/icons/Returns.js';
import '../Icon/icons/Revert.js';
import '../Icon/icons/Rewards.js';
import '../Icon/icons/RightArrow.js';
import '../Icon/icons/RxBarcode.js';
import '../Icon/icons/Route.js';
import '../Icon/icons/Sandwich.js';
import '../Icon/icons/SaveTime.js';
import '../Icon/icons/Scan.js';
import '../Icon/icons/ScanBagGo.js';
import '../Icon/icons/ScheduleAppointment.js';
import '../Icon/icons/Search.js';
import '../Icon/icons/SecurityQuestions.js';
import '../Icon/icons/SecurityTag.js';
import '../Icon/icons/Send.js';
import '../Icon/icons/Settings.js';
import '../Icon/icons/Share.js';
import '../Icon/icons/ShipToHome.js';
import '../Icon/icons/Ship.js';
import '../Icon/icons/ShopInStore.js';
import '../Icon/icons/ShopOnline.js';
import '../Icon/icons/Shop.js';
import '../Icon/icons/ShoppingList.js';
import '../Icon/icons/SkipToEnd.js';
import '../Icon/icons/SkipToStart.js';
import '../Icon/icons/SocialServices.js';
import '../Icon/icons/Sort.js';
import '../Icon/icons/SortAz.js';
import '../Icon/icons/SortZa.js';
import '../Icon/icons/SoyFree.js';
import '../Icon/icons/Status.js';
import '../Icon/icons/Stencil.js';
import '../Icon/icons/Stethoscope.js';
import '../Icon/icons/Stock.js';
import '../Icon/icons/StoreLocker.js';
import '../Icon/icons/Store.js';
import '../Icon/icons/Success.js';
import '../Icon/icons/Sun.js';
import '../Icon/icons/Supercharged.js';
import '../Icon/icons/TableView.js';
import '../Icon/icons/TermsConditions.js';
import '../Icon/icons/ThumbsDown.js';
import '../Icon/icons/ThumbsUp.js';
import '../Icon/icons/Transfer.js';
import '../Icon/icons/TransferCircle.js';
import '../Icon/icons/TransferCircleSolid.js';
import '../Icon/icons/Trash.js';
import '../Icon/icons/Trending.js';
import '../Icon/icons/Undo.js';
import '../Icon/icons/Unlocked.js';
import '../Icon/icons/UpArrow.js';
import '../Icon/icons/Upload.js';
import '../Icon/icons/VaccinationsMadeEasy.js';
import '../Icon/icons/Vaccinations.js';
import '../Icon/icons/Value.js';
import '../Icon/icons/Vegan.js';
import '../Icon/icons/Vegetarian.js';
import '../Icon/icons/Voice.js';
import '../Icon/icons/Wallet.js';
import '../Icon/icons/Warning.js';
import '../Icon/icons/WeeklyAd.js';
import '../Icon/icons/Weigh.js';
import '../Icon/icons/WorldwideAcceptance.js';
import '../Icon/icons/WrongItem.js';
import '../Icon/icons/YouthPrograms.js';
import '../Icon/icons/YtdSavings.js';
import '../Icon/icons/ZoomIn.js';
import '../Icon/icons/ZoomOut.js';
import Heading from '../Heading/Heading.js';
import { callSafely } from '../../utils/functionUtils.js';
import { generateUniqueString } from '../../utils/generationUtils.js';
import Text from '../Text/Text.js';
import Button from '../Button/Button.js';
import DismissalButton from '../DismissalButton/DismissalButton.js';
import '../Image/Image.js';
import ReactModal from 'react-modal';

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}
var validActionButtonKinds = {
  primary: ['primary', 'primary-positive', 'primary-negative'],
  secondary: ['secondary', 'secondary-positive', 'secondary-negative', 'cancel']
};
/**
 * Modal provides a focused view of important details, objects, or workflows.
 *
 * Use modal only when absolutely necessary, not for simple notifications or excessive confirmations. A modal always has a header. It does not have a disabled state.
 *
 * > **Note:** `<Modal>` renders a `react-modal` with most props spread on. Documentation about the API can be accessed [here](https://github.com/reactjs/react-modal#api-documentation). Additional props (for example, `data-*` attributes for testing ) can be provided through `data` prop, and will be added to the content div. See [General Usage](http://reactcommunity.org/react-modal/#general-usage) section for more details.
 */

var Modal =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Modal, _PureComponent);

  var _super = _createSuper(Modal);

  function Modal() {
    var _this;

    _classCallCheck(this, Modal);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "uniqueId", _this.props.id || "kds-Modal-".concat(generateUniqueString()));

    return _this;
  }

  _createClass(Modal, [{
    key: "getSnapshotBeforeUpdate",
    value: function getSnapshotBeforeUpdate(prevProps) {
      var modalIsOpening = this.props.isOpen && !prevProps.isOpen;

      if (modalIsOpening) {
        return getWindow().scrollY;
      }

      return null;
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState, snapshot) {
      var modalHasClosed = !this.props.isOpen && prevProps.isOpen;

      if (modalHasClosed && snapshot !== null) {
        getWindow().scrollTo(0, snapshot);
      }
    }
  }, {
    key: "configureModal",
    value: function configureModal(kind) {
      return {
        // MODAL CONFIGURATION
        isOnboarding: kind === 'onboardingCover' || kind === 'onboardingStep',
        allowClose: kind !== 'actionRequired',
        showProgressIndicator: kind === 'onboardingStep',
        showSecondaryButton: kind !== 'onboardingCover',
        showFooter: kind !== 'base'
      };
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          headingText = _this$props.headingText,
          isOpen = _this$props.isOpen,
          className = _this$props.className,
          kind = _this$props.kind,
          PrimaryActionButton = _this$props.primaryActionButton,
          SecondaryActionButton = _this$props.secondaryActionButton,
          handleCloseModal = _this$props.handleCloseModal,
          children = _this$props.children,
          HeaderIcon = _this$props.headerIcon,
          onboardingStepCurrent = _this$props.onboardingStepCurrent,
          onboardingStepTotal = _this$props.onboardingStepTotal,
          onboardingHeaderImage = _this$props.onboardingHeaderImage,
          onAfterOpen = _this$props.onAfterOpen,
          _this$props$contentLa = _this$props.contentLabel,
          contentLabel = _this$props$contentLa === void 0 ? '' : _this$props$contentLa,
          _this$props$aria = _this$props.aria,
          aria = _this$props$aria === void 0 ? {} : _this$props$aria,
          props = _objectWithoutProperties(_this$props, ["headingText", "isOpen", "className", "kind", "primaryActionButton", "secondaryActionButton", "handleCloseModal", "children", "headerIcon", "onboardingStepCurrent", "onboardingStepTotal", "onboardingHeaderImage", "onAfterOpen", "contentLabel", "aria"]); // MODAL CONFIGURATION


      var _this$configureModal = this.configureModal(kind),
          isOnboarding = _this$configureModal.isOnboarding,
          allowClose = _this$configureModal.allowClose,
          showProgressIndicator = _this$configureModal.showProgressIndicator,
          showSecondaryButton = _this$configureModal.showSecondaryButton,
          showFooter = _this$configureModal.showFooter;

      var heading = React.createElement(React.Fragment, null, HeaderIcon && React.createElement(HeaderIcon, {
        size: "utility-l",
        className: "kds-Modal-headingIcon"
      }), React.createElement(Heading, {
        className: "kds-Modal-heading",
        size: "l",
        tag: "h2",
        id: "".concat(this.uniqueId, "-heading")
      }, headingText));

      var computeAria = function computeAria() {
        // Drop labelledby from aria object because when aria-labelledby is present, it takes precedence over aria-label and some browsers don't focus on the dialog at all when using a screen reader.
        var labelledby = aria.labelledby,
            describedby = aria.describedby,
            rest = _objectWithoutProperties(aria, ["labelledby", "describedby"]);

        var computedDescribedby = _this2.props['aria-describedby'] || describedby || '';

        var computedAria = _objectSpread({
          modal: true
        }, rest);

        if (computedDescribedby.length) {
          computedAria.describedby = computedDescribedby;
        }

        return computedAria;
      };

      return React.createElement(ReactModal, _extends({
        isOpen: isOpen,
        ariaHideApp: false,
        onAfterOpen: callSafely(onAfterOpen),
        onRequestClose: callSafely(handleCloseModal),
        closeTimeoutMS: 100,
        overlayRef: function overlayRef(_overlayRef) {
          return _overlayRef && _overlayRef.setAttribute('role', 'presentation');
        },
        overlayClassName: {
          base: 'kds-Modal-overlay',
          afterOpen: 'kds-Modal-overlay--afterOpen is-open',
          beforeClose: 'kds-Modal-overlay--beforeClose is-closed'
        } // We want to make sure aria-label is always set, even if it is an empty string.
        // Alternatively we could set aria-labelledby="id-for-heading", but aria-label turns out be more accessible when using screen reader after testing.
        // When aria-label="", screen reader announces "dialog" by default
        // When aria-labelledby is present, it takes precedence over aria-label and some browsers don't focus on the dialog when using a screen reader.
        ,
        contentLabel: contentLabel.length ? contentLabel : headingText || '',
        aria: computeAria(),
        id: "".concat(this.uniqueId),
        className: {
          base: cx("kds-Modal kds-Modal--l ".concat(isOnboarding ? 'kds-Modal--onboarding ' : '', "kds-Modal--").concat(kind), className),
          afterOpen: 'kds-Modal--afterOpen is-open',
          beforeClose: 'kds-Modal--beforeClose is-closed'
        },
        bodyOpenClassName: "overflow-hidden",
        htmlOpenClassName: "overflow-hidden",
        shouldCloseOnOverlayClick: allowClose
      }, props), allowClose && React.createElement(DismissalButton, {
        className: "kds-Modal-closeButton",
        "aria-label": "Close pop-up",
        onClick: callSafely(handleCloseModal)
      }), React.createElement("header", {
        className: "kds-Modal-header"
      }, isOnboarding ? React.createElement("div", {
        className: "kds-Modal-headerImage"
      }, onboardingHeaderImage) : heading), React.createElement("div", {
        className: "kds-Modal-body"
      }, React.createElement("div", {
        className: "kds-Modal-content"
      }, isOnboarding && heading, children)), showProgressIndicator && React.createElement(Text, {
        size: "l",
        className: "kds-Modal-progressIndicator"
      }, "".concat(onboardingStepCurrent, " of ").concat(onboardingStepTotal)), showFooter && (PrimaryActionButton || SecondaryActionButton) && React.createElement("footer", {
        className: cx("kds-Modal-footer ".concat(showProgressIndicator ? 'kds-Modal-footer-columns' : ''), className)
      }, React.createElement("div", {
        className: "kds-Modal-footer-buttons"
      }, standardizeActionButton(PrimaryActionButton, 'primary'), showSecondaryButton && standardizeActionButton(SecondaryActionButton, "".concat(PrimaryActionButton ? 'secondary' : 'primary')))));
    }
  }]);

  return Modal;
}(PureComponent);

_defineProperty(Modal, "defaultProps", {
  kind: 'base',
  isOpen: false
});

var isValidButtonKind = function isValidButtonKind(kind, validKinds) {
  return validKinds.filter(function (k) {
    return k === kind;
  }).length > 0;
};

var standardizeActionButton = function standardizeActionButton(button, actionKind) {
  if (button && React.isValidElement(button) && button.type === Button) {
    return React.cloneElement(button, {
      // Make sure button kind is valid. If not use default for the action type
      kind: isValidButtonKind(button.props.kind, validActionButtonKinds[actionKind]) ? button.props.kind : actionKind,
      className: cx(button.props.className, 'kds-Modal-actionButton', "kds-Modal-actionButton--".concat(actionKind))
    });
  }

  return null;
};

export default Modal;
