// kds-react/build/cjs/components/Select/Select.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var PropTypes = _interopDefault(require('prop-types'));
var Label = require('../Label/Label.js');

/**
 * `<Select />` lets users choose from a list of items. Use them for things like:
 * - selecting an item from a list of 3&ndash;15 items
 *
 * > **Note:** `<Select />` renders a `<select>` with all props spread on, like so: `<select {...props} />`.
 * So you can pass any native props such as `id`, `className`, `aria-*` attributes, etc.
 */

var Select = function Select(_ref, ref) {
  var compact = _ref.compact,
      selectProps = _objectWithoutProperties(_ref, ["compact"]);

  return React__default.createElement(Label.LabelContext.ReverseProvider, {
    value: {
      disabled: selectProps.disabled,
      hasField: true
    }
  }, React__default.createElement("select", _extends({
    ref: ref
  }, selectProps, {
    className: cx('kds-Select', {
      'kds-Select--compact': compact
    }, selectProps.className)
  })));
}; // we do this because forwardRef render functions do not support propTypes or defaultProps


var SelectWithForwardRef = React__default.forwardRef(Select);
SelectWithForwardRef.displayName = 'Select';
SelectWithForwardRef.propTypes = {
  /** Compact select size. */
  compact: PropTypes.bool
};
SelectWithForwardRef.defaultProps = {
  compact: false
};

exports.default = SelectWithForwardRef;
