// kds-react/build/es/components/Pagination/Pagination.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React, { useState, useRef, useEffect } from 'react';
import cx from 'classnames';
import { getWindow } from '../../utils/windowUtils.js';
import '../Icon/icons/AcceptEntry.js';
import '../Icon/icons/AccountSummary.js';
import '../Icon/icons/Account.js';
import '../Icon/icons/AddCard.js';
import '../Icon/icons/AddColumn.js';
import '../Icon/icons/AddPatient.js';
import '../Icon/icons/AddPlusCard.js';
import '../Icon/icons/AddPrescription.js';
import '../Icon/icons/AdultBeverage.js';
import '../Icon/icons/Airplane.js';
import '../Icon/icons/Android.js';
import '../Icon/icons/Animals.js';
import '../Icon/icons/ApiAuthorize.js';
import '../Icon/icons/ApiProduct.js';
import '../Icon/icons/Approved.js';
import '../Icon/icons/Ar.js';
import '../Icon/icons/Associate.js';
import '../Icon/icons/Asterisk.js';
import '../Icon/icons/Audience.js';
import '../Icon/icons/AutoRefill.js';
import '../Icon/icons/Baby.js';
import '../Icon/icons/Backstock.js';
import '../Icon/icons/BagBarcode.js';
import '../Icon/icons/Bags.js';
import '../Icon/icons/BarGraph.js';
import '../Icon/icons/BarLineGraph.js';
import '../Icon/icons/Barcode.js';
import '../Icon/icons/BasketTime.js';
import '../Icon/icons/Basket.js';
import '../Icon/icons/BeautyPersonalCare.js';
import '../Icon/icons/Beta.js';
import '../Icon/icons/BlankBox.js';
import '../Icon/icons/BuyWireless.js';
import '../Icon/icons/Calendar.js';
import '../Icon/icons/Camera.js';
import '../Icon/icons/Candy.js';
import '../Icon/icons/CapsLock.js';
import '../Icon/icons/CardView.js';
import '../Icon/icons/CarouselView.js';
import { IconCaretLeft } from '../Icon/icons/CaretLeft.js';
import { IconCaretRight } from '../Icon/icons/CaretRight.js';
import '../Icon/icons/Cart.js';
import '../Icon/icons/CashBack.js';
import '../Icon/icons/CashRegister.js';
import '../Icon/icons/CcAmex.js';
import '../Icon/icons/CcDinersClub.js';
import '../Icon/icons/CcDiscover.js';
import '../Icon/icons/CcMastercard.js';
import '../Icon/icons/CcVisa.js';
import '../Icon/icons/CellPhone.js';
import '../Icon/icons/CenterAlign.js';
import '../Icon/icons/CheckMark.js';
import '../Icon/icons/ChevronDown.js';
import '../Icon/icons/ChevronUp.js';
import '../Icon/icons/CivicOrganizations.js';
import '../Icon/icons/CleaningHousehold.js';
import '../Icon/icons/Clock.js';
import '../Icon/icons/Close.js';
import '../Icon/icons/CodePhone.js';
import '../Icon/icons/Coffee.js';
import '../Icon/icons/ContainsDairy.js';
import '../Icon/icons/ContainsArtificialIngredients.js';
import '../Icon/icons/Comment.js';
import '../Icon/icons/CompanyDirect.js';
import '../Icon/icons/ContractViewport.js';
import '../Icon/icons/ControlPayment.js';
import '../Icon/icons/Convenience.js';
import '../Icon/icons/CollapseAll.js';
import '../Icon/icons/Copy.js';
import '../Icon/icons/Coupons.js';
import '../Icon/icons/CreateNew.js';
import '../Icon/icons/CreditCard.js';
import '../Icon/icons/CurrentLocationAndroid.js';
import '../Icon/icons/CurrentLocationIos.js';
import '../Icon/icons/DamagedItem.js';
import '../Icon/icons/DarkLight.js';
import '../Icon/icons/Dashboard.js';
import '../Icon/icons/DebitCard.js';
import '../Icon/icons/DeliBakery.js';
import '../Icon/icons/Delivery.js';
import '../Icon/icons/DiabeticNeeds.js';
import '../Icon/icons/Directions.js';
import '../Icon/icons/Diversity.js';
import '../Icon/icons/DownArrow.js';
import '../Icon/icons/Download.js';
import '../Icon/icons/DragHandle.js';
import '../Icon/icons/Draggable.js';
import '../Icon/icons/DrugInteractions.js';
import '../Icon/icons/DrugSearch.js';
import '../Icon/icons/Ebt.js';
import '../Icon/icons/Edit.js';
import '../Icon/icons/Education.js';
import '../Icon/icons/EggFree.js';
import { IconEllipsis } from '../Icon/icons/Ellipsis.js';
import '../Icon/icons/Email.js';
import '../Icon/icons/EmergencyContact.js';
import '../Icon/icons/Environmental.js';
import '../Icon/icons/Error.js';
import '../Icon/icons/ExclusiveSavings.js';
import '../Icon/icons/Exit.js';
import '../Icon/icons/ExpandAll.js';
import '../Icon/icons/ExpandViewport.js';
import '../Icon/icons/ExpenseReport.js';
import '../Icon/icons/Expiration.js';
import '../Icon/icons/Export.js';
import '../Icon/icons/External.js';
import '../Icon/icons/EyeSlash.js';
import '../Icon/icons/Eye.js';
import '../Icon/icons/Failure.js';
import '../Icon/icons/FairService.js';
import '../Icon/icons/FaithBased.js';
import '../Icon/icons/Faq.js';
import '../Icon/icons/FatFree.js';
import '../Icon/icons/FavoriteSolid.js';
import '../Icon/icons/Favorite.js';
import '../Icon/icons/Fee.js';
import '../Icon/icons/Feedback.js';
import '../Icon/icons/File.js';
import '../Icon/icons/Filters.js';
import '../Icon/icons/Flag.js';
import '../Icon/icons/FlagSolid.js';
import '../Icon/icons/FlashOff.js';
import '../Icon/icons/FlashOn.js';
import '../Icon/icons/Flashlight.js';
import '../Icon/icons/Floral.js';
import '../Icon/icons/FmJewelers.js';
import '../Icon/icons/FuelPump.js';
import '../Icon/icons/Fuel.js';
import '../Icon/icons/GanttChart.js';
import '../Icon/icons/GiftCardMall.js';
import '../Icon/icons/GiftCard.js';
import '../Icon/icons/GlutenFree.js';
import '../Icon/icons/GoodService.js';
import '../Icon/icons/GoogleHome.js';
import '../Icon/icons/GridView.js';
import '../Icon/icons/GroceriesHome.js';
import '../Icon/icons/Groceries.js';
import '../Icon/icons/HandSanitizer.js';
import '../Icon/icons/HamburgerMenuNative.js';
import '../Icon/icons/HamburgerMenu.js';
import '../Icon/icons/HealthNutrition.js';
import '../Icon/icons/HealthWellness.js';
import '../Icon/icons/HeatEat.js';
import '../Icon/icons/Hierarchy.js';
import '../Icon/icons/HighProtein.js';
import '../Icon/icons/History.js';
import '../Icon/icons/HomeElectronics.js';
import '../Icon/icons/Home.js';
import '../Icon/icons/HungerRelief.js';
import '../Icon/icons/ImageMissing.js';
import '../Icon/icons/Import.js';
import '../Icon/icons/Info.js';
import '../Icon/icons/Ingredients.js';
import '../Icon/icons/InProgress.js';
import '../Icon/icons/InsuranceInfo.js';
import '../Icon/icons/IosShare.js';
import '../Icon/icons/Ios.js';
import '../Icon/icons/Kosher.js';
import '../Icon/icons/KrogerCreditCard.js';
import '../Icon/icons/KrogerRewards.js';
import '../Icon/icons/LactoseFree.js';
import '../Icon/icons/Layers.js';
import '../Icon/icons/LeftAlign.js';
import '../Icon/icons/LeftArrow.js';
import '../Icon/icons/Leftover.js';
import '../Icon/icons/LightBulb.js';
import '../Icon/icons/Link.js';
import '../Icon/icons/ListView.js';
import '../Icon/icons/LoadedSavings.js';
import '../Icon/icons/LocationA.js';
import '../Icon/icons/LocationB.js';
import '../Icon/icons/LocationRx.js';
import '../Icon/icons/Location.js';
import '../Icon/icons/Locked.js';
import '../Icon/icons/LockerPasscode.js';
import '../Icon/icons/LowFat.js';
import '../Icon/icons/Map.js';
import '../Icon/icons/MealKit.js';
import '../Icon/icons/MeatSeafood.js';
import '../Icon/icons/MedicationToolkit.js';
import '../Icon/icons/Members.js';
import '../Icon/icons/Merge.js';
import '../Icon/icons/MilitaryPoliceFire.js';
import '../Icon/icons/Minus.js';
import '../Icon/icons/MissingItem.js';
import '../Icon/icons/MoneyServices.js';
import '../Icon/icons/Moon.js';
import '../Icon/icons/MoreOptions.js';
import '../Icon/icons/Natural.js';
import '../Icon/icons/Newest.js';
import '../Icon/icons/NoFees.js';
import '../Icon/icons/NonGmo.js';
import '../Icon/icons/NotAvailable.js';
import '../Icon/icons/Notification.js';
import '../Icon/icons/NuroDeliver.js';
import '../Icon/icons/NuroPickup.js';
import '../Icon/icons/NutFree.js';
import '../Icon/icons/Offline.js';
import '../Icon/icons/OnlinePayment.js';
import '../Icon/icons/OptUp.js';
import '../Icon/icons/Organic.js';
import '../Icon/icons/OtherMisc.js';
import '../Icon/icons/Oven.js';
import '../Icon/icons/Paleo.js';
import '../Icon/icons/Partnership.js';
import '../Icon/icons/Paste.js';
import '../Icon/icons/PatientAccess.js';
import '../Icon/icons/PatientWellness.js';
import '../Icon/icons/PaymentSuccess.js';
import '../Icon/icons/Pencil.js';
import '../Icon/icons/PersonHome.js';
import '../Icon/icons/PersonStore.js';
import '../Icon/icons/PersonalCare.js';
import '../Icon/icons/PetMedications.js';
import '../Icon/icons/PharmacyNotifications.js';
import '../Icon/icons/PharmacySearch.js';
import '../Icon/icons/Pharmacy.js';
import '../Icon/icons/Phone.js';
import '../Icon/icons/Pickup.js';
import '../Icon/icons/PieChart.js';
import '../Icon/icons/PiggyBank.js';
import '../Icon/icons/PillIdentifier.js';
import '../Icon/icons/PlusCard.js';
import '../Icon/icons/Plus.js';
import '../Icon/icons/PoorService.js';
import '../Icon/icons/PoorSubstitution.js';
import '../Icon/icons/Popularity.js';
import '../Icon/icons/PreferredCard.js';
import '../Icon/icons/PrescriptionHistory.js';
import '../Icon/icons/PrescriptionStatus.js';
import '../Icon/icons/Print.js';
import '../Icon/icons/PrivacyPolicy.js';
import '../Icon/icons/Produce.js';
import '../Icon/icons/PromotionOnAd.js';
import '../Icon/icons/PromotionOnDisplay.js';
import '../Icon/icons/Promotions.js';
import '../Icon/icons/Publish.js';
import '../Icon/icons/PurchaseHistory.js';
import '../Icon/icons/PushNotification.js';
import '../Icon/icons/PushPin.js';
import '../Icon/icons/PushPinSolid.js';
import '../Icon/icons/PuzzlePiece.js';
import '../Icon/icons/QrCode.js';
import '../Icon/icons/RadioChecked.js';
import '../Icon/icons/RadioUnchecked.js';
import '../Icon/icons/RatingsHalf.js';
import '../Icon/icons/RatingsSolid.js';
import '../Icon/icons/Ratings.js';
import '../Icon/icons/ReWeigh.js';
import '../Icon/icons/React.js';
import '../Icon/icons/Recall.js';
import '../Icon/icons/Receipt.js';
import '../Icon/icons/RecentSearch.js';
import '../Icon/icons/Recipes.js';
import '../Icon/icons/Recycle.js';
import '../Icon/icons/Redo.js';
import '../Icon/icons/RefillFromList.js';
import '../Icon/icons/RefillPrograms.js';
import '../Icon/icons/RejectEntry.js';
import '../Icon/icons/Relevance.js';
import '../Icon/icons/RemoveColumn.js';
import '../Icon/icons/Repurchase.js';
import '../Icon/icons/Reset.js';
import '../Icon/icons/Restrooms.js';
import '../Icon/icons/Returns.js';
import '../Icon/icons/Revert.js';
import '../Icon/icons/Rewards.js';
import '../Icon/icons/RightArrow.js';
import '../Icon/icons/RxBarcode.js';
import '../Icon/icons/Route.js';
import '../Icon/icons/Sandwich.js';
import '../Icon/icons/SaveTime.js';
import '../Icon/icons/Scan.js';
import '../Icon/icons/ScanBagGo.js';
import '../Icon/icons/ScheduleAppointment.js';
import '../Icon/icons/Search.js';
import '../Icon/icons/SecurityQuestions.js';
import '../Icon/icons/SecurityTag.js';
import '../Icon/icons/Send.js';
import '../Icon/icons/Settings.js';
import '../Icon/icons/Share.js';
import '../Icon/icons/ShipToHome.js';
import '../Icon/icons/Ship.js';
import '../Icon/icons/ShopInStore.js';
import '../Icon/icons/ShopOnline.js';
import '../Icon/icons/Shop.js';
import '../Icon/icons/ShoppingList.js';
import '../Icon/icons/SkipToEnd.js';
import '../Icon/icons/SkipToStart.js';
import '../Icon/icons/SocialServices.js';
import '../Icon/icons/Sort.js';
import '../Icon/icons/SortAz.js';
import '../Icon/icons/SortZa.js';
import '../Icon/icons/SoyFree.js';
import '../Icon/icons/Status.js';
import '../Icon/icons/Stencil.js';
import '../Icon/icons/Stethoscope.js';
import '../Icon/icons/Stock.js';
import '../Icon/icons/StoreLocker.js';
import '../Icon/icons/Store.js';
import '../Icon/icons/Success.js';
import '../Icon/icons/Sun.js';
import '../Icon/icons/Supercharged.js';
import '../Icon/icons/TableView.js';
import '../Icon/icons/TermsConditions.js';
import '../Icon/icons/ThumbsDown.js';
import '../Icon/icons/ThumbsUp.js';
import '../Icon/icons/Transfer.js';
import '../Icon/icons/TransferCircle.js';
import '../Icon/icons/TransferCircleSolid.js';
import '../Icon/icons/Trash.js';
import '../Icon/icons/Trending.js';
import '../Icon/icons/Undo.js';
import '../Icon/icons/Unlocked.js';
import '../Icon/icons/UpArrow.js';
import '../Icon/icons/Upload.js';
import '../Icon/icons/VaccinationsMadeEasy.js';
import '../Icon/icons/Vaccinations.js';
import '../Icon/icons/Value.js';
import '../Icon/icons/Vegan.js';
import '../Icon/icons/Vegetarian.js';
import '../Icon/icons/Voice.js';
import '../Icon/icons/Wallet.js';
import '../Icon/icons/Warning.js';
import '../Icon/icons/WeeklyAd.js';
import '../Icon/icons/Weigh.js';
import '../Icon/icons/WorldwideAcceptance.js';
import '../Icon/icons/WrongItem.js';
import '../Icon/icons/YouthPrograms.js';
import '../Icon/icons/YtdSavings.js';
import '../Icon/icons/ZoomIn.js';
import '../Icon/icons/ZoomOut.js';
import { callSafely } from '../../utils/functionUtils.js';
import { withRouter } from 'react-router-dom';
import Link from '../Link/Link.js';
import Button from '../Button/Button.js';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import useOnResize from '../../hooks/useOnResize.js';
import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import { normalizePageLinks, generateVisiblePageLinks, getNumberOfDisplayedPageLinks, captureFocusAfterDisablingButton } from './_paginationFunctions.js';

/**
 * Pagination provides an interface for users to navigate multiple pages of content
 *
 * Pagination will attempt to 'display' as many page links as it can fit (up to 10) on one line,
 * responsively hiding any 'extra' links that cannot fit.
 * **/

var PaginationComponent = function PaginationComponent(_ref) {
  var className = _ref.className,
      maxNumberOfDisplayedPages = _ref.maxNumberOfDisplayedPages,
      numberOfPages = _ref.numberOfPages,
      forcePage = _ref.forcePage,
      onPageChange = _ref.onPageChange,
      hrefBuilder = _ref.hrefBuilder,
      disableNextButton = _ref.disableNextButton,
      withRouter = _ref.withRouter,
      history = _ref.history,
      props = _objectWithoutProperties(_ref, ["className", "maxNumberOfDisplayedPages", "numberOfPages", "forcePage", "onPageChange", "hrefBuilder", "disableNextButton", "withRouter", "history"]);

  var _useState = useState(1),
      _useState2 = _slicedToArray(_useState, 2),
      currentPage = _useState2[0],
      setCurrentPage = _useState2[1];

  var _useState3 = useState(1),
      _useState4 = _slicedToArray(_useState3, 2),
      numberOfDisplayedPages = _useState4[0],
      setNumberOfDisplayedPages = _useState4[1];

  var pagination = useRef();
  var prevButton = useRef();
  var nextButton = useRef();
  var pageLinks = normalizePageLinks(generateVisiblePageLinks(numberOfPages, numberOfDisplayedPages, currentPage));
  useOnResize({
    onResize: setLinks,
    debounceDuration: 200
  });
  useEffect(setLinks, []);
  useEffect(function () {
    if (forcePage) {
      setPage(forcePage);
    }
  }, [forcePage]);

  function setPage(page) {
    /* Snap page within range of valid pages, 1 to (numberOfPages or Infinity) */
    var clampedPage = numberOfPages ? Math.min(numberOfPages, Math.max(1, page)) : Math.max(1, page);
    setCurrentPage(clampedPage);
  }

  function changePage(page, e) {
    callSafely(onPageChange)(page, e);
    setPage(page);
  }

  function navigateToPage(page, e) {
    var _getWindow = getWindow(),
        location = _getWindow.location;

    var href = callSafely(hrefBuilder)(page);

    if (!e.defaultPrevented) {
      if (withRouter && history) {
        history.push(href);
      } else if (location) {
        location.assign(href);
      }
    }
  }

  function setLinks() {
    if (pagination.current) {
      /* Total space the links are able to occupy, uses parent container width does not factor in padding */
      var containerWidth = pagination.current.parentNode.getBoundingClientRect().width;
      /* Width of the prev/next button */

      var buttonWidth = prevButton.current ? prevButton.current.getBoundingClientRect().width : 32;
      var numberOflinks = getNumberOfDisplayedPageLinks(numberOfPages, containerWidth, buttonWidth);
      setNumberOfDisplayedPages(numberOflinks);
    }
  }

  return React.createElement("nav", _extends({
    ref: pagination,
    "aria-label": "Pagination",
    className: cx('kds-Pagination', 'justify-center', className)
  }, props), React.createElement(Button, {
    ref: prevButton,
    className: "kds-Pagination-prev",
    "aria-label": "Previous page",
    kind: "tertiary",
    compact: true,
    disabled: currentPage <= 1,
    onClick: function onClick(e) {
      var prevPage = currentPage - 1;
      changePage(prevPage, e);
      navigateToPage(prevPage, e);
      captureFocusAfterDisablingButton(prevButton);
    }
  }, React.createElement(IconCaretLeft, null)), pageLinks.reduce(function (acc, page, i) {
    var isCurrentPage = currentPage === page;
    /* Only show overlay when multiple pages are visible */

    var showCurrentPageOverlay = isCurrentPage && pageLinks.length > 1;
    /* Distance between adjacent page links */

    var pageLinkGap = pageLinks[i + 1] - page;
    /* Show ellipsis when 'multiple' pages are skipped */

    var appendEllipsis = pageLinkGap > 1;
    return [].concat(_toConsumableArray(acc), [React.createElement(Link, {
      key: page,
      className: "kds-Pagination-link",
      "aria-current": isCurrentPage ? 'page' : null,
      "aria-label": "Page ".concat(page),
      inverse: showCurrentPageOverlay,
      onClick: function onClick(e) {
        return changePage(page, e);
      },
      href: callSafely(hrefBuilder)(page),
      withRouter: withRouter,
      implied: true
    }, page), appendEllipsis && React.createElement(IconEllipsis, {
      key: "".concat(page, "-ellipsis"),
      className: "kds-Pagination-ellipsis"
    })]);
  }, []), React.createElement(Button, {
    ref: nextButton,
    className: "kds-Pagination-next",
    "aria-label": "Next page",
    kind: "tertiary",
    compact: true,
    disabled: numberOfPages ? currentPage >= numberOfPages : disableNextButton,
    onClick: function onClick(e) {
      var nextPage = currentPage + 1;
      changePage(nextPage, e);
      navigateToPage(nextPage, e);
      captureFocusAfterDisablingButton(nextButton);
    }
  }, React.createElement(IconCaretRight, null)));
};

var Pagination = function Pagination(props) {
  if (props.withRouter) {
    var PaginationWithRouter = withRouter(PaginationComponent);
    return React.createElement(PaginationWithRouter, props);
  } else {
    return React.createElement(PaginationComponent, props);
  }
};

Pagination.defaultProps = {
  disableNextButton: false,
  withRouter: false
};
var Pagination$1 = React.memo(Pagination);

export default Pagination$1;
