// kds-react/build/cjs/components/Tabs/Tabs.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var generationUtils = require('../../utils/generationUtils.js');
var numberUtils = require('../../utils/numberUtils.js');

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}
/**
 * Use tabs to organize related content into multiple views.
 * Tabs are useful for switching between a small number of views.
 * For example, switching between products, coupons and recipes in a search results page.
 *
 ```Message { "props" : { "className" : "mb-16" } }
 * Note: Tabs should not be used when displaying a single Tab.
 ```
 */

var Tabs =
/*#__PURE__*/
function (_Component) {
  _inherits(Tabs, _Component);

  var _super = _createSuper(Tabs);

  function Tabs() {
    var _this;

    _classCallCheck(this, Tabs);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "uniqueId", generationUtils.generateUniqueString());

    _defineProperty(_assertThisInitialized(_this), "tabEls", {});

    _defineProperty(_assertThisInitialized(_this), "state", {
      selectedIndex: _this.props.defaultSelected,
      loadedTabs: _this.props.tabs.map(function (_ref, index) {
        var _ref$lazyLoad = _ref.lazyLoad,
            lazyLoad = _ref$lazyLoad === void 0 ? true : _ref$lazyLoad;
        return _this.props.defaultSelected === index || !lazyLoad;
      })
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyDown", function (event) {
      switch (event.key) {
        case 'ArrowUp':
        case 'ArrowLeft':
          event.preventDefault();

          _this.selectTab(_this.state.selectedIndex - 1);

          break;

        case 'ArrowDown':
        case 'ArrowRight':
          event.preventDefault();

          _this.selectTab(_this.state.selectedIndex + 1);

          break;

        case 'Home':
          event.preventDefault();

          _this.selectTab(0);

          break;

        case 'End':
          event.preventDefault();

          _this.selectTab(_this.props.tabs.length - 1);

          break;
      }
    });

    _defineProperty(_assertThisInitialized(_this), "selectTab", function (index) {
      if (index === _this.state.selectedIndex) {
        return;
      }

      var newIndex = numberUtils.wrappingModulus(index, _this.props.tabs.length);

      var loadedState = _this.state.loadedTabs.map(function (loaded, loadedIndex) {
        return loadedIndex === newIndex || loaded;
      });

      _this.setState({
        selectedIndex: newIndex,
        loadedTabs: loadedState
      }, function () {
        _this.tabEls[_this.state.selectedIndex].focus();

        _this.props.onChange(_this.state.selectedIndex);
      });
    });

    return _this;
  }

  _createClass(Tabs, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          tabs = _this$props.tabs,
          tabsClass = _this$props.tabsClass,
          contentClass = _this$props.contentClass;
      var selectedIndex = this.state.selectedIndex;
      return React__default.createElement(React__default.Fragment, null, React__default.createElement("div", {
        role: "tablist",
        className: cx('kds-Tabs', tabsClass),
        onKeyDown: this.handleKeyDown
      }, tabs.map(function (_ref2, index) {
        var tab = _ref2.tab;
        return React__default.createElement("button", {
          id: "Tabs-tab-".concat(_this2.uniqueId, "-").concat(index),
          key: index,
          ref: function ref(el) {
            return _this2.tabEls[index] = el;
          },
          role: "tab",
          tabIndex: index === selectedIndex ? 0 : -1,
          "aria-selected": index === selectedIndex ? 'true' : 'false',
          "aria-controls": "Tabs-content-".concat(_this2.uniqueId, "-").concat(index),
          className: cx('kds-Tabs-tab', {
            selected: index === selectedIndex
          }),
          onClick: function onClick() {
            _this2.selectTab(index);
          }
        }, tab);
      })), tabs.map(function (_ref3, index) {
        var content = _ref3.content;
        return content && _this2.state.loadedTabs[index] ? React__default.createElement("section", {
          id: "Tabs-content-".concat(_this2.uniqueId, "-").concat(index),
          key: index,
          role: "tabpanel",
          "aria-labelledby": "Tabs-tab-".concat(_this2.uniqueId, "-").concat(index),
          hidden: index !== selectedIndex,
          className: cx('kds-Tabs-content', contentClass)
        }, content) : null;
      }));
    }
  }]);

  return Tabs;
}(React.Component);

Tabs.defaultProps = {
  tabs: [],
  defaultSelected: 0,
  onChange: function onChange() {}
};

exports.default = Tabs;
