// kds-react/build/cjs/components/Heading/Heading.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var React = require('react');
var React__default = _interopDefault(React);
var cx = _interopDefault(require('classnames'));

var sizeOpts = ['xl', 'l', 'm', 's'];
var defaultTags = {
  xl: 'h1',
  l: 'h1',
  m: 'h2',
  s: 'h3'
};

var checkSize = function checkSize(size) {
  return sizeOpts.includes(size) ? size : 'l';
};

var Heading = function Heading(_ref) {
  var size = _ref.size,
      Tag = _ref.tag,
      hProps = _objectWithoutProperties(_ref, ["size", "tag"]);

  size = checkSize(size);
  Tag = Tag || defaultTags[size];
  return React__default.createElement(Tag, _extends({}, hProps, {
    className: cx("kds-Heading", "kds-Heading--".concat(size), hProps.className)
  }));
};

Heading.defaultProps = {
  size: 'l'
};
var Heading$1 = React__default.memo(Heading);

exports.default = Heading$1;
exports.sizeOpts = sizeOpts;
