// kds-react/tokens.js
export const breakpointS = "37.5em";
export const breakpointM = "56.25em";
export const breakpointL = "75em";
export const bodyXs = "0.625rem";
export const bodyS = "0.750rem";
export const bodyM = "0.844rem";
export const bodyL = "1.00rem";
export const headingS = "1.00rem";
export const headingM = "1.125rem";
export const headingL = "1.250rem";
export const headingXl = "2.00rem";
export const shadow0 = "none";
export const shadow2 = "0 2px 2px 0 #00000024, 0 3px 1px -2px #0000001f, 0 1px 5px 0 #00000033";
export const shadow4 = "0 4px 5px 0 #00000024, 0 1px 10px 0 #0000001f, 0 2px 4px -1px #00000033";
export const shadow8 = "0 8px 10px 1px #00000024, 0 3px 14px 2px #0000001f, 0 5px 5px -3px #00000033";
export const shadow12 = "0 12px 7px 2px #00000024, 0 5px 22px 4px #0000001f, 0 7px 8px -4px #00000033";
export const shadow16 = "0 16px 24px 2px #00000024, 0 6px 30px 5px #0000001f, 0 8px 10px -5px #00000033";
export const shadow24 = "0 24px 38px 3px #00000024, 0 9px 46px 8px #0000001f, 0 11px 15px -7px #00000033";
export const fontBase = "-apple-system, BlinkMacSystemFont, Roboto, Segoe UI, system-ui, Helvetica Neue, Ubuntu, sans-serif";
export const fontSecondary = "Nunito, -apple-system, BlinkMacSystemFont, Roboto, Segoe UI, system-ui, Helvetica Neue, Ubuntu, sans-serif";
export const fontMono = "Menlo, Monaco, Consolas, Liberation Mono, Courier New, monospace";
export const fontWeightNormal = "400";
export const fontWeightMedium = "500";
export const fontWeightBold = "700";
export const leadingBodyXs = "1.6";
export const leadingBodyS = "1.666666666666667";
export const leadingBodyM = "1.481481481481481";
export const leadingBodyL = "1.5";
export const leadingHeadingS = "1.25";
export const leadingHeadingM = "1.111111111111111";
export const leadingHeadingL = "1.2";
export const leadingHeadingXl = "1.25";
export const maxLineLengthNarrowViewport = "24.25rem";
export const minLineLengthWideViewport = "17.75rem";
export const maxLineLengthWideViewport = "33rem";
export const borderStyleDefault = "solid";
export const borderStyleCallout = "dashed";
export const borderWidthThin = "1px";
export const borderWidthThick = "2px";
export const borderRadiusNone = "0";
export const borderRadiusExtraSmall = "0.125rem";
export const borderRadiusSmall = "0.25rem";
export const borderRadiusMedium = "0.50rem";
export const borderRadiusLarge = "1.00rem";
export const borderRadiusExtraLarge = "1.25rem";
export const borderRadiusFull = "625rem";
export const zIndexPageCrucial = "5";
export const zIndexPageAlerts = "4";
export const zIndexPageModals = "3";
export const zIndexPagePopovers = "2";
export const zIndexPageNavs = "1";
export const zIndexStackHighest = "3";
export const zIndexStackHigher = "2";
export const zIndexStackHigh = "1";
export const zIndexStackBase = "0";
export const zIndexStackBelow = "-1";
export const colorBakersPrimary = "#ed3024";
export const colorCitymarketPrimary = "#ed3024";
export const colorDillonsPrimary = "#ca1626";
export const colorFredmeyerPrimary = "#d9272e";
export const colorFrysfoodPrimary = "#e2231a";
export const colorGerbesPrimary = "#ca1626";
export const colorJaycfoodsPrimary = "#d72a28";
export const colorKingsoopersPrimary = "#ed3024";
export const colorMetromarketPrimary = "#678339";
export const colorOwensmarketPrimary = "#ec1c24";
export const colorOptupPrimary = "#95c93f";
export const colorPayLessPrimary = "#ec1c24";
export const colorPharmacyPrimary = "#238dc1";
export const colorPicknsavePrimary = "#68944f";
export const colorRalphsPrimary = "#eb0029";
export const colorSmithsPrimary = "#d21245";
export const colorKrogerPrimary = "#084999";
export const colorKrogerSea = "#4471b7";
export const colorKrogerSky = "#418fde";
export const colorKrogerIce = "#8dc8e8";
export const colorKrogerStone = "#b3d7ed";
export const colorBrandPrimary = "#084999";
export const colorDefault900 = "#1d1e1f";
export const colorDefault800 = "#3f4041";
export const colorDefault700 = "#5d5e60";
export const colorDefault600 = "#727374";
export const colorDefault500 = "#9b9c9d";
export const colorDefault400 = "#bbbcbe";
export const colorDefault300 = "#dfe1e1";
export const colorDefault200 = "#eff0f1";
export const colorDefault100 = "#f8f9f9";
export const colorDefault50 = "#ffffff";
export const colorBrand900 = "#084999";
export const colorBrand800 = "#1567b8";
export const colorBrand700 = "#1b77ca";
export const colorBrand600 = "#2389dc";
export const colorBrand500 = "#2997ea";
export const colorBrand400 = "#47a6ed";
export const colorBrand300 = "#67b5ef";
export const colorBrand200 = "#91caf4";
export const colorBrand100 = "#bbdef8";
export const colorBrand50 = "#e3f2fc";
export const colorNeutral900 = "#385397";
export const colorNeutral800 = "#4471b7";
export const colorNeutral700 = "#4b81c9";
export const colorNeutral600 = "#5493dc";
export const colorNeutral500 = "#5ba1ea";
export const colorNeutral400 = "#6aafed";
export const colorNeutral300 = "#80bdf0";
export const colorNeutral200 = "#a0d0f6";
export const colorNeutral100 = "#c3e2f9";
export const colorNeutral50 = "#e6f3fc";
export const colorAction900 = "#385397";
export const colorAction800 = "#4471b7";
export const colorAction700 = "#4b81c9";
export const colorAction600 = "#5493dc";
export const colorAction500 = "#5ba1ea";
export const colorAction400 = "#6aafed";
export const colorAction300 = "#80bdf0";
export const colorAction200 = "#a0d0f6";
export const colorAction100 = "#c3e2f9";
export const colorAction50 = "#e6f3fc";
export const colorSpecial900 = "#492d7c";
export const colorSpecial800 = "#64378c";
export const colorSpecial700 = "#733d97";
export const colorSpecial600 = "#8444a0";
export const colorSpecial500 = "#9149a8";
export const colorSpecial400 = "#a15eb5";
export const colorSpecial300 = "#b178c3";
export const colorSpecial200 = "#c79dd5";
export const colorSpecial100 = "#ddc4e5";
export const colorSpecial50 = "#f1e7f4";
export const colorCallout900 = "#ff6700";
export const colorCallout800 = "#ff8900";
export const colorCallout700 = "#ff9b00";
export const colorCallout600 = "#feae00";
export const colorCallout500 = "#febd00";
export const colorCallout400 = "#fec600";
export const colorCallout300 = "#fed241";
export const colorCallout200 = "#ffde7a";
export const colorCallout100 = "#ffebaf";
export const colorCallout50 = "#fff7df";
export const colorNegative900 = "#ca0013";
export const colorNegative800 = "#d70022";
export const colorNegative700 = "#e4002a";
export const colorNegative600 = "#f61930";
export const colorNegative500 = "#ff282f";
export const colorNegative400 = "#ff444d";
export const colorNegative300 = "#f36c71";
export const colorNegative200 = "#f99699";
export const colorNegative100 = "#ffcbd3";
export const colorNegative50 = "#ffeaee";
export const colorPositive900 = "#006334";
export const colorPositive800 = "#018249";
export const colorPositive700 = "#0c9455";
export const colorPositive600 = "#18a662";
export const colorPositive500 = "#21b56c";
export const colorPositive400 = "#4ec183";
export const colorPositive300 = "#71cc98";
export const colorPositive200 = "#9cdab5";
export const colorPositive100 = "#c3e8d2";
export const colorPositive50 = "#e6f6ed";
