// kds-react/build/es/components/StatusIndicator/StatusIndicator.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React from 'react';
import cx from 'classnames';

var kindOpts = ['positive', 'callout', 'negative'];

var getValidKind = function getValidKind(kind) {
  return kindOpts.includes(kind) ? kind : 'positive';
};
/**
 * `<StatusIndicator>` provides context to the user on the status of an order, system process, or a datapoint within the UI.
 */


var StatusIndicator = function StatusIndicator(_ref) {
  var children = _ref.children,
      Tag = _ref.tag,
      className = _ref.className,
      kind = _ref.kind,
      props = _objectWithoutProperties(_ref, ["children", "tag", "className", "kind"]);

  return children ? React.createElement(Tag, _extends({}, props, {
    className: cx('kds-StatusIndicator', "kds-StatusIndicator--".concat(getValidKind(kind)), className)
  }), children) : '';
};

StatusIndicator.defaultProps = {
  kind: 'positive',
  tag: 'span'
};

export default StatusIndicator;
export { kindOpts };
