// kds-react/build/es/components/Heading/Heading.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Fri Nov 13 2020 21:58:49 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React from 'react';
import cx from 'classnames';

var sizeOpts = ['xl', 'l', 'm', 's'];
var defaultTags = {
  xl: 'h1',
  l: 'h1',
  m: 'h2',
  s: 'h3'
};

var checkSize = function checkSize(size) {
  return sizeOpts.includes(size) ? size : 'l';
};

var Heading = function Heading(_ref) {
  var size = _ref.size,
      Tag = _ref.tag,
      hProps = _objectWithoutProperties(_ref, ["size", "tag"]);

  size = checkSize(size);
  Tag = Tag || defaultTags[size];
  return React.createElement(Tag, _extends({}, hProps, {
    className: cx("kds-Heading", "kds-Heading--".concat(size), hProps.className)
  }));
};

Heading.defaultProps = {
  size: 'l'
};
var Heading$1 = React.memo(Heading);

export default Heading$1;
export { sizeOpts };
