// kds-react/build/cjs/components/Toast/ToastProvider.js
/* eslint-disable */

/***********************************************************************************************
  THIS IS AN AUTO-GENERATED FILE. DO NOT MAKE CHANGES HERE

  Generated on: Wed Dec 09 2020 21:14:13 GMT+0000 (Coordinated Universal Time)
************************************************************************************************/

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var _classCallCheck = _interopDefault(require('@babel/runtime/helpers/classCallCheck'));
var _createClass = _interopDefault(require('@babel/runtime/helpers/createClass'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inherits = _interopDefault(require('@babel/runtime/helpers/inherits'));
var _possibleConstructorReturn = _interopDefault(require('@babel/runtime/helpers/possibleConstructorReturn'));
var _getPrototypeOf = _interopDefault(require('@babel/runtime/helpers/getPrototypeOf'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var windowUtils = require('../../utils/windowUtils.js');
var generationUtils = require('../../utils/generationUtils.js');
var reactTransitionGroup = require('react-transition-group');
var Toast = require('./Toast.js');
var _ToastContext = require('./_ToastContext.js');
var _Portal = require('../Portal/_Portal.js');

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}
/** See [`Toast`](./toast) documentation for further details. */

var ToastProvider =
/*#__PURE__*/
function (_Component) {
  _inherits(ToastProvider, _Component);

  var _super = _createSuper(ToastProvider);

  function ToastProvider(props) {
    var _this;

    _classCallCheck(this, ToastProvider);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "timeoutId", null);

    _defineProperty(_assertThisInitialized(_this), "add", function (content) {
      var id = generationUtils.generateUniqueString();

      if (_this.state.toasts.length > 0) {
        _this.remove(_this.state.toasts[0] && _this.state.toasts[0].id);
      }

      _this.setState(function (state) {
        var toasts = state.toasts.slice();
        var toast = {
          content: content,
          id: id
        };
        toasts.push(toast);
        return {
          toasts: toasts
        };
      });

      _this.timeoutId = setTimeout(_this.onDismiss(id), _this.readTime(content['text']));
    });

    _defineProperty(_assertThisInitialized(_this), "readTime", function (str) {
      var words = str.split(' ').length;
      return 5000 + words * 300;
    });

    _defineProperty(_assertThisInitialized(_this), "remove", function (id) {
      clearTimeout(_this.timeoutId);

      _this.setState(function (state) {
        var toasts = state.toasts.filter(function (t) {
          return t.id !== id;
        });
        return {
          toasts: toasts
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onDismiss", function (id) {
      return function () {
        _this.remove(id);
      };
    });

    _this.state = {
      toasts: []
    };
    return _this;
  }
  /** The `add` function can be wrapped around an element that will trigger a `Toast` message. */


  _createClass(ToastProvider, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var add = this.add;
      var toasts = this.state.toasts;
      return React__default.createElement(_ToastContext.Provider, {
        value: add
      }, this.props.children, windowUtils.canUseDOM() ? React__default.createElement(_Portal.default, {
        id: "toast"
      }, React__default.createElement(reactTransitionGroup.TransitionGroup, {
        className: "kds-ToastGroup"
      }, toasts.map(function (_ref) {
        var content = _ref.content,
            id = _ref.id;
        return React__default.createElement(reactTransitionGroup.CSSTransition, {
          key: id,
          timeout: 400,
          classNames: {
            enter: 'isVisible',
            exit: 'isRemoved'
          }
        }, React__default.createElement(Toast.default, {
          kind: content.kind,
          heading: content.heading,
          onClose: _this2.onDismiss(id)
        }, content.text));
      }))) : null);
    }
  }]);

  return ToastProvider;
}(React.Component);

exports.default = ToastProvider;
